---
title: "SCAHN figures"
output: html_document
date: "2025-04-28"
author: "Hong Zheng"
editor_options: 
  chunk_output_type: inline
---

```{r packages}
library(data.table)
library(effsize)
library(DESeq2)
library(ggalluvial)
library(ggpubr)
library(ggrastr)
library(ggtext)
library(ggsci)
library(ggrepel)
library(ggforce)
library(pBrackets)
library(magick)
library(ComplexHeatmap)
library(rmeta)
library(circlize)
library(viridis)
library(gridExtra)
library(scales)
library(cowplot)
library(fgsea)
library(dplyr)
library(ggbio)
library(survival)
library(ggfortify)
```

```{r functions}
geomMean <- function(x, na.rm = T){
  if (!is.numeric(x) && !is.complex(x) && !is.logical(x)) {
    warning("argument is not numeric or logical: returning NA")
    return(as.numeric(NA))
  }
  if (na.rm)
    x <- x[!is.na(x)]
  if (any(x < 0))
    stop("'x' contains negative value(s)")
  
  ### direct method causes overflow errors, use log method instead
  ### return(prod(x)^(1/length(x)))
  return(exp(sum(log(x[x>0]))/length(x)))
}

getGeneScores <- function(geneMtx, pos, neg, makePos = T, out.missing = T){
  if(out.missing){
    genes=unique(c(pos,neg))
    #missingpos = pos[!(pos %in% rownames(geneMtx))]
    #missingneg = neg[!(neg %in% rownames(geneMtx))]
    #missing = unique(c(missingpos,missingneg))
    #missing=genes[!(genes %in% rownames(geneMtx))]
    missing=setdiff(genes,rownames(geneMtx))
    #if(length(missing)>=1){
    cat("Missing these genes:",missing,"\n")
    #}
  }
  pos = pos[pos %in% rownames(geneMtx)]
  neg = neg[neg %in% rownames(geneMtx)]
  
  if(makePos){ #If I need to make everything positive
    if(any(geneMtx < 0, na.rm=T)){
      geneMtx <- geneMtx - min(geneMtx, na.rm=T)
    }
  }
  
  if(length(neg)>=1 && length(pos)>=1){
    scores <- apply(geneMtx[pos, , drop=F], 2, geomMean, na.rm=T) - apply(geneMtx[neg, , drop=F], 2, geomMean, na.rm=T)
  }else if(length(pos)>=1){
    scores <- apply(geneMtx[pos, , drop=F], 2, geomMean, na.rm=T)
  }else if(length(neg)>=1){
    scores <- -1*apply(geneMtx[neg, , drop=F], 2, geomMean, na.rm=T)
  }
  
  if(any(is.nan(scores))){ #this means all upgenes or all downgenes were missing from some samples
    nanIndex = which(is.nan(scores))
    for(i in nanIndex){
      my.score = c(geomMean(geneMtx[pos,i,drop=F],na.rm=T),geomMean(geneMtx[neg,i,drop=F],na.rm=T))
      my.score = my.score[!is.nan(my.score)]
      if(length(my.score) == 1){
        scores[i] = my.score
      }else if(length(my.score) == 2){
        scores[i] = my.score[1] - my.score[2]
      }
    }
  }
  
  return(scores)
}

ScatterPlotRast <- function(df, featureX, featureY, groupby = NA, colorby = NA, 
                            label = NA, labeldf = data.frame(), labelfont = "plain", labelsize = 2, repel = F,
                            mycol = NA, colcateg = "continuous", backgroudcol = "grey97", panelbackgroundcol = "white",nacol = "grey87", 
                            linesize = 0.2, pointsize = 0.5, rasterdpi = 100, alphalevel = 0.4, 
                            legendpointsize = 2, legendColumn = 1, legendposition = "right", legendtitle = NA, legendsize = 1, 
                            numberofrows = 1, facetdirection = "h", scale = "free", facetlabel = "all", facetlabelface = "italic", 
                            xlab = "", ylab = "", plottitle = "", showaxis = F, showgrid = F, 
                            textsize = 10, legendtextsize = 10, axistextsize = 10, 
                            contvarlower = 0.01, contvarupper = 0.99, seed = 42, shuffle = T, addstat = F)
{
  textcolor = "#252525"; linecolor = "#252525"
    
  if(!is.na(mycol[1]) & colcateg == "discrete"){
    df[, colorby] = factor(df[, colorby], levels = names(mycol))
  }
  
  if(is.na(mycol[1]) & colcateg == "continuous"){
    mycol = dichromat_pal("DarkRedtoBlue.12")(12)
  }
  
  if(colcateg == "continuous"){
    n1 = quantile(df[, colorby], contvarlower)
    n2 = quantile(df[, colorby], contvarupper)
    df[, colorby][which(df[, colorby] >= n2)] = n2
    df[, colorby][which(df[, colorby] <= n1)] = n1
  }
  
  if(shuffle == T){
    set.seed(seed)
    df = df[sample(1:nrow(df)),]}
  
  if(is.na(groupby)){df$groupdummy = facetlabel}else{df$groupdummy = df[, groupby]}
  
  theme_plot = theme(text = element_text(size = textsize, color = textcolor),
                     panel.grid.minor = element_blank(),
                     panel.grid.major =element_line(color = "grey87", size = 0.05, linetype = 2),
                     panel.border = element_blank(),
                     legend.position = legendposition,
                     legend.title = element_blank(),
                     legend.background =  element_blank(),
                     legend.box.background =  element_blank(),
                     legend.box.margin = margin(0, 0, 0, 0, "null"),
                     legend.margin = margin(0, 0, 0, 0, "null"),
                     legend.justification = c(0, 0.5),
                     legend.text = element_text(size = legendtextsize, color = textcolor),
                     legend.key.width = unit(0.3, "line"), legend.key.height = unit(0.5, "line"),
                     axis.line = element_line(color = linecolor, size = linesize), 
                     axis.ticks = element_line(color = linecolor, size = linesize),
                     axis.text =  element_text(size = axistextsize, color = textcolor),
                     axis.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "null")),
                     plot.title = element_text(size = textsize, color = textcolor, face = "italic", margin = margin(0, 0, 0, 0, "null")),
                     plot.background = element_rect(color = backgroudcol, fill = backgroudcol),
                     panel.background = element_rect(color = panelbackgroundcol, fill= panelbackgroundcol),
                     strip.background = element_rect(color = NA, fill = "grey91"),
                     strip.text = element_text(size = textsize, face = facetlabelface, color = textcolor, 
                                               margin = margin(0.1, 0.1, 0.1, 0.1, "line"))
  )
  
  if(colcateg == "continuous"){
    p =
      ggplot(df, aes_string(x = featureX, y = featureY)) +
      theme_void() +
      geom_point_rast(aes_string(color = colorby), size = pointsize, stroke = 0, alpha = alphalevel, shape = 16, raster.dpi = rasterdpi) +
      scale_x_continuous(breaks = pretty_breaks()) +
      scale_y_continuous(breaks = pretty_breaks()) +
      facet_wrap(. ~ groupdummy, nrow = numberofrows, scales = scale, dir = facetdirection) +
      labs(title = plottitle) +
      theme_plot +
      theme(legend.key.width = unit(legendsize/10, "line"), legend.key.height = unit(legendsize, "line")) +
      scale_color_gradientn(colours = mycol, name = legendtitle, 
                            breaks = c(0, round(max(df[, colorby])/2.2,1), round(max(df[, colorby])/2.2,1)*2))
    
  }
  else{
    p=
      ggplot(df, aes_string(x = featureX, y = featureY)) +
      theme_void() +
      geom_point_rast(aes_string(color = colorby), size = pointsize, stroke = 0, alpha = alphalevel, shape = 16, raster.dpi = rasterdpi) +
      scale_x_continuous(breaks = pretty_breaks()) +
      scale_y_continuous(breaks = pretty_breaks()) +
      facet_wrap(. ~ groupdummy, nrow = numberofrows, scales = scale, dir = facetdirection) +
      labs(title = plottitle) +
      theme_plot +
      scale_color_manual(values = mycol, name = NULL, na.value = "grey93") +
      guides(color = guide_legend(ncol = legendColumn, label.hjust = 0, override.aes = list(size = legendpointsize, alpha = 0.7))) 
  }
  
  
  if(!is.na(label) & repel == T){
    p = p +
      geom_label_repel(data = df, aes(label=labeltext), size = labelsize, color = "#252525", force = 0.05, fontface = labelfont)
  }
  
  if(!is.na(label) & repel == F){
    p = p +
      geom_label(data = df, aes(label = labeltext), size = labelsize, color = "#252525", fontface = labelfont)
  }
  
  if(nrow(labeldf) > 0 & repel == T){
    p = p +
      geom_text_repel(data = labeldf, aes(x = x, y = y, label = labeltext),
                      size = labelsize, color = "#252525", force = 0.05, fontface = labelfont)
  }  
  
  if(nrow(labeldf) > 0  & repel == F){
    p = p +
      geom_text(data = labeldf, aes(x = x, y = y, label = labeltext),
                size = labelsize, color = "#252525", fontface = labelfont)
  }  
  
  if(addstat == T){
    p = p +
      stat_cor(method = "spearman", label.x.npc = 0, label.y.npc = 0.95, size = labelsize, color = "grey67")
  }
  
  if(showaxis == F){
    p = p + theme(axis.line = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())
  }
  
  if(showgrid == F){
    p = p + theme(panel.grid.major = element_blank())
  }
  
  return(p)
}

ScatterPlotRastNoFacet <- function(df, featureX, featureY, colorby = NA, 
                                   label = NA, labeldf = data.frame(), labelfont = "plain", labelsize = 2, repel = F,
                                   mycol = NA, colcateg = "continuous", backgroudcol = "grey97", nacol = "grey87", 
                                   linesize = 0.2, pointsize = 0.5, rasterdpi = 100, alphalevel = 0.4, 
                                   legendpointsize = 2, legendColumn = 1, legendposition = "right", legendtitle = NA, legendsize = 1, 
                                   numberofrows = 1, facetdirection = "h", scale = "free", facetlabel = "all", facetlabelface = "italic", 
                                   xlab = "", ylab = "", plottitle = "", showaxis = F, showgrid = F, 
                                   textsize = 10, legendtextsize = 10, axistextsize = 10, 
                                   contvarlower = 0.01, contvarupper = 0.99, seed = 42, shuffle = T, addstat = F)
{
  textcolor = "#252525"; linecolor = "#252525"
    
  if(!is.na(mycol[1]) & colcateg == "discrete"){
    df[, colorby] = factor(df[, colorby], levels = names(mycol))
  }
  
  if(is.na(mycol[1]) & colcateg == "continuous"){
    mycol = dichromat_pal("DarkRedtoBlue.12")(12)
  }
  
  if(colcateg == "continuous"){
    n1 = quantile(df[, colorby], contvarlower)
    n2 = quantile(df[, colorby], contvarupper)
    df[, colorby][which(df[, colorby] >= n2)] = n2
    df[, colorby][which(df[, colorby] <= n1)] = n1
  }
  
  if(shuffle == T){
    set.seed(seed)
    df = df[sample(1:nrow(df)),]}
  
  theme_plot = theme(text = element_text(size = textsize, color = textcolor),
                     panel.grid.minor = element_blank(),
                     panel.grid.major = element_line(color = "grey87", size = 0.05, linetype = 2),
                     panel.border = element_blank(),
                     legend.position = legendposition,
                     legend.title = element_blank(),
                     legend.background =  element_blank(),
                     legend.box.background =  element_blank(),
                     legend.box.margin = margin(0, 0, 0, 0, "null"),
                     legend.margin = margin(0, 0, 0, 0, "null"),
                     legend.justification = c(0, 0.5),
                     legend.text = element_text(size = legendtextsize, color = textcolor),
                     legend.key.width = unit(0.3, "line"), legend.key.height = unit(0.5, "line"),
                     axis.line = element_line(color = linecolor, size = linesize), 
                     axis.ticks = element_line(color = linecolor, size = linesize),
                     axis.text =  element_text(size = axistextsize, color = textcolor),
                     axis.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "null")),
                     plot.title = element_text(size = textsize, color = textcolor, face = "italic", margin = margin(0, 0, 0, 0, "null")),
                     plot.background = element_rect(color = backgroudcol, fill = backgroudcol),
                     strip.background = element_rect(color = NA, fill = "grey91"),
                     strip.text = element_text(size = textsize, face = facetlabelface, color = textcolor, 
                                               margin = margin(0.1, 0.1, 0.1, 0.1, "line"))
  )
  
  
  if(colcateg == "continuous"){
    p =
      ggplot(df, aes_string(x = featureX, y = featureY)) +
      theme_void() +
      geom_point_rast(aes_string(color = colorby), size = pointsize, stroke = 0, alpha = alphalevel, shape = 16, raster.dpi = rasterdpi) +
      scale_x_continuous(breaks = pretty_breaks()) +
      scale_y_continuous(breaks = pretty_breaks()) +
      labs(title = plottitle) +
      theme_plot +
      theme(legend.key.width = unit(legendsize/10, "line"), legend.key.height = unit(legendsize, "line")) +
      scale_color_gradientn(colours = mycol, name = legendtitle, 
                            breaks = c(0, round(max(df[, colorby])/2.2,1), round(max(df[, colorby])/2.2,1)*2))
  }
  else{
    p=
      ggplot(df, aes_string(x = featureX, y = featureY)) +
      theme_void() +
      geom_point_rast(aes_string(color = colorby), size = pointsize, stroke = 0, alpha = alphalevel, shape = 16, raster.dpi = rasterdpi) +
      scale_x_continuous(breaks = pretty_breaks()) +
      scale_y_continuous(breaks = pretty_breaks()) +
      labs(title = plottitle) +
      theme_plot +
      scale_color_manual(values = mycol, name = NULL, na.value = "grey93") +
      guides(color = guide_legend(ncol = legendColumn, label.hjust = 0, override.aes = list(size = legendpointsize, alpha = 0.7))) 
  }
  
  
  if(!is.na(label) & repel == T){
    p = p +
      geom_label_repel(data = df, aes(label=labeltext), size = labelsize, color = "#252525", force = 0.05, fontface = labelfont)
  }
  
  if(!is.na(label) & repel == F){
    p = p +
      geom_label(data = df, aes(label = labeltext), size = labelsize, color = "#252525", fontface = labelfont)
  }
  
  if(nrow(labeldf) > 0 & repel == T){
    p = p +
      geom_text_repel(data = labeldf, aes(x = x, y = y, label = labeltext),
                      size = labelsize, color = "#252525", force = 0.05, fontface = labelfont)
  }  
  
  if(nrow(labeldf) > 0  & repel == F){
    p = p +
      geom_text(data = labeldf, aes(x = x, y = y, label = labeltext),
                size = labelsize, color = "#252525", fontface = labelfont)
  }  
  
  if(addstat == T){
    p = p +
      stat_cor(method = "spearman", label.x.npc = 0, label.y.npc = 0.95, size = labelsize, color = "grey67")
  }
  
  if(showaxis == F){
    p = p + theme(axis.line = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())
  }
  
  if(showgrid == F){
    p = p + theme(panel.grid.major = element_blank())
  }
  
  return(p)
}

GetRownamesCombineDuplicate <- function(geneMtx, keys){
  if(length(keys) == length(unique(keys))){
    return(geneMtx)
  }
  dupvals = unique(keys[duplicated(keys)])
  duplicateindex = keys %in% dupvals
  duplicategenes = geneMtx[duplicateindex,,drop= F]
  uniquegenes = geneMtx[!duplicateindex,,drop= F]
  rownames(uniquegenes)= as.character(keys[!duplicateindex])
  
  newvals = do.call(rbind, lapply(dupvals, function(val){
    dupindex = which(keys[rownames(duplicategenes)] == val)
    dupmat = duplicategenes[dupindex, , drop = F]
    # replace NA values with the average of the other values in that column
    if(any(is.na(dupmat))){
      dupmat = apply(dupmat, 2, function(x){
        if(any(is.na(x))){
          avg = mean(x[!is.na(x)])
          x[is.na(x)] = avg
        }
        return(x)
      })
    }
    # average all rows
    return(colMeans(dupmat))
  }))
  
  rownames(newvals) = dupvals
  if(any(is.na(rownames(newvals)))){ #removing the row named "NA"
    NArow = which(is.na(rownames(newvals)))
    newvals = newvals[-NArow, ]
  }
  return(rbind(uniquegenes, newvals))
}

# score_comparison <- function(dat, cases, controls){
#   res = dat[, .(
#     pval = wilcox.test(value[condition %in% cases], value[condition %in%  controls])$p.value,
#     ES.g =  cohen.d(value[condition %in%cases], value[condition %in% controls], hedges.correction=TRUE)$estimate,
#     ES.se =  sqrt(cohen.d(value[condition %in%cases], value[condition %in% controls], hedges.correction=TRUE)$var),
#     Ncase = length(value[condition %in% cases]),
#     Ncontrol = length(value[condition %in% controls]),
#     samplesize = length(value[condition %in% c(cases, controls)])
#   ),
#   by = .(variable, pid)]
# }

score_comparison <- function(dat, cases, controls){
  res = dat[, {
    # Extract values for cases and controls
    case_values <- value[condition %in% cases]
    control_values <- value[condition %in% controls]
    
    # Check if the expression is invariable across both groups:
    if(length(unique(c(case_values, control_values))) == 1){
      # If invariable, assign neutral estimates
      list(pval = 1,
           ES.g = 0,
           ES.se = 1,
           Ncase = length(case_values),
           Ncontrol = length(control_values),
           samplesize = length(c(case_values, control_values)))
    } else {
      # Otherwise, compute as usual
      wt <- wilcox.test(case_values, control_values)
      cd <- cohen.d(case_values, control_values, hedges.correction=TRUE)
      list(pval = wt$p.value,
           ES.g = cd$estimate,
           ES.se = sqrt(cd$var),
           Ncase = length(case_values),
           Ncontrol = length(control_values),
           samplesize = length(c(case_values, control_values)))
    }
  }, by = .(variable, pid)]
  return(res)
}

metaStats <- function(dd, geneid, pids, case, control){
  dfes = score_comparison(dd[variable %in% geneid & pid %in% pids], 
                          cases = case, controls = control)
  dfes = dfes[order(variable)]
  dfes$rowname = paste0(dfes$pid, " (N1 = ",dfes$Ncase, ", N0 = ",dfes$Ncontrol, ")")
  dfes$index = nrow(dfes):1
  dfes = dfes[, c("ES.g", "ES.se", "samplesize", "pval", "rowname", "index")]
  dfes.summary = pool.inverseVar(t(data.frame(genes = dfes$ES.g)), t(data.frame(genes = dfes$ES.se)),method = "random")
  dfes.summary = as.data.frame(dfes.summary)
  dfes.summary$ES.g = dfes.summary$summary
  dfes.summary$ES.se = dfes.summary$se.summary
  dfes.summary$samplesize = NA
  dfes.summary$pval = dfes.summary$p.value
  dfes.summary$rowname = "summary"
  dfes.summary$index = 0
  dfes = rbind(dfes,dfes.summary[, c("ES.g", "ES.se", "samplesize", "pval", "rowname", "index")])
  dfes$gene = geneid
  dfes[, label:=ifelse(rowname == "summary", paste0("ES:", round(ES.g, 2), "\npval:", formatC(pval, format = "e", digits = 1)), NA)]
  return(dfes)
}

metaPlots<- function(datadf, boxsize = 1, conf.level = 0.95, mycol = "grey",
                     plottitle = NULL, titleface = "plain", stripxface = "plain",labex = 2, textsize = 8, summaryy = c(0, -0.4, 0, 0.4))
{
  nth <- function(x, i) {
    x[(i - 1)%%length(x) + 1]
  }
  ci.value <- -qnorm((1 - conf.level)/2)
  
  datadf$lower <- datadf$ES.g - ci.value * datadf$ES.se
  datadf$upper <- datadf$ES.g + ci.value * datadf$ES.se
  datadf$samplesize = datadf$samplesize / max(datadf$samplesize, na.rm = T)
  
  # summary
  if(!"summary" %in% rownames(datadf)){
    datadfsummary = datadf[nrow(datadf),]}
  else{
    datadfsummary = datadf[rowname == "summary"]}
  
  dfpolygon = data.table(x = c(datadfsummary$lower, datadfsummary$ES.g, datadfsummary$upper, datadfsummary$ES.g),
                         y = summaryy)
  
  datadf[nrow(datadf), c("lower", "upper", "samplesize") := list(NA, NA, NA)]
  
  # dfpolygon = data.table(x = c(datadf[rowname == "summary"]$lower, datadf[rowname == "summary"]$ES.g, datadf[rowname == "summary"]$upper, datadf[rowname == "summary"]$ES.g),
  #                        y = summaryy)
  # 
  # datadf[rowname == "summary"]$lower = NA
  # datadf[rowname == "summary"]$upper = NA
  # datadf[rowname == "summary"]$samplesize = NA
  
  p = ggplot(datadf, aes_string(x = "ES.g", y = "index")) +
    theme_bw() +
    geom_vline(xintercept = 0, linetype = 2, alpha = 1, color = "grey62", size = 0.4) +
    geom_hline(yintercept = c(-0.5, 0.5), linetype = 1, alpha = 1, color = textcolor, size = linesize) +
    geom_errorbarh(alpha = 0.8, height = 0.4, size = 0.4, aes_string(xmax = "upper", xmin = "lower"), color = mycol) +
    geom_tile(aes(width = samplesize / 20 * boxsize), height = 0.5, fill = mycol, color = NA) +
    geom_polygon(data = dfpolygon, aes_string(x = "x", y = "y"), fill = mycol, color = NA) +
    scale_y_continuous(position = "right", breaks = datadf$index, labels = datadf$rowname, limits = c(-0.51, max(datadf$index) + 0.51), expand = c(0,0)) +
    labs(x = "effect size", title = plottitle, y = NULL) +
    facet_wrap(. ~ gene, scales = "free_x", nrow = 1) +
    geom_text(data = data.frame(x = labex, y = 0, label = datadfsummary$label), mapping = aes(x = x, y = y, label = label), color = "grey17", size = 2, parse = F, hjust = 0, lineheight = 0.7) +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major.x =element_line(color = "grey67", size = 0.05, linetype = 2),
      panel.grid.major.y = element_blank(),
      panel.background  = element_rect(color = NA, size = 0, fill = "grey98"),
      panel.border = element_blank(),
      axis.ticks = element_blank(),
      axis.line = element_blank(),
      axis.text = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      axis.text.y = element_text(size = textsize, color = textcolor),
      plot.title = element_text(size = textsize, face = titleface, margin = margin(0, 0, 0, 0, "line")),
      axis.title = element_text(size = textsize, color = textcolor),
      axis.title.x = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "null")),
      axis.title.y = element_text(size = textsize, color = textcolor),
      strip.background = element_rect(color = NA, fill = "grey89", size = 1),
      strip.text.x = element_text(margin = margin(0.08, 0, 0.08, 0, "line"), face = stripxface),
      strip.text.y = element_text(size = textsize, margin = margin(0, 0, 0, 0, "line"))
    )
  return(p)
}

metaPlotsNoRowName <- function(datadf, boxsize = 1, conf.level = 0.95, mycol = "grey",
                               plottitle = NULL, titleface = "plain", stripxface = "plain",labex = 2, textsize = 8, summaryy = c(0, -0.4, 0, 0.4))
{
  nth <- function(x, i) {
    x[(i - 1)%%length(x) + 1]
  }
  ci.value <- -qnorm((1 - conf.level)/2)
  
  datadf$lower <- datadf$ES.g - ci.value * datadf$ES.se
  datadf$upper <- datadf$ES.g + ci.value * datadf$ES.se
  datadf$samplesize = datadf$samplesize / max(datadf$samplesize, na.rm = T)
  
  # summary
  datadfsummary = datadf[nrow(datadf),]
  dfpolygon = data.table(x = c(datadfsummary$lower, datadfsummary$ES.g, datadfsummary$upper, datadfsummary$ES.g),
                         y = summaryy)
  
  datadf[nrow(datadf), c("lower", "upper", "samplesize") := list(NA, NA, NA)]
  
  p = ggplot(datadf, aes_string(x = "ES.g", y = "index")) +
    theme_bw() +
    geom_vline(xintercept = 0, linetype = 2, alpha = 1, color = "grey62", size = 0.4) +
    geom_hline(yintercept = c(-0.5, 0.5), linetype = 1, alpha = 1, color = textcolor, size = linesize) +
    geom_errorbarh(alpha = 0.8, height = 0.4, size = 0.4, aes_string(xmax = "upper", xmin = "lower"), color = mycol) +
    geom_tile(aes(width = samplesize / 20 * boxsize), height = 0.5, fill = mycol, color = NA) +
    geom_polygon(data = dfpolygon, aes_string(x = "x", y = "y"), fill = mycol, color = NA) +
    scale_y_continuous(position = "right", breaks = datadf$index, labels = datadf$rowname, limits = c(-0.51, max(datadf$index) + 0.51), expand = c(0,0)) +
    labs(x = "effect size", title = plottitle, y = NULL) +
    facet_wrap(. ~ gene, scales = "free_x", nrow = 1) +
    geom_text(data = data.frame(x = labex, y = 0, label = datadfsummary$label), mapping = aes(x = x, y = y, label = label), color = "grey17", size = 2, parse = F, hjust = 0, lineheight = 0.7) +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major.x =element_line(color = "grey67", size = 0.05, linetype = 2),
      panel.grid.major.y = element_blank(),
      panel.background  = element_rect(color = NA, size = 0, fill = "grey98"),
      panel.border = element_blank(),
      axis.ticks = element_blank(),
      axis.line = element_blank(),
      axis.text = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      axis.text.y = element_text(size = textsize, color = textcolor),
      plot.title = element_text(size = textsize, face = titleface, margin = margin(0, 0, 0, 0, "line")),
      axis.title = element_text(size = textsize, color = textcolor),
      axis.title.x = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "null")),
      axis.title.y = element_text(size = textsize, color = textcolor),
      strip.background = element_rect(color = NA, fill = "grey89", size = 1),
      strip.text.x = element_text(margin = margin(0.08, 0, 0.08, 0, "line"), face = stripxface),
      strip.text.y = element_text(size = textsize, margin = margin(0, 0, 0, 0, "line"))
    )
  return(p)
}

pool.inverseVar < function(g, se.g, method){
  require(rmeta)
  stopifnot( identical( rownames(g), rownames(se.g) ) )
  out <- matrix( nr=nrow(g), nc=8,
                 dimnames=list( rownames(g), c("n.studies", "summary", "se.summary", "tau2", "p.value", "Q", "df", "pval.het") ) )
  
  cleanNA <- function(x) return( x[!is.na(x) & is.finite(x) ] )
  for(j in 1:nrow(g)){
    
    e  <- cleanNA(    g[j, ] )
    se <- cleanNA( se.g[j, ] )
    n  <- length(e)
    
    if(n==1){
      summ <- e;   se.summ <- se;   tau2 <- NA
      Q.het = NA
      df.het = NA
      pval.het = NA
    } else {
      fit <- meta.summaries(e, se, method = method)
      summ <- fit$summary
      se.summ <- fit$se.summary
      tau2 <- ifelse( method=="fixed", NA, fit$tau2 )
      Q.het = fit$het[1]
      df.het = fit$het[2]
      pval.het = fit$het[3]
      rm(fit)
    }
    
    pval     <- 2*pnorm( abs(summ/se.summ), lower.tail=FALSE )
    
    out[j, ] <- c(n, summ, se.summ, tau2, pval, Q.het, df.het, pval.het)
    rm(e, se, n, summ, se.summ, tau2, pval, Q.het, df.het, pval.het)
  }
  return(out)
}
```

```{r figure settings}
textsize = 10
axistextsize = 8
legendtextsize = 8
legendpointsize = 2
pointsize = 0.3
pointsizen = 0.1
linesize = 0.1
gridlwd = 0.4
clustercolumns = F
clusterrows = F
showrowdend = F
showcolumndend = F
showheatmaplegend = F

textcolor = linecolor = linecolor = "#252525" 
  
backgroundcol = "white"
  
mythemenull = theme_bw() + 
  theme(axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank())

mytheme =
  theme(text = element_text(size = textsize, color = textcolor),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_line(color = "grey67", linewidth = 0.05, linetype = 2),
        panel.border = element_blank(),
        legend.title =  element_blank(),
        legend.position = "right",
        legend.key.width = unit(0.3, "line"),
        legend.key.height = unit(0.7, "line"),
        legend.text = element_text(margin = margin(0, 0, 0, 0.1, "line")),
        legend.key.spacing.x = unit(.1, "line"),
        legend.key.spacing.y = unit(.1, "line"),
        legend.key.size = unit(1, 'lines'),
        axis.line = element_line(color = linecolor,linewidth = linesize),
        axis.ticks = element_line(color = linecolor,linewidth = linesize),
        axis.text =  element_text(size = textsize, color = textcolor),
        axis.title = element_text(size = textsize, color = textcolor),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain"),
        strip.background = element_rect(color = NA, fill = "grey91"),
        strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.05, 0, 0.05,0, "line"))
  )

col_pid = c(chan2021 = "#FFF143", chen2020 = "#FDD834", combes2021 = "#000080", 
            deng2021 = "#EE4C97", garridotrigo2023 = "#FFE099", gupta2020 = "#A60021", 
            habermann2020 = "#2F4F4F", han2020 = "#F3D3E7", hu2022 = "#5E2D30", 
            hu2023 = "#73DAFF", jardine2021 = "#FFC0CB", kaiser2024 = "#F47F17", 
            kwok2023 = "grey67", liao2020 = "#2A0BD9", maynard2020 = "#40A1FF", 
            montaldo2022 = "#FFAD73", moreno2022 = "#28231D", myin2024 = "#EAFF56", 
            qian2020 = "#8CC269", qiu2021 = "#1A5E1F", ren2021 = "#FFFFBF", 
            reyfman2019 = "#FF00FF", salcher2022 = "#DDB952", schrepping2020 = "#1C77A3", 
            schupp2020 = "#FD8CC1", sinha2021 = "#D1C4E9", suo2022 = "#E0FFFF", 
            tabulasapiens2022 = "#6A1A99", wang2021 = "#FFFF00", wang2023 = "#CD92D8", 
            wigerblad2022 = "#E9D097", wilhelm2024 = "#CE5E8A", wilk2021 = "#D92632", 
            wu2024 = "#264EFF", xue2023 = "#2BAE85", yang2022 = "#ABF8FF", yang2023 = "#7E57C1", 
            zhang2023 = "#00FF00", zhu2022 = "#F76E5E", zilionis2019 = "#C7E5C9"
)

col_group = c(healthy = "#1A6840", convalescent = "#4EAE3C", `COVID-19` = "#EFC000", 
              `COVID-19,nonsevere` = "#7b72c5", `COVID-19,severe` = "#EFC000", 
              `bacterial sepsis` = "#F97D1C", flu = "cyan", `flu,pregnancy` = "green", 
              pregnancy = "lightgreen", `lung cancer` = "#DC3023", `GI cancer` = "#FF0097", 
              `kidney cancer` = "#6A1A99", `colorectal cancer` = "#CE5E8A", 
              `brain tumor` = "#7C1823", `other cancer` = "#EEA2A4", asthma = "#142334", 
              `other lung disease` = "#352A87", CVD = "#0E5FDB", diabetes = "#C5BB5C", 
              `autoimmune disease` = "#06A5C7", `HSC-T` = "#FFDFB2", `healthy,G-CSF/IFN` = "#C7E5C9FF", 
              fetus = "#D6BBCF", `other disease` = "grey87")

col_tissue = c(`peripheral blood` = "#EDD1D8", `cord blood` = "#FBC831", `bone marrow` = "#73DAFF", 
               lung = "#DC3023", BALF = "#BEC936", sputum = "#0E5FDB", liver = "#B35C44", 
               colonrectum = "#EE6C00", pancreas = "#FF0097", stomach = "#CE5E8A", 
               kidney = "#B0A4E3", brain = "#7C1823", spleen = "#A5D6A6", `other GI organs` = "#FFDFB2", 
               `urinary system` = "#352A87", other = "grey77")

col_celltype_v1 = c("AZU1,MPO" = "#DC3023",
                    "LTF,CAMP" = "#F47B00", 
                    "MMP9" = "#F8D626",
                    "MME" = "#80C684",
                    "IL1R2" = "cyan",
                    "TXNIP" = "#D1BA58",
                    "Jun/Fos" = "#2BAE85",
                    "EGR1" = "#1A5E1F", 
                    "PLAUR" = "#FFDFB2",
                    "IFN 1" = "#2E42B8",
                    "IFN 2" = "#BADEFA",
                    "IFN 3" = "#264EFF",
                    "NF-κB" = "#5E34B1", 
                    "IL1B" = "#ccabb4", 
                    "VEGFA" = "#B29DDA", 
                    "CCL3/4" = "#acf300",  
                    "SLPI,PI3" = "#73DAFF",  
                    "ARHGAP26" = "#C7E5C9",
                    "HSP" = "#C0D8D8", 
                    "CD74" = "#e9e4c4",  
                    "Neu-Lym" = "grey77"
)

col_celltype = c("AZU1" = "#DC3023",
                 "LTF" = "#F47B00", 
                 "MMP9" = "#F8D626",
                 "MME" = "#80C684",
                 "IL1R2" = "cyan",
                 "TXNIP" = "#D1BA58",
                 "AP-1" = "#2BAE85",
                 "EGR1" = "#1A5E1F", 
                 "G0S2" = "#FFDFB2",
                 "IFN1" = "#2E42B8",
                 "IFN2" = "#BADEFA",
                 "IFN3" = "#264EFF",
                 "NF-κB" = "#5E34B1", 
                 "IL1B" = "#ccabb4", 
                 "VEGFA" = "#B29DDA", 
                 "CCL3/4" = "#acf300",  
                 "SLPI" = "#73DAFF",  
                 "ARHGAP26" = "#C7E5C9",
                 "HSP" = "#C0D8D8", 
                 "CD74" = "#e9e4c4",  
                 "Neu-Lym" = "grey77"
)

celltypenamechange = names(col_celltype)
names(celltypenamechange) = names(col_celltype_v1)

col_set = c("IMNeuS" = "#a93434",
            "TNeuS" = "#1A5E1F",
            "IFN genes" = "#2E42B8",
            "other" = "grey77"
)

col_set2 = c("IMNeuS" = "#a93434",
             "TNeuS" = "#1A5E1F",
             "IFN genes" = "#2E42B8",
             "other" = "grey67"
)
```

```{r load data}
dir = "/labs/khatrilab/hongzheng/singlecell/neutrophil/rds/"
dir.fig = "/labs/khatrilab/hongzheng/singlecell/neutrophil/figures/"

# meta data of samples
samplemeta = fread(paste0(dir, "samplemta.v10f.csv"))
samplemeta$cohort = samplemeta$pid
samplemeta$pid = gsub("a$|b$|c$|d$", "", samplemeta$pid)
samplemeta$pid = gsub("myin2024b", "yin2024", samplemeta$pid)
samplemeta$pid = gsub("reyfman2018", "reyfman2019", samplemeta$pid)
samplemeta$pid = gsub("xue2022", "xue2023", samplemeta$pid)

# meta data of cells
dtf = readRDS(paste0(dir, "dtf.pc25.v10f.rds"))
dtfmore = readRDS(paste0(dir, "dtf.v10f.cellmeta.rds"))

dtf[, celltype:=ifelse(celltype %in% names(celltypenamechange), celltypenamechange[celltype], celltype)]

dtf$pid = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$pid
dtf$cohort = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$cohort
dtf$subjectid = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$subjectid
dtf$sex = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$sex
dtf$age = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$age
dtf$tissue = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$tissue
dtf$tissue.detail = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$tissue.detail
dtf$group = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$group
dtf$group.detail = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$group.detail
dtf$severity = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$severity
dtf$source = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$source
dtf$other = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$other
dtf$technique = samplemeta[match(dtf$sampleid.pid, samplemeta$sampleid.pid)]$technique

dtf = cbind(dtf, dtfmore[,c("nCount_RNA", "nFeature_RNA", "pct.mt", "S.Score", "G2M.Score", "cxds_score", "bcds_score", "hybrid_score")])
```

```{r gene signatures}
genes.immature.neu = c("AZU1", "BPI", "CAMP", "CEACAM8", "CRISP3", "DEFA3", "DEFA4", 
                       "ELANE", "LCN2", "LTF", "MMP8", "MPO", "OLFM4")
genes.total.neu = c("ALPL", "ANXA3", "CXCR1", "CYP4F3", "KCNJ15", "MME", "MMP9", 
                    "PI3", "SLPI", "TNFRSF10C")
genes.ifn = c("APOL6", "CMPK2", "DDX58", "EPSTI1", "GBP1", "GBP2", "GBP4", 
              "GBP5", "HERC5", "HES4", "IFI44", "IFI44L", "IFI6", "IFIH1", 
              "IFIT1", "IFIT2", "IFIT3", "IFIT5", "IFITM1", "IFITM3", "IRF7", 
              "ISG15", "LY6E", "MT1X", "MT2A", "MX1", "OAS1", "OAS2", "OAS3", 
              "OASL", "PARP14", "RSAD2", "SAMD9", "SAMD9L", "TNFSF10", "TNFSF13B", 
              "XAF1")

geneSets = list(genes.immature.neu, genes.total.neu, genes.ifn)
names(geneSets) = c("genes.immature.neu", "genes.total.neu", "genes.ifn")

signaturegenes = c("STMN1", "MKI67", "FUT4", "AZU1", "MPO", "PRTN3", "CTSG", "ELANE", 
                   "DEFA4", "CEACAM6", "DEFA3", "BPI", "LYZ", "RETN", "ANXA1", "CEACAM8", 
                   "OLFM4", "TCN1", "LTF", "LCN2", "MMP8", "CAMP", "CRISP3", "HP", 
                   "PGLYRP1", "ANXA3", "FCN1", "ARG1", "PADI4", "CD177", "MMP9", 
                   "NQO2", "CYP4F3", "S100A8", "S100A9", "S100A12", "TXN", "TSPO", 
                   "PLAC8", "GPI", "OLR1",  "ORM1", "MME", "IL1R2", "TXNIP", 
                   "FOS", "FOSB", "JUN", "EGR1", "G0S2", 
                   "MT2A", "MT1X", "HES4", 
                   "LY6E", "ISG15", "IFI44L", "IFI44", "OASL", "IFI6", "OAS3", "OAS2", 
                   "RSAD2", "HERC5", "MX1", "IFIT1", "IFIT2", "IFIT3", "IFIT5", 
                   "DDX58", "XAF1", "IRF7", "EPSTI1", "APOL6", "GBP5", "GBP4", "GBP2", 
                   "GBP1", "PARP14",  "IFIH1", "TNFSF10", "IFITM1", "IFITM3", "CD274",
                   
                   "TNFAIP3", "NFKBIA", "CXCL8", "IL1B", "IL1RN", "CXCL1", "CXCL2", 
                   "PPIF", "SPP1", "CDKN1A", "CD83", "SQSTM1", "PLAU", "CXCR4", 
                   "CSTB", "LGALS3", "HMOX1", "VEGFA", "CCL3", "CCL4", "SLPI", "PI3", 
                   "SIGLEC10", "ARHGAP26", "LUCAT1", "DOCK5" , "HSPA1B", 
                   "DNAJB1", "HSPA5", "CD74", "HLA-DRA", "VCAN", "CD69", "IL7R", 
                   "GNLY", "SELL",
                   "KCNJ15", "TNFRSF10C", "ALPL", "CSF3R", "NAMPT", "FCGR3B", "CXCR1", 
                   "CXCR2")
```

# --------------------- FIGURES ---------------------

# UMAPs and ggalluvial ----

```{r}
### pid ----
dtf$facetvar = "study"
p.umap.pid =
  ScatterPlotRast(df = as.data.frame(dtf), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.01, alphalevel = 0.4, colorby = "pid", groupby = "facetvar", legendColumn = 5, colcateg= "discrete", numberofrows = 5, legendposition = "none", seed = 13, mycol = col_pid, backgroudcol = NA, plottitle = "", rasterdpi = 500, nacol = "grey42", facetlabelface = "plain") +
  labs(title = NULL) + 
  theme(text = element_text(size = textsize),
        legend.text = element_text(size = 6, margin = margin(0, 0, 0, 0.05, "line")),
        legend.position = c(1.03, 1.15),
        legend.justification = c(0, 1),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.2, "line"),
        legend.key.spacing.x = unit(.01, "line"),
        legend.key.spacing.y = unit(.1, "line"),
        legend.key.size = unit(1, 'lines'),
        legend.box.margin =  margin(0, 0, 0, 0.5, "line"),
        plot.margin = margin(0.05, 0.05, 0, 0.05, "line"),
        strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.05, 0, 0.08, 0, "line"))
  ) + 
  guides(color = guide_legend(nrow = 20, label.hjust = 0, override.aes = list(size = 1.5, alpha = 1))) 

cairo_pdf(file = paste0(dir.fig, "HNeuSA.fig1.umap.pid.pdf"), width = 2.8, height = 2.4)
pushViewport(viewport(layout = grid.layout(nrow = 38, ncol = 41)))
print(p.umap.pid,vp = viewport(layout.pos.row = 2:22, layout.pos.col = 1:18))
dev.off()

### ggalluvial ----
dtf.summary = as.data.table(table(dtf$tissue, dtf$group))
dtf.summary = dtf.summary[N > 0]
colnames(dtf.summary)[1:2] = c("tissue", "group")
dtf.summary$tissue = factor(dtf.summary$tissue, levels = rev(names(col_tissue)))
dtf.summary$group = factor(dtf.summary$group, levels = rev(names(col_group)))

col_tissue_group = c(col_tissue, col_group)

dtf.summary_long = to_lodes_form(data.frame(dtf.summary),
                                 key = "category", value = "item", id = "index",
                                 axes = 2:1)
dtf.summary_long = as.data.table(dtf.summary_long)
dtf.summary_long[, colflow:=ifelse(category == "group", item, NA)]

p.ggalluvial = ggplot(data = dtf.summary_long,
                      aes(x = category, stratum = item, alluvium = index, y = N)) +
  geom_flow(aes(fill = item), width = 1/5) +
  geom_stratum(aes(fill = item), width = 1/5, size = .1) +
  #geom_text(stat = "stratum", aes(label = item)) +
  scale_fill_manual(values = col_tissue_group) +
  mythemenull +
  coord_flip() +
  theme(legend.position = "none",
        plot.margin = unit(c(-0.8, -1.1, -0.8, -1.1), "line"),) +
  labs(x = NULL, y= NULL) 

### tissue ----
dtf$facetvar = "tissue"
p.umap.tissue =
  ScatterPlotRast(df = as.data.frame(dtf), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.01, alphalevel = 0.4, colorby = "tissue", groupby = "facetvar", legendColumn = 4, colcateg = "discrete", numberofrows = 5, legendposition = "none", seed = 13, mycol = col_tissue, backgroudcol = NA, plottitle = "", rasterdpi = 500, nacol = "grey42", facetlabelface = "plain") +
  labs(title = NULL) + 
  theme(text = element_text(size = textsize),
        legend.text = element_text(margin = margin(0, 0, 0, 0.1, "line")),
        legend.position = c(1.01, 0.5),
        legend.justification = c(0, 0.5),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.2, "line"),
        legend.key.spacing.x = unit(.1, "line"),
        legend.key.spacing.y = unit(.1, "line"),
        legend.key.size = unit(1, 'lines'),
        plot.margin = margin(0.05, 0.05, 0, 0.05, "line"),
        strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.05, 0, 0.08, 0, "line"))
  ) + 
  guides(color = guide_legend(nrow = 4, label.hjust = 0, override.aes = list(size = 2, alpha = 1))) 

### group ----
dtf$facetvar = "condition"
p.umap.group =
  ScatterPlotRast(df = as.data.frame(dtf), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.01, alphalevel = 0.4, colorby = "group", groupby = "facetvar", legendColumn = 4, colcateg = "discrete", numberofrows = 5, legendposition = "none", seed = 13, mycol = col_group, backgroudcol = NA, plottitle = "", rasterdpi = 500, nacol = "grey42", facetlabelface = "plain") +
  labs(title = NULL) +
  theme(text = element_text(size = textsize),
        legend.text = element_text(margin = margin(0, 0, 0, 0.1, "line")),
        legend.position = c(1.01, 0.55),
        legend.justification = c(0, 0.5),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.2, "line"),
        legend.key.spacing.x = unit(.1, "line"),
        legend.key.spacing.y = unit(.1, "line"),
        legend.key.size = unit(1, 'lines'),
        plot.margin = margin(0.05, 0.05, 0, 0.05, "line"),
        strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.05, 0, 0.08, 0, "line"))
  ) + 
  guides(color = guide_legend(nrow = 7, label.hjust = 0, override.aes = list(size = 2, alpha = 1))) 

### subsets ----
dtf$celltype = factor(dtf$celltype, levels = names(col_celltype))
df.label = as.data.table(dtf)[, lapply(.SD, median), by = celltype, .SDcols=c("UMAP_1", "UMAP_2")]
colnames(df.label)[2:3] = c("x", "y")
df.label$labeltext = df.label[, "celltype", with= F]
df.label[, x:=ifelse(celltype == "IL1B", x-0.2, x)]
df.label[, x:=ifelse(celltype == "VEGFA", x-0.5, x)]
df.label[, y:=ifelse(celltype == "ARHGAP26", y + 0.2, y)]
df.label[, y:=ifelse(celltype == "EGR1", y - 0.2 , y)]
dtf$facetvar = "neutrophil subset"
p.umap.celltype = 
  ScatterPlotRastNoFacet(df = as.data.frame(dtf), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = .01, alphalevel = 0.4, colorby = "celltype", legendColumn = 4, colcateg= "discrete", numberofrows = 5, legendposition = "none", seed = 13, mycol = col_celltype, backgroudcol = NA, plottitle = "", rasterdpi = 300, nacol = "grey42", facetlabelface = "plain", labeldf = df.label, labelfont = "plain", labelsize = 2.7) +
  labs(title = NULL) + 
  theme(text = element_text(size = textsize),
        legend.text = element_text(margin = margin(0, 0, 0, 0.1, "line")),
        legend.position = "none",
        legend.justification = c(0, 0.5),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.2, "line"),
        legend.key.size = unit(1, 'lines'),
        plot.margin = margin(0.05, 0.05, 0, 0.05, "line"),
        strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.05, 0, 0.1, 0, "line"))
  )  + 
  guides(color = guide_legend(ncol = 1, label.hjust = 0, override.aes = list(size = 2, alpha = 1))) 

### mature and immature ----
dtf[, celltype_matureornot:=ifelse(celltype %in% c("AZU1", "LTF"), "immature", "mature")]
df.label2 = as.data.table(dtf)[,lapply(.SD, median), by = celltype_matureornot, .SDcols=c("UMAP_1", "UMAP_2")]
colnames(df.label2)[2:3] = c("x", "y")
df.label2$labeltext = df.label2[, "celltype_matureornot", with = F]
df.label2[, x:=ifelse(celltype_matureornot == "immature", x + 1.5 , x)]

p.umap.celltype.matureornot = 
  ScatterPlotRastNoFacet(df = as.data.frame(dtf), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.00001, alphalevel = 0.3, colorby = "celltype_matureornot", legendColumn = 4, colcateg= "discrete", numberofrows = 5, legendposition = "none", seed = 13, mycol = c("immature" =  "#F47B00", "mature" = "grey87"), backgroudcol = NA, plottitle = "", rasterdpi = 400, nacol = "grey42", facetlabelface = "plain", labeldf = df.label2, labelfont = "plain", labelsize = 2.7) +
  labs(title = NULL) + 
  theme(text = element_text(size = textsize),
        legend.text = element_text(margin = margin(0, 0, 0, 0.1, "line")),
        legend.position = "none",
        legend.justification = c(0, 0.5),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.2, "line"),
        legend.key.size = unit(1, 'lines'),
        plot.margin = margin(0.05, 0.05, 0, 0.05, "line"),
        strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.05, 0, 0.1, 0, "line"))
  )  + 
  guides(color = guide_legend(ncol = 1, label.hjust = 0, override.aes = list(size = 2, alpha = 1))) 

### genes ----
mat.3genes = readRDS(paste0(dir, "mat.3genes.neu.v10f.rds"))
dtfgenes = as.data.table(cbind(dtf, mat.3genes))
dtf.score = readRDS(paste0(dir, "dtf.score.v10f.rds"))
dtfgenes$`granule\ngenes` = dtf.score$granules.score - 1
```

```{r}
pgenes = list()
for(gene in c("FCGR3B", "CXCR2", "MME", "granule\ngenes")){
  dtfgenes$id = dtfgenes[, gene, with = F]
  dtfgenes$facetvar = gene
  pgenes[[gene]] = ScatterPlotRast(as.data.frame(dtfgenes), featureX = "UMAP_1", featureY = "UMAP_2", groupby = "facetvar", pointsize = 0.0001, alphalevel = 0.1, colorby = "id", legendColumn = 2, colcateg= "continuous", numberofrows = 4, legendtextsize = 8, legendposition = c(0.96, 0.5), plottitle = NULL, seed = 1, backgroudcol = "white", mycol = c("grey99",dichromat_pal("DarkRedtoBlue.12")(12)[c(7:12)]), legendsize = 0.6, rasterdpi = 400, shuffle = T, contvarlower = 0.001, contvarupper = 0.999)
}

mat.neu = readRDS(paste0(dir, "mat.gene267.neu.v10f.rds"))
pgenes2 = list()
for(gene in c("AZU1", "MPO", "PRTN3", "CTSG", "ELANE", "DEFA4","CEACAM8", "LTF", "MMP8", "CAMP", "ARG1", "MMP9",
              "S100A8", "OLR1", "ORM1", "IL1R2", "TXNIP", "FOS", "JUN", "EGR1" , "IFI6", "GBP5", "CD274", "NFKBIA", "IL1B", "CXCL1", "CXCL2", "CXCR4", "VEGFA", 
              "CCL3", "SLPI", "PI3"
)){
  dtfgenes$id = mat.neu[,gene]
  dtfgenes$facetvar = gene
  pgenes2[[gene]] = ScatterPlotRast(as.data.frame(dtfgenes), featureX = "UMAP_1", featureY = "UMAP_2", groupby = "facetvar", pointsize = 0.00001, alphalevel = 0.1, colorby = "id", legendColumn = 2, colcateg= "continuous", numberofrows = 4, legendtextsize = 8, legendposition = "none", plottitle = NULL, seed = 1, backgroudcol = "white", mycol = c("grey99",dichromat_pal("DarkRedtoBlue.12")(12)[c(7:12)]), legendsize = 0.6, rasterdpi = 400, shuffle = T, contvarlower = 0.001, contvarupper = 0.999) + theme(plot.background = element_rect(color = "transparent", fill = "transparent"), strip.background = element_rect(color = "transparent", fill = "transparent"), strip.text.x = element_text(size = 8, color = textcolor, margin = margin(0.02, 0.02, 0.05,0.02, "line")))
}
```

# Neutrophil proportion in peripheral blood

```{r}
dtf.s.pct.blood = dtf.s.pct[tissue %in% c("peripheral blood") & !is.na(number.allcell)][!tissue.detail %in% c("neutrophil", "granulocyte") & !pid %in% c("deng2021", "tabulasapiens2022", "han2020", "wilhelm2024")] 
# wilhelm2024: not sure if it is fresh or frozen
# deng2021: it says PBMC in the paper, but very high proportion (>60%) of neutrophils was observed in 2 out of 3 samples

dtf.s.pct.blood[,lapply(.SD, function(x){median(x)}), by = c("source2"), .SDcols = c("neu.pct")]
dtf.s.pct.blood[grepl("freshWB",source)][,lapply(.SD, function(x){median(x)}), by = c("technique"), .SDcols = c("neu.pct")]

p.neu.pct.source2 = 
  ggplot(dtf.s.pct.blood, aes(x = source2, y = neu.pct)) +
  ggbeeswarm::geom_quasirandom(size = 0.2, width = 0.2, show.legend = T, alpha = 0.7, varwidth = F, color = "grey42", shape = 16) +
  geom_boxplot(width = 0.35, size = 0.3, outlier.shape = NA, alpha = 1, fill = NA, color = "grey17") +
  theme_bw() + mytheme +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 1.5),
        panel.grid.major.y = element_blank(),
        plot.margin = unit(c(.1, .1, .2, .1), "line"),
        axis.text.x = element_text(angle = 0, hjust = 0.5)
  ) +
  labs(x = NULL, y = NULL,  title = "") +
  guides(color = guide_legend(ncol = 1, label.hjust = 0, override.aes = list(size = 2, alpha = 1))) +
  scale_x_discrete(limits = rev, position = "bottom") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  coord_flip()

dtf.s.pct.blood$technique = as.character(dtf.s.pct.blood$technique)
dtf.s.pct.blood[,technique2:=ifelse(technique %in% c("10x Chromium", "BD Rhapsody"), technique, "other")]
p.neu.pct.technique2 = 
  ggplot(dtf.s.pct.blood[grepl("freshWB", source)], aes(x = technique2, y = neu.pct)) +
  ggbeeswarm::geom_quasirandom(size = 0.2, width = 0.2, show.legend = T, alpha = 0.7, varwidth = F, shape = 16, color = "grey42") +
  geom_boxplot(width = 0.4, size = 0.3, outlier.shape = NA, alpha = 1, fill = NA, color = "grey17") +
  theme_bw() + mytheme +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 0, hjust = 0.5),
        plot.margin = unit(c(.1, .1, .2, .23), "line")
  ) +
  labs(x = NULL, y = NULL,  title = NULL) +
  guides(color = guide_legend(ncol = 1, label.hjust = 0, override.aes = list(size = 2, alpha = 1))) +
  scale_x_discrete(limits = rev, position = "bottom") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  coord_flip()
```

# Neutrophil subsets distribution in tissues

```{r}
# proportion of immature subsets
dtf.s.pct$immature = dtf.s.pct$AZU1 + dtf.s.pct$LTF

dtf.s.pct.m0 = melt(dtf.s.pct,
                    id.vars = c("group", "pid", "sampleid.pid", "tissue", "tissue.detail"),
                    measure.vars = c(names(col_celltype), "immature"))

dtf.s.pct.m3 = melt(dtf.s.pct[source %in% c("freshWB_RBClysis", "frozenPBMC", "freshWB_RBCdepletion", 
                                            "freshPBMC", "neutrophil")],
                    id.vars = c("group", "pid", "sampleid.pid", "source"),
                    measure.vars = c(names(col_celltype), "immature"))
dtf.s.pct.m3 = as.data.table(dtf.s.pct.m3)
dtf.s.pct.m3[,tissue:=ifelse(grepl("WB", source), "whole blood", source)]
dtf.s.pct.m3[,tissue:=ifelse(grepl("neutrophil", source), "neutrophil", tissue)]
dtf.s.pct.m3[,tissue:=ifelse(grepl("PBMC", source), "PBMC", tissue)]

dtf.s.pct.m = melt(dtf.s.pct,
                   id.vars = c("group", "pid", "sampleid.pid", "tissue", "tissue.detail"),
                   measure.vars = names(col_celltype))

dtf.s.pct.m = as.data.table(dtf.s.pct.m)
length(table(dtf.s.pct.m$sampleid.pid)) # 879

dtf.s.pct.m.agg = dtf.s.pct.m[,lapply(.SD, function(x){mean(x)}), by = c("tissue",  "variable"), .SDcols = c("value")]
dtf.s.pct.m.aggmedian = dtf.s.pct.m[,lapply(.SD, function(x){median(x)}), by = c("tissue",  "variable"), .SDcols = c("value")]
dtf.s.pct.m.aggsd = dtf.s.pct.m[,lapply(.SD, function(x){sd(x)}), by = c("tissue",  "variable"), .SDcols = c("value")]
dtf.s.pct.m.agg[,lapply(.SD, function(x){sum(x)}), by = c("tissue"), .SDcols = c("value")]

dtf.s.pct.m1 = melt(dtf.s.pct[source %in% c("freshWB_RBClysis", "frozenPBMC", "freshWB_RBCdepletion", 
                                            "freshPBMC", "neutrophil")],
                    id.vars = c("group", "pid", "sampleid.pid", "source"),
                    measure.vars = names(col_celltype))
dtf.s.pct.m1 = as.data.table(dtf.s.pct.m1)
dtf.s.pct.m1[,tissue:=ifelse(grepl("WB", source), "whole blood", source)]
dtf.s.pct.m1[,tissue:=ifelse(grepl("neutrophil", source), "neutrophil", tissue)]
dtf.s.pct.m1[,tissue:=ifelse(grepl("PBMC", source), "PBMC", tissue)]
table(dtf.s.pct.m1$tissue)

dtf.s.pct.m1.agg = dtf.s.pct.m1[,lapply(.SD, function(x){mean(x)}), by = c("tissue", "variable"), .SDcols = c("value")]
dtf.s.pct.m1.aggmedian = dtf.s.pct.m1[,lapply(.SD, function(x){median(x)}), by = c("tissue", "variable"), .SDcols = c("value")]
dtf.s.pct.m1.aggsd = dtf.s.pct.m1[,lapply(.SD, function(x){sd(x)}), by = c("tissue", "variable"), .SDcols = c("value")]
dtf.s.pct.m1.agg[,lapply(.SD, function(x){sum(x)}), by = c("tissue"), .SDcols = c("value")]

dtf.s.pct.m.agg.combined = rbind(dtf.s.pct.m.agg, dtf.s.pct.m1.agg[, colnames(dtf.s.pct.m.agg), with = F])

dtf.s.pct.m.aggmedian.combined = rbind(dtf.s.pct.m.aggmedian,
                                       dtf.s.pct.m1.aggmedian[, colnames(dtf.s.pct.m.aggmedian), with = F])

dtf.s.pct.m.aggsd.combined = rbind(dtf.s.pct.m.aggsd,
                                   dtf.s.pct.m1.aggsd[, colnames(dtf.s.pct.m.aggsd), with = F])

dtf.s.pct.m.agg.combined$variable = factor(dtf.s.pct.m.agg.combined$variable, levels = rev(names(col_celltype)))
dtf.s.pct.m.aggmedian.combined$variable = factor(dtf.s.pct.m.aggmedian.combined$variable, levels = rev(names(col_celltype)))
dtf.s.pct.m.aggsd.combined$variable = factor(dtf.s.pct.m.aggsd.combined$variable, levels = rev(names(col_celltype)))

tissueorder = c("other", "BALF", "lung", "sputum",  
                "other GI organs", "stomach", "pancreas", 
                "brain", "colonrectum", "kidney", "liver", 
                "spleen", "bone marrow", "cord blood", 
                "PBMC", "neutrophil", "whole blood",
                "peripheral blood")

dtf.s.pct.m.agg.combined$tissue = factor(dtf.s.pct.m.agg.combined$tissue, levels = tissueorder)
dtf.s.pct.m.aggmedian.combined$tissue = factor(dtf.s.pct.m.aggmedian.combined$tissue, levels = tissueorder)
dtf.s.pct.m.aggsd.combined$tissue = factor(dtf.s.pct.m.aggsd.combined$tissue, levels = tissueorder)

# bar ----
p.tissuedistribution = ggplot(dtf.s.pct.m.agg.combined[!tissue %in% c("other")], aes(x = tissue, y = value, fill = variable)) +
  geom_bar(stat = "identity", alpha = 0.8) +
  theme_bw() + mytheme +
  scale_y_continuous(expand = c(0, 0), breaks = seq(0, 1, 0.2), labels = seq(0, 1, 0.2), position = "left") + 
  scale_x_discrete(position = "top") + 
  labs(x = NULL, y = NULL , title = NULL) +
  theme(
    panel.grid.major = element_blank(), 
    plot.margin = unit(c(.2, .2, .2, .2), "line"),
    panel.border = element_blank(),
    legend.title =  element_blank(),
    legend.text = element_text(color = textcolor, face = "italic"),
    legend.position = "none",
    legend.spacing.y = unit(0.01, "line"), 
    legend.spacing.x = unit(0.04, "line"),
    legend.key.width = unit(0.2, "line"),
    legend.key.height = unit(0.2, "line"),
    plot.title = element_text(size = textsize, face = "plain", hjust = 0.5, margin = margin(t = 1, r = 0, b = 1, l = 0, unit = "pt"))) +
  scale_fill_manual(values = col_celltype, name= "") + coord_flip() +
  guides(fill = guide_legend(ncol = 1, byrow = T, size = 2.3)) 

# box ----
tmp = dtf.s.pct.m[!tissue %in% c("cord blood", "bone marrow", "spleen", "peripheral blood", "BALF", "sputum", "other") | tissue.detail %in% c("bladder", "breast", "cervix", "heart", "ovarian", "ovary", "uterus")][, c("tissue", "variable", "value")]
tmp$tissue = "other\ntissue"

dtf.tmp = rbind(
  dtf.s.pct.m1[tissue %in% c("whole blood", "PBMC")][, c("tissue", "variable", "value")],
  dtf.s.pct.m[tissue %in% c("cord blood", "bone marrow", "spleen")][, c("tissue", "variable", "value")],
  tmp
)

dtf.tmp$tissue = gsub(" blood", "\nblood", dtf.tmp$tissue)
dtf.tmp$tissue = gsub("bone marrow", "bone\nmarrow", dtf.tmp$tissue)
dtf.tmp$tissue = factor(dtf.tmp$tissue, levels = c("PBMC", "whole\nblood", "cord\nblood", "bone\nmarrow", "spleen", "other\ntissue"))
p.immature.pct = 
  ggplot(dtf.tmp[variable %in% c("AZU1", "LTF")], aes(x = tissue, y = value)) +
  ggbeeswarm::geom_quasirandom(color = "grey42", size = .1, width = 0.2, show.legend = T, alpha = 0.5, varwidth = F, shape = 16) +
  geom_boxplot(width = 0.2, size = 0.2, outlier.shape = NA, alpha = 1, fill = NA, color = "grey17") +
  labs(y = "proportion in total neutrophils", x = NULL, title = "immature neutrophil subsets") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  theme_bw() + mytheme + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1, lineheight = 0.7),
        plot.margin = unit(c(.2, .2, .2, .2), "line"),
        panel.grid.major.x = element_blank(),
        plot.title = element_text(size = textsize, face = "plain", hjust = 0, margin = margin(t = .1, r = 0, b = .1, l = 0, unit = "pt"))
  ) +
  facet_wrap(. ~ variable, nrow = 1, scales = "free") +
  stat_compare_means(comparisons = list(c("whole\nblood", "PBMC"), c("whole\nblood", "cord\nblood"), c("whole\nblood", "bone\nmarrow"), c("whole\nblood", "spleen")), method = "wilcox.test",vjust = 1.5, size =2, label = "p.format", color = "grey17", bracket.size = 0.1, tip.length = 0.01, step.increase = c(0, 0, -0.06, 0.04))

p.allsubset.pct = 
  ggplot(dtf.tmp, aes(x = tissue, y = value)) +
  ggbeeswarm::geom_quasirandom(color = "grey42",size = .1, width = 0.2, show.legend = T, alpha = 0.5, varwidth = F, shape = 16) +
  geom_boxplot(width = 0.2, size = 0.2, outlier.shape = NA, alpha = 1, fill = NA, color = "grey17") +
  labs(y = "proportion in total neutrophils", x = NULL, title = NULL) +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  theme_bw() + mytheme + 
  theme(axis.text.x = element_text(angle = 50, hjust = 1, lineheight = 0.7),
        plot.margin = unit(c(.2, .2, .2, .2), "line"),
        panel.grid.major.x = element_blank(),
        plot.title = element_text(size = textsize, face = "plain", hjust = 0,margin = margin(t = .1, r = 0, b = .1, l = 0, unit = "pt"))
  ) +
  facet_wrap(. ~ variable, nrow = 3, scales = "free") +
  stat_compare_means(comparisons = list(c("whole\nblood", "PBMC"), c("whole\nblood", "bone\nmarrow"), c("whole\nblood", "cord\nblood"), c("whole\nblood", "spleen"), c("whole\nblood", "other\ntissue")), method = "wilcox.test",vjust = 1.5, size =2, label = "p.format", color = "grey17", bracket.size = 0.1,tip.length = 0.01, step.increase = c(0, 0, -0.06, 0.04, 0.06))

tmp = dtf.s.pct.m[tissue %in% c("lung", "colonrectum", "other GI organs", 
                                "liver", "pancreas", "brain", "kidney", "stomach")][, c("tissue", "variable", "value")]
tmp$tissue = "tissue\ncombined"

dtf.tmp2 = rbind(
  dtf.s.pct.m1[tissue %in% c("whole blood")][, c("tissue", "variable", "value")],
  dtf.s.pct.m[tissue %in% c("lung", "colonrectum", "other GI organs", 
                            "liver", "pancreas", "brain", "kidney", "stomach")][, c("tissue", "variable", "value")],
  tmp
)
dtf.tmp2$tissue = gsub(" blood", "\nblood", dtf.tmp2$tissue)
dtf.tmp2$tissue = gsub("other GI organs", "other\nGI organs", dtf.tmp2$tissue)

p.tissuesubsets.pct = 
  ggplot(dtf.tmp2[variable %in% c("NF-κB", "IL1B", 
                                  "VEGFA", "CCL3/4", "SLPI")], aes(x = tissue, y = value)) +
  ggbeeswarm::geom_quasirandom(color = "grey42",size = .1, width = 0.2, show.legend = T, alpha = 0.5, varwidth = F, shape = 16) +
  geom_boxplot(width = 0.2, size = 0.2, outlier.shape = NA, alpha = 1, fill = NA, color = "grey17") +
  labs(y = "proportion in total neutrophils", x = NULL, title = "cytokine/inflammation subsets") +
  scale_y_continuous(breaks = seq(0, 1, 0.1)) +
  theme_bw() + mytheme + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1, lineheight = 0.7),
        plot.margin = unit(c(.2,.2,.2,.2), "line"),
        panel.grid.major.x = element_blank(),
        plot.title = element_text(size = textsize, face = "plain", hjust = 0, margin = margin(t = .1, r = 0, b = .1, l = 0, unit = "pt"))
  ) +
  facet_wrap(. ~ variable, nrow = 3, scales = "free") +
  stat_compare_means(comparisons = list(c("whole\nblood", "tissue\ncombined")), method = "wilcox.test",vjust = 1.5, size =2, label = "p.format", color = "grey17", bracket.size = 0.1, tip.length = 0.01)

# ht ----
dtf.s.pct.m.agg.combined.mat = dcast(dtf.s.pct.m.aggmedian.combined[!tissue %in% c("other")], tissue ~ variable, value.var = "value")
rowname = dtf.s.pct.m.agg.combined.mat$tissue
dtf.s.pct.m.agg.combined.mat$tissue = NULL
dtf.s.pct.m.agg.combined.mat = as.matrix(dtf.s.pct.m.agg.combined.mat)
rownames(dtf.s.pct.m.agg.combined.mat) = rowname
dtf.s.pct.m.agg.combined.mat = dtf.s.pct.m.agg.combined.mat[, intersect(names(col_celltype), colnames(dtf.s.pct.m.agg.combined.mat))]

dtf.s.pct.m.aggsd.combined.mat = dcast(dtf.s.pct.m.aggsd.combined[!tissue %in% c("other")], tissue ~ variable, value.var = "value")
rowname = dtf.s.pct.m.aggsd.combined.mat$tissue
dtf.s.pct.m.aggsd.combined.mat$tissue = NULL
dtf.s.pct.m.aggsd.combined.mat = as.matrix(dtf.s.pct.m.aggsd.combined.mat)
rownames(dtf.s.pct.m.aggsd.combined.mat) = rowname
dtf.s.pct.m.aggsd.combined.mat = dtf.s.pct.m.aggsd.combined.mat[, intersect(names(col_celltype), colnames(dtf.s.pct.m.aggsd.combined.mat))]

col_fun.tissueht = circlize::colorRamp2(seq(0, max(dtf.s.pct.m.agg.combined.mat), length =9), c("grey99", pal_material("blue")(10)[3:10]))
dtf.s.pct.m.agg.combined.mat = dtf.s.pct.m.agg.combined.mat[rev(rownames(dtf.s.pct.m.agg.combined.mat)),]
dtf.s.pct.m.aggsd.combined.mat = dtf.s.pct.m.aggsd.combined.mat[rev(rownames(dtf.s.pct.m.aggsd.combined.mat)),]
mat.size = 1 - sqrt(dtf.s.pct.m.aggsd.combined.mat)

p.tissue.ht = Heatmap(dtf.s.pct.m.agg.combined.mat,
                      rect_gp = gpar(type = "none"),
                      col = col_fun.tissueht,
                      na_col = "grey93",
                      cluster_columns = F, show_row_names = T,
                      cluster_rows = F, show_row_dend = T, show_column_dend = T, row_dend_side = "left", column_dend_side = "top",
                      clustering_method_rows = "ward.D2", clustering_method_columns  = "ward.D2",
                      clustering_distance_rows  = "euclidean", clustering_distance_columns  = "euclidean", 
                      row_names_side = "right",
                      column_names_side = "top", column_names_rot = 55,
                      column_names_gp = gpar(fontsize = 10, fontface = "plain", col = col_celltype),
                      row_names_gp = gpar(fontsize = 10, fontface = "plain"),
                      column_title = NULL, column_title_side = "top", column_title_gp = gpar(fontsize = 8, fontface = "plain"),
                      show_heatmap_legend = T,
                      cell_fun =  function(j, i, x, y, width, height, fill){
                        grid.rect(x = x, y = y, width = width, height = height, gp = gpar(lwd = 0.2, col = "grey90",fill = NA, alpha = 0.2))
                        grid.circle(x = x, y = y, r = 0.6 * mat.size[i, j]  * min(unit.c(width, height)), gp = gpar(fill = col_fun.tissueht(dtf.s.pct.m.agg.combined.mat[i, j]), col = NA, alpha = 0.9))},
                      heatmap_legend_param = list(title = "median proportion", title_gp = gpar(fontsize = 9), labels_gp = gpar(fontsize = 9), legend_width = unit(5, "line"), legend_height = unit(.2, "line"), grid_width = unit(5, "line"), grid_height = unit(.1, "line"), by_row = F, direction = "horizontal", title_position = "lefttop"))
```

# Marker heatmap

```{r}
mat.expr = readRDS(paste0(dir, "mat.expr.v10f.rds"))
genes.geomMean = readRDS(paste0(dir, "mat.genes.geomMean.v10f.rds"))

mat.expr = mat.expr[, neu.cluster.markers2]
mat.expr = apply(mat.expr, 2, function(x) (x - min(x)) / (max(x) - min(x)))
mat.expr = mat.expr[intersect(names(col_celltype), rownames(mat.expr)),]

matcolnames = data.table(v1 = colnames(mat.expr))
matcolnames[,v2:=ifelse(v1 %in% neu.cluster.markers, paste0(v1, "*"), v1)]

colnames(mat.expr) = matcolnames$v2

genes.geomMean = genes.geomMean[neu.cluster.markers2]
col_fun = circlize::colorRamp2(seq(0, max(mat.expr), length = 7), c("grey99", pal_material("red")(10)[1:6]))

column_ha = HeatmapAnnotation(expression = anno_barplot(genes.geomMean, bar_width = 0.8, axis = F, border = F,
                                                        height = unit(1, "line"),
                                                        gp = gpar(fill = "grey78", col = "grey78")),
                              show_annotation_name = T,
                              annotation_name_side = "left", annotation_name_gp = gpar(fontsize = 8)
)

p.markerht = Heatmap(mat.expr,
                     rect_gp = gpar(col = "white", lwd = 0.5),
                     col = col_fun,
                     na_col = "grey93",
                     bottom_annotation = column_ha,
                     cluster_columns = T, show_row_names = T,
                     cluster_rows = F, show_row_dend = F, show_column_dend = F, row_dend_side = "left", column_dend_side = "top",
                     clustering_method_rows = "ward.D2", clustering_method_columns  = "ward.D2",
                     clustering_distance_rows  = "euclidean", clustering_distance_columns  = "euclidean", 
                     row_names_side = "left",
                     column_names_side = "top", 
                     column_names_gp = gpar(fontsize = 8, fontface = "italic"),
                     row_names_gp = gpar(fontsize = 9, fontface = "plain", col = col_celltype),
                     column_title = NULL, column_title_side = "top", column_title_gp = gpar(fontsize = 8, fontface = "plain"),
                     show_heatmap_legend = T,
                     heatmap_legend_param = list(title = "scaled expression", title_gp = gpar(fontsize = 8), labels_gp = gpar(fontsize = 8), legend_width = unit(0.2, "line"), legend_height = unit(1.8, "line"), grid_width = unit(0.1, "line"), grid_height = unit(1.5, "line"), at = c(0.3, max(mat.expr) - 0.3), labels = c("low", "high"), by_row = F, direction = "vertical", title_position = "leftcenter-rot"))
```

```{r}
mat.expr = readRDS(paste0(dir, "mat.expr.v10f.rds"))
genes.geomMean = readRDS(paste0(dir, "mat.genes.geomMean.v10f.rds"))

mat.expr = mat.expr[, signaturegenes]
mat.expr <- apply(mat.expr, 2, function(x) (x - min(x)) / (max(x) - min(x)))

mat.expr = mat.expr[intersect(names(col_celltype), rownames(mat.expr)),]

genes.geomMean = genes.geomMean[signaturegenes]
col_fun = circlize::colorRamp2(seq(0, max(mat.expr), length = 7), c("grey99", pal_material("red")(10)[1:6]))

column_ha = HeatmapAnnotation(expression = anno_barplot(genes.geomMean, bar_width = 0.8, axis = F, border = F,
                                                        height = unit(1, "line"),
                                                        gp = gpar(fill = "grey78", col = "grey78")),
                              show_annotation_name = T,
                              annotation_name_side = "left", annotation_name_gp = gpar(fontsize = 8))

p.markerht2 = Heatmap(mat.expr,
                      rect_gp = gpar(col = "white", lwd = 0.5),
                      col = col_fun,
                      na_col = "grey93",
                      bottom_annotation = column_ha,
                      cluster_columns = F, show_row_names = T,
                      cluster_rows = F, show_row_dend = T, show_column_dend = T, row_dend_side = "left", column_dend_side = "top",
                      clustering_method_rows = "ward.D2", clustering_method_columns  = "ward.D2",
                      clustering_distance_rows  = "euclidean", clustering_distance_columns  = "euclidean", 
                      row_names_side = "left",
                      column_names_side = "top", 
                      column_names_gp = gpar(fontsize = 8, fontface = "italic"),
                      row_names_gp = gpar(fontsize = 9, fontface = "plain", col = col_celltype),
                      column_title = NULL, column_title_side = "top", column_title_gp = gpar(fontsize = 8, fontface = "plain"),
                      show_heatmap_legend = T,
                      heatmap_legend_param = list(title = "scaled expression", title_gp = gpar(fontsize = 8), labels_gp = gpar(fontsize = 8), legend_width = unit(0.2, "line"), legend_height = unit(1.8, "line"), grid_width = unit(0.1, "line"), grid_height = unit(1.5, "line"), at = c(0.3, max(mat.expr) - 0.3), labels = c("low", "high"), by_row = F, direction = "vertical", title_position = "leftcenter-rot"))
```

# Figure 1B-C

```{r}
cairo_pdf(file = paste0(dir.fig, "HNeuSA.fig1.umap.tissue.group.number.pdf"), width = 8.75, height = 3)
pushViewport(viewport(layout = grid.layout(nrow = 49, ncol = 140)))
print(p.ggalluvial, vp = viewport(layout.pos.row = 20:28, layout.pos.col = 1:85))
print(p.umap.tissue, vp = viewport(layout.pos.row = 1:20, layout.pos.col = 1:18))
print(p.umap.group, vp = viewport(layout.pos.row = 29:48, layout.pos.col = 1:18))
print(p.neu.pct.source2, vp = viewport(layout.pos.row = 4:29, layout.pos.col = 98:136))
print(p.neu.pct.technique2, vp = viewport(layout.pos.row = 30:47, layout.pos.col = 95:136))
dev.off()
```

# Figure 2&S1

```{r figure2}
cairo_pdf(file = paste0(dir.fig, "figure2.pdf"), width = 14.2, height = 8.6)
pushViewport(viewport(layout = grid.layout(nrow = 122, ncol = 115)))

genes2display = c("AZU1", "MPO", "CTSG", "DEFA4", "CEACAM8", "LTF", "MMP8", "CAMP", "ARG1", "MMP9", "S100A8", "OLR1", "ORM1", "IL1R2", "FOS", "EGR1", "IFI6", "GBP5", "CD274", "NFKBIA", "IL1B", "CXCR4", "VEGFA",  "CCL3", "SLPI", "PI3")

pushViewport(viewport(layout.pos.row = 115:122, layout.pos.col = 2:114))
grid.arrange(grobs=pgenes2[genes2display], nrow = 1, as.table = TRUE, newpage = FALSE)
upViewport(1)

print(pgenes$FCGR3B, vp = viewport(layout.pos.row = 2:15, layout.pos.col = 3:10))
print(pgenes$CXCR2, vp = viewport(layout.pos.row = 16:29, layout.pos.col = 3:10))
print(pgenes$MME, vp = viewport(layout.pos.row = 30:42, layout.pos.col = 3:10))
print(pgenes$`granule\ngenes`,vp = viewport(layout.pos.row = 43:58, layout.pos.col = 3:10))

print(p.umap.celltype, vp = viewport(layout.pos.row = 3:56, layout.pos.col = 20:52))
print(ggdraw() + draw_image(paste0(dir.fog, "Asset 5.png"), x = 0.5, y = 0.5, width = 0.91, height = 0.88, hjust = 0.5, vjust = 0.4),vp = viewport(layout.pos.row = 3:56, layout.pos.col = 20:52))

grid.text(x = unit(0.422, "npc"), y = unit(0.89, "npc"), label = "granule\nsubsets", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))
grid.text(x = unit(0.16, "npc"), y = unit(0.88, "npc"), label = "interferon-\nrelated\nsubsets", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))
grid.text(x = unit(0.17, "npc"), y = unit(0.73, "npc"), label = "cytokine/\ninflammation\nsubsets", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

print(p.umap.celltype.matureornot, vp = viewport(layout.pos.row = 38:55, layout.pos.col = 15:25))

print(p.tissuedistribution + theme(axis.text.y = element_blank(), plot.margin = margin(0.45,0.05,0.1,0.05, "line")),
      vp = viewport(layout.pos.row = 13:51, layout.pos.col = 61:71))

pushViewport(viewport(layout.pos.row = 3:52, layout.pos.col = 72:109))
draw(p.tissue.ht, merge_legend = T, heatmap_legend_side = "bottom", newpage = FALSE)
upViewport(1)

pushViewport(viewport(layout.pos.row = 5:56, layout.pos.col = 54:58))
draw(Legend(title = "", title_gp = gpar(fontsize = 10, col = textcolor),
            title_position = "topcenter", background = "white",
            at = names(col_celltype[1:21]), type = "points", pch = 16 , ncol = 1, by_row = F, labels_gp = gpar(fontsize = 10, col = col_celltype[1:21], fontface = "plain"), grid_height = unit(0.7, "line"), grid_width = unit(0.3, "line"), gap = unit(.4, "line"), legend_gp = gpar(alpha = 0.8, col = col_celltype[1:21])))
upViewport(1)

pushViewport(viewport(layout.pos.row = 61:114, layout.pos.col = 8:115))
draw(p.markerht2, merge_legend = T, heatmap_legend_side = "right",newpage = FALSE)
upViewport(1)

grid.brackets(0.141, 0.485, 0.325, 0.485, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.text(x = unit(0.233, "npc"), y = unit(0.505, "npc"), label = "granule genes", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor))

grid.brackets(0.45, 0.485, 0.66, 0.485, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.text(x = unit(0.5575, "npc"), y = unit(0.505, "npc"), label = "interferon-related genes", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor))

grid.brackets(0.667, 0.485, 0.876, 0.485, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.text(x = unit(0.7415, "npc"), y = unit(0.505, "npc"), label = "cytokine/inflammation genes", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor))

grid.brackets(0.09, .39, 0.09, .404, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.lines(x = unit(c(0.084, 0.084), "npc"), y = unit(c(0.397, 0.415), "npc"), gp = gpar(fill = "grey42", lwd = 0.7, col = "grey42"))
grid.text(x = unit(0.074, "npc"), y = unit(0.424, "npc"), label = "immature subsets", just = "center", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

grid.brackets(0.08, .37, 0.08, .41, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.text(x = unit(0.073, "npc"), y = unit(0.39, "npc"), label = "granule\nsubsets", just = "right", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

grid.brackets(0.08, .239, 0.08, .277, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.text(x = unit(0.073, "npc"), y = unit(0.258, "npc"), label = "interferon-\nrelated\nsubsets", just = "right", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

grid.brackets(0.08, .164, 0.08, .235, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.text(x = unit(0.073, "npc"), y = unit(0.1995, "npc"), label = "cytokine/\ninflammation\nsubsets", just = "right", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

grid.brackets(0.93, .856, 0.93, .817, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.lines(x = unit(c(0.942, 0.945), "npc"), y = unit(c(0.872, 0.872), "npc"), gp = gpar(fill = "grey42", lwd = 0.7, col = "grey42"))
grid.lines(x = unit(c(0.935, 0.945), "npc"), y = unit(c(0.8365, 0.8365), "npc"), gp = gpar(fill = "grey42", lwd = 0.7, col = "grey42"))
grid.lines(x = unit(c(0.945, 0.945), "npc"), y = unit(c(0.8365, 0.872), "npc"), gp = gpar(fill = "grey42", lwd = 0.7, col = "grey42"))

grid.text(x = unit(0.01, "npc"), y = unit(0.99, "npc"), label = "A", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.12, "npc"), y = unit(0.67, "npc"), label = "B", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.18, "npc"), y = unit(0.96, "npc"), label = "C", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.01, "npc"), y = unit(0.46, "npc"), label = "D", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.53, "npc"), y = unit(0.96, "npc"), label = "E", gp = gpar(fontsize = textsize, fontface = "bold"))

print(pthreecircle + theme(plot.margin = margin(0.08,0.03,0.08,0, "line")),
      vp = viewport(layout.pos.row = 53:55, layout.pos.col = 83:93))

dev.off()
```

```{r figureS1}
cairo_pdf(file = paste0(dir.fig, "figureS1.pdf"), width = 24, height = 12)
pushViewport(viewport(layout = grid.layout(nrow = 120, ncol = 100)))

pushViewport(viewport(layout.pos.row = 2:49, layout.pos.col = 1:100))
draw(p.markerht, merge_legend = T, heatmap_legend_side = "right",newpage = FALSE)
upViewport(1)

print(p.allsubset.pct + theme(plot.margin = margin(0.1, 0.2, 0.1, 0.1, "line")), vp = viewport(layout.pos.row = 51:120, layout.pos.col = 2:100))

grid.text(x = unit(0.01, "npc"), y = unit(0.98, "npc"), label = "A", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.01, "npc"), y = unit(0.58, "npc"), label = "B", gp = gpar(fontsize = textsize, fontface = "bold"))

dev.off()
```

# Figure S2

```{r umap}
dtf.subset = readRDS(paste0(dir, "dtf.cell911733.subset.sepsis_covid.v2a.rds"))
dtf.subset.genes = readRDS(paste0(dir, "pseudobulk.cell911733.genes267.rds"))
dtf.subset.genes = dtf.subset.genes + 1
min(dtf.subset.genes)
dtf.subset[,.N]

dtf.subset = cbind(dtf.subset, dtf.subset.genes)

dtf.subset$IMNeuS = getGeneScores(t(dtf.subset.genes), genes.immature.neu, "")
dtf.subset$TNeuS = getGeneScores(t(dtf.subset.genes), genes.total.neu, "")
dtf.subset$IFNS = getGeneScores(t(dtf.subset.genes), genes.ifn, "")

dtf.subset$group = samplemeta[match(dtf.subset$sampleid.pid, samplemeta$sampleid.pid)]$group
dtf.subset = dtf.subset[sampleid.pid %in%  dtf.s.pct[neu.pct >= 0.1]$sampleid.pid]
dtf.subset$pid = gsub("a$|b$|c$|d$","",dtf.subset$pid)
dtf.subset$celltype.neu = dtf[match(dtf.subset$cell.pid, dtf$cell.pid),]$celltype

length(unique(dtf.subset$sampleid.pid))

dtf.subset[,celltype:=ifelse(celltype.v1 == "Neu", "Mature Neu", celltype.v1)]

### pid ----
col_pid = c("#FFDC91FF","#4DBBD5FF", "#F39B7FFF","#3C5488FF","#7876B1FF"); names(col_pid)=sort(unique(dtf.subset$pid))

dtf.subset$facetvar = "study"
p.umap.pid =
  ScatterPlotRast(df = as.data.frame(dtf.subset), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.02, alphalevel = 0.2, colorby = "pid", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed =13, mycol = col_pid, backgroudcol = "white", plottitle = "", rasterdpi = 500, nacol = "grey42", facetlabelface = "plain", legendtextsize=8) +
  labs(title = NULL)+ 
  theme(text = element_text(size = textsize),
        legend.position = c(1.01,0.5),
        legend.justification = c(0,0.5),
        legend.spacing.x = unit(0.1,"line"),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.25,"line"),
        legend.background = element_blank(),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) + 
  guides(color = guide_legend(ncol=1, label.hjust = 0, override.aes = list(size=2.7,alpha=0.7))) 

### group ----
dtf.subset$facetvar = "condition"
dtf.subset$group = factor(dtf.subset$group,levels = names(col_group))
p.umap.group =
  ScatterPlotRast(df = as.data.frame(dtf.subset), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.02, alphalevel = 0.2, colorby = "group", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed =13, mycol = col_group, backgroudcol = "white", plottitle = "", rasterdpi = 500, nacol = "grey42", facetlabelface = "plain", legendtextsize=8) +
  labs(title = NULL)+
  theme(text = element_text(size = textsize),
        legend.position = c(1.01,0.5),
        legend.justification = c(0,0.5),
        legend.spacing.x = unit(0.1,"line"),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.25,"line"),
        legend.background = element_blank(),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) +
  guides(color = guide_legend(ncol=1, label.hjust = 0, override.aes = list(size=2.7,alpha=0.7)))

### celltype ----
col_celltype2 = c(`Mature Neu` = "#2E42B8", `Immat Neu` = "#0A73DC",
                  `CD14 Monocyte` = "#8CC269", `CD16 Monocyte` = "#70F3FF", cDC = "#B0A4E3", pDC = "#CCA4E3",
                  `Neu-Lym` = "#F39B7FFF", 
                  "Platelet" = "grey88", Erythrocyte = "#FD8CC1FF", HSPC = "#DC3023", "Mast" = "#352A87",
                  `CD4 T Naive` = "#7397AB", `CD8 T Naive` = "#D2F0F4", `CD4 T Eff/Mem` = "#33B7A0", `CD8 T Eff/Mem` = "#F86B1D",
                  Treg = "#1A6840", `Prolif T/NK` = "#F9906F", NK = "#FFA631", NKT = "#EFC000FF", 
                  B = "#F9FB0E", PB = "#FFDC91FF", "B-T" = "#003C67FF")

df.label= as.data.table(dtf.subset)[, lapply(.SD,median), by = celltype,.SDcols = c("UMAP_1", "UMAP_2")]
colnames(df.label)[2:3] = c("x","y")
df.label$labeltext=df.label[,"celltype", with = F]
df.label=df.label[!is.na(celltype)]
df.label = as.data.table(df.label)
df.label[, y:=ifelse(celltype == "CD4 T Naive", y + 1.5, y)]
df.label[, y:=ifelse(celltype == "CD4 T Eff/Mem", y - 0.5, y)]
df.label[, y:=ifelse(celltype == "NK", y + 0.5, y)]
df.label[, y:=ifelse(celltype == "CD16 Monocyte", y - 0.5, y)]
df.label[, x:=ifelse(celltype == "CD16 Monocyte", x + 0.9, x)]
df.label[, y:=ifelse(celltype == "CD14 Monocyte", y + 1, y)]

dtf.subset$facetvar = "cell type"
p.umap.celltype =
  ScatterPlotRast(df = as.data.frame(dtf.subset), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.01, alphalevel = 0.3, colorby = "celltype", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed =13, mycol = col_celltype2, backgroudcol = "white", plottitle = "", rasterdpi = 300, nacol = "grey42", labeldf = df.label, facetlabelface = "plain",repel = F, legendtextsize=8) +
  labs(title = NULL)+ 
  theme(text = element_text(size = textsize),
        legend.position = c(1.01,0.5),
        legend.justification = c(0,0.5),
        legend.spacing.x = unit(0.1,"line"),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.2,"line"),
        legend.background = element_blank(),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) + 
  guides(color = guide_legend(ncol = 1, label.hjust = 0, override.aes = list(size=2.7,alpha=0.7))) 

### celltype.neu ----
df.label= as.data.table(dtf.subset)[, lapply(.SD,median), by = celltype.neu,.SDcols = c("UMAP_1", "UMAP_2")]
colnames(df.label)[2:3] = c("x","y")
df.label$labeltext=df.label[,"celltype.neu", with = F]
df.label=df.label[!is.na(celltype.neu)]
df.label = as.data.table(df.label)

dtf.subset$facetvar = "cell type"
p.umap.celltype.neu =
  ScatterPlotRast(df = as.data.frame(dtf.subset), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.01, alphalevel = 0.3, colorby = "celltype.neu", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed =13, mycol = col_celltype, backgroudcol = "white", plottitle = "", rasterdpi = 300, nacol = "grey42", labeldf = df.label, facetlabelface = "plain",repel = F, legendtextsize=8) +
  labs(title = NULL)+ 
  theme(text = element_text(size = textsize),
        legend.position = c(1.01,0.5),
        legend.justification = c(0,0.5),
        legend.spacing.x = unit(0.1,"line"),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.2,"line"),
        legend.background = element_blank(),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) + 
  guides(color = guide_legend(ncol = 1, label.hjust = 0, override.aes = list(size=2.7,alpha=0.7))) 

### celltype.neu.main ----
dtf.subset[,celltype.neu.main:=ifelse(celltype.v1 == "Neu", "Mature Neu", celltype.v1)]
dtf.subset[,celltype.neu.main:=ifelse(grepl("Mono", celltype.neu.main), "Monocyte", celltype.neu.main)]
dtf.subset[,celltype.neu.main:=ifelse(grepl("CD4|CD8|Treg|NK", celltype.neu.main), "T and NK", celltype.neu.main)]
dtf.subset[,celltype.neu.main:=ifelse(grepl("Mast|Erythrocyte|HSPC|DC", celltype.neu.main), "other", celltype.neu.main)]
table(dtf.subset$celltype.neu.main)

col_celltype.neu.main = c(other = "grey80", Monocytes = "grey80", `T and NK` = "grey80", 
                          `Mature Neu` = "#2E42B8", PB = "grey80", Platelet = "grey80", 
                          `Immat Neu` = "#0A73DC", B = "grey80")

df.label= as.data.table(dtf.subset)[, lapply(.SD,median), by = celltype.neu.main,.SDcols = c("UMAP_1", "UMAP_2")]
colnames(df.label)[2:3] = c("x","y")
df.label$labeltext=df.label[,"celltype.neu.main", with = F]
df.label=df.label[!is.na(celltype.neu.main)][!celltype.neu.main %in% "other"]
df.label = as.data.table(df.label)

dtf.subset$facetvar = "cell type"
p.umap.celltype.neu.main =
  ScatterPlotRast(df = as.data.frame(dtf.subset), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = 0.0001, alphalevel = 0.05, colorby = "celltype.neu.main", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed =13, mycol = col_celltype.neu.main, backgroudcol = "grey97", plottitle = "", rasterdpi = 400, nacol = "grey42", labeldf = df.label, facetlabelface = "plain",repel = F, legendtextsize=8) +
  labs(title = NULL)+ 
  theme(text = element_text(size = textsize),
        legend.position = "none",
        legend.justification = c(0,0.5),
        legend.spacing.x = unit(0.1,"line"),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.2,"line"),
        legend.background = element_blank(),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) + 
  guides(color = guide_legend(ncol = 1, label.hjust = 0, override.aes = list(size=2.7,alpha=0.7))) 

### genes ----
pgenes.blood = list()
for(gene in c("S100A8", "S100A9", "PRTN3", "CTSG", "ELANE", "ARG1", "OLR1", "MMP9", "CD274", "CYP4F3", "CXCL1", "IL1R2", "ORM1", "SLPI", "PI3")){
  dtf.subset$id = dtf.subset[, gene, with = F]
  dtf.subset$facetvar = gene
  dtf.subset1 = dtf.subset[id>0]
  dtf.subset2 = dtf.subset[id<=0]
  tmp = rbind(dtf.subset2,dtf.subset1)
  pgenes.blood[[gene]] = ScatterPlotRast(as.data.frame(dtf.subset), featureX = "UMAP_1", featureY = "UMAP_2", groupby = "facetvar", pointsize = 0.0001, alphalevel = 0.1, colorby = "id", legendColumn = 2, colcateg= "continuous", numberofrows = 4, legendtextsize = 8, legendposition = "none", plottitle = NULL, seed = 1, backgroudcol = "white", mycol = c("grey99",dichromat_pal("DarkRedtoBlue.12")(12)[c(7:12)]), legendsize = 0.6, rasterdpi = 400, shuffle = T, contvarlower = 0.001, contvarupper = 0.999) + theme( plot.margin = unit(c(0.2,0.2,0.2,0.2), "line"))
}

pscores.blood = list()
for(gene in c("IMNeuS", "TNeuS", "IFNS")){
  dtf.subset$id = dtf.subset[, gene, with = F]
  dtf.subset$facetvar = gene
  dtf.subset1 = dtf.subset[id>0]
  dtf.subset2 = dtf.subset[id<=0]
  tmp = rbind(dtf.subset2,dtf.subset1)
  pscores.blood[[gene]] = ScatterPlotRast(as.data.frame(dtf.subset), featureX = "UMAP_1", featureY = "UMAP_2", groupby = "facetvar", pointsize = 0.0001, alphalevel = 0.1, colorby = "id", legendColumn = 2, colcateg= "continuous", numberofrows = 4, legendtextsize = 8, legendposition = "none", plottitle = NULL, seed = 1, backgroudcol = "grey97", mycol = c("grey99",dichromat_pal("DarkRedtoBlue.12")(12)[c(7:12)]), legendsize = 0.6, rasterdpi = 400, shuffle = T, contvarlower = 0.001, contvarupper = 0.999, facetlabelface = "plain") + theme( plot.margin = unit(c(0.2,0.2,0.2,0.2), "line"))
}

```

```{r ht}
dtfgenes.sum = readRDS(paste0(dir, "pseudobulk.cell911733.sum.sepsis_covid.v2.rds")) #  4472 14126
dtfgenes.mean = readRDS(paste0(dir, "pseudobulk.cell911733.mean.sepsis_covid.v2.rds")) #  4472 14126

dtfgenes.sum[,celltype.v1:=ifelse(celltype.v1 == "Neu", "Mature Neu", celltype.v1)]
dtfgenes.mean[,celltype.v1:=ifelse(celltype.v1 == "Neu", "Mature Neu", celltype.v1)]

dtf.s.pct = readRDS(paste0(dir, "dtf.s.pct.v10f.rds"))

length(unique(dtfgenes.sum$sampleid.pid)) # 262
length(unique(dtfgenes.sum[sampleid.pid %in% dtf.s.pct[neu.pct >= 0.1]$sampleid.pid]$sampleid.pid)) # 169

dtfgenes.sum.mean = dtfgenes.sum[sampleid.pid %in% dtf.s.pct[neu.pct >= 0.1]$sampleid.pid][,lapply(.SD, function(x){geomMean(x + 1)}), by = c("celltype.v1"), .SDcols = signaturegenes]
dtfgenes.mean.mean = dtfgenes.mean[sampleid.pid %in% dtf.s.pct[neu.pct >= 0.1]$sampleid.pid][,lapply(.SD, function(x){geomMean(x + 1)}), by = c("celltype.v1"), .SDcols = signaturegenes]

celltypeorder = c("Mature Neu", "Immat Neu", "CD14 Monocyte", "CD16 Monocyte", "cDC", 
                  "pDC", "Platelet", "Erythrocyte", "Mast", "HSPC", "CD4 T Naive", "CD8 T Naive", 
                  "CD4 T Eff/Mem", "CD8 T Eff/Mem", "Treg", "Prolif T/NK", "NK", 
                  "NKT", "B", "PB")

dtfgenes.sum[,celltype.v2:=ifelse(grepl("Neu",celltype.v1), "Neutrophil", "other")]

dtfgenes.sum.sum = dtfgenes.sum[sampleid.pid %in% dtf.s.pct[neu.pct >= 0.1]$sampleid.pid][,lapply(.SD, function(x){sum(x)}), by = c("celltype.v2"), .SDcols = signaturegenes]

# mean ht ----
mat.rownames = dtfgenes.mean.mean$celltype.v1
mat.expr = as.matrix(dtfgenes.mean.mean[,2:ncol(dtfgenes.mean.mean)])
rownames(mat.expr) = mat.rownames
mat.expr = mat.expr[celltypeorder, signaturegenes]
col_fun = circlize::colorRamp2(seq(min(mat.expr), max(mat.expr), length=11), c("#fdf8f3",rev(rocket(10))))

p.mean = Heatmap(mat.expr,
                 rect_gp = gpar(col = "white", lwd = 0.5),
                 col = col_fun,
                 na_col = "grey93",
                 cluster_columns = F,show_row_names = T,
                 cluster_rows = F,show_row_dend = T,show_column_dend = T,row_dend_side = "left", column_dend_side = "top",
                 clustering_method_rows = "ward.D2",clustering_method_columns  = "ward.D2",
                 clustering_distance_rows  = "euclidean", clustering_distance_columns  = "euclidean", 
                 row_names_side = "left",
                 column_names_side = "top", #column_names_rot = 90,
                 column_names_gp = gpar(fontsize = 8, fontface = "italic"),
                 row_names_gp = gpar(fontsize = 8, fontface = "plain"),
                 column_title = NULL, column_title_side = "top", column_title_gp = gpar(fontsize = 8, fontface = "plain"),
                 show_heatmap_legend = T,
                 heatmap_legend_param = list(title = "expression", title_gp = gpar(fontsize = 8), labels_gp = gpar(fontsize = 8), legend_width = unit(0.2,"line"), legend_height=unit(1.8,"line"),grid_width = unit(0.1,"line"),grid_height=unit(1.5,"line"), at = c(0.3, max(mat.expr) - 0.3), labels = c("low","high"), by_row = F, direction = "vertical", title_position = "leftcenter-rot"))

# mean ht scaled ----
neu.expr =  readRDS(paste0(dir, "neu.genes.expr.rds")) 
mat.colanno2 = as.data.frame(neu.expr[gene %in% signaturegenes][,c("pct.total", "gene")])
mat.colanno2$other = 100 - mat.colanno2$pct.total
colnames(mat.colanno2)[1] = "Neutrophil"
rownames(mat.colanno2) = mat.colanno2$gene
mat.colanno2$gene = NULL
mat.colanno2 = as.matrix(mat.colanno2)
mat.colanno2 = mat.colanno2[signaturegenes,]

column_ha2 = HeatmapAnnotation("Neu reads PCT" = anno_barplot(mat.colanno2, bar_width = 0.8, axis = F, border = F,
                                                              height = unit(2,"line"),
                                                              gp = gpar(fill = c("#2E42B8","grey92"), col = c("#2E42B8","grey92"))),
                               show_annotation_name = T,
                               annotation_name_side = "left", annotation_name_gp = gpar(fontsize = 8))

p.mean.scaled2 = Heatmap(mat.expr,
                         rect_gp = gpar(col = "white", lwd = 0.5),
                         col = col_fun,
                         na_col = "grey93",
                         bottom_annotation = column_ha2,
                         cluster_columns = F,show_row_names = T,
                         cluster_rows = F,show_row_dend = T,show_column_dend = T,row_dend_side = "left", column_dend_side = "top",
                         clustering_method_rows = "ward.D2",clustering_method_columns  = "ward.D2",
                         clustering_distance_rows  = "euclidean", clustering_distance_columns  = "euclidean", 
                         row_names_side = "left",
                         column_names_side = "top", #column_names_rot = 90,
                         column_names_gp = gpar(fontsize = 8, fontface = "italic"),
                         row_names_gp = gpar(fontsize = 8, fontface = "plain"),
                         column_title = NULL, column_title_side = "top", column_title_gp = gpar(fontsize = 8, fontface = "plain"),
                         show_heatmap_legend = T,
                         heatmap_legend_param = list(title = "scaled expression", title_gp = gpar(fontsize = 8), labels_gp = gpar(fontsize = 8), legend_width = unit(0.2,"line"), legend_height=unit(1.8,"line"),grid_width = unit(0.1,"line"),grid_height=unit(1.5,"line"), at = c(0.3, max(mat.expr) - 0.3), labels = c("low","high"), by_row = F, direction = "vertical", title_position = "leftcenter-rot"))
```

```{r figure}
cairo_pdf(file = paste0(dir.fig, "figureS2.pdf"), width = 14, height = 9.5)
pushViewport(viewport(layout=grid.layout(nrow=284,ncol = 200)))

print(p.umap.pid,vp=viewport(layout.pos.row = 4:42, layout.pos.col = 5:21))
print(p.umap.group,vp=viewport(layout.pos.row = 45:83, layout.pos.col = 5:21))
print(p.umap.celltype,vp=viewport(layout.pos.row = 4:83, layout.pos.col = 41:77))


pushViewport(viewport(layout.pos.row = 4:87, layout.pos.col = 108:180))
grid.arrange(grobs=pgenes.blood, nrow = 3, as.table = TRUE, newpage = FALSE)
upViewport(1)

pushViewport(viewport(layout.pos.row = 85:183, layout.pos.col = 2:199))
draw(p.mean,merge_legend=F,heatmap_legend_side = "right",newpage = FALSE)
upViewport(1)

pushViewport(viewport(layout.pos.row = 185:283, layout.pos.col = 2:199))
draw(p.mean.scaled2,merge_legend=F,heatmap_legend_side = "right",newpage = FALSE)
upViewport(1)

grid.text(x = unit(0.012,"npc"), y=unit(0.99,"npc"), label="A",gp=gpar(fontsize = textsize,fontface = "bold"))
grid.text(x = unit(0.012,"npc"), y=unit(0.845,"npc"), label="B",gp=gpar(fontsize = textsize,fontface = "bold"))
grid.text(x = unit(0.192,"npc"), y=unit(0.99,"npc"), label="C",gp=gpar(fontsize = textsize,fontface = "bold"))
grid.text(x = unit(0.52,"npc"), y=unit(0.99,"npc"), label="D",gp=gpar(fontsize = textsize,fontface = "bold"))

grid.text(x = unit(0.012,"npc"), y=unit(0.65,"npc"), label="E",gp=gpar(fontsize = textsize,fontface = "bold"))
grid.text(x = unit(0.012,"npc"), y=unit(0.3,"npc"), label="F",gp=gpar(fontsize = textsize,fontface = "bold"))
grid.text(x = unit(0.01,"npc"), y=unit(0.05,"npc"), label="G",gp=gpar(fontsize = textsize,fontface = "bold"))
dev.off()
```

# Figure 3&S3

```{r subset distribution}
col_infect = c(`COVID-19/flu,nonsevere` = "#7b72c5", `COVID-19,severe` = "#EFC000", `bacterial sepsis` = "#F97D1C")
col_infect2 = c(`healthy/convalescent` = "#1A6840", `COVID-19/flu,nonsevere` = "#7b72c5",
                `COVID-19,severe` = "#EFC000",  `bacterial sepsis` = "#F97D1C")

dtf.s.pct.infection = dtf.s.pct[pid %in% unique(dtf[grepl("COVID|sepsis|flu|conv",group)]$pid)][grepl("COVID|sepsis|flu|conv|healthy", group)][tissue %in% c("peripheral blood")][grepl("WB", source)] # 179
table(dtf.s.pct.infection$group) 

dtf.s.pct.infection[,condition:=ifelse(group %in% c("COVID-19,nonsevere", "flu", "flu,pregnancy"), "COVID-19/flu,nonsevere",
                                       ifelse(group %in% c("healthy", "convalescent"), "healthy/convalescent", group
                                       ))]
dtf.s.pct.infection = dtf.s.pct.infection[neu.pct >= 0.05][number.neutrophil >= 100] 

table(dtf.s.pct.infection$condition)
table(dtf.s.pct.infection$condition, dtf.s.pct.infection$pid)

dtf.s.pct.m = melt(dtf.s.pct.infection,
                   id.vars = c("condition", "pid", "sampleid.pid"),
                   measure.vars = names(col_celltype))

dtf.s.pct.m = as.data.table(dtf.s.pct.m)

dtf.s.pct.m.aggmedian = dtf.s.pct.m[,lapply(.SD, function(x){median(x)}), by = c("condition",  "variable"), .SDcols = c("value")]
dtf.s.pct.m.aggsd = dtf.s.pct.m[,lapply(.SD, function(x){sd(x)}), by = c("condition",  "variable"), .SDcols = c("value")]
dtf.s.pct.m.agg = dtf.s.pct.m[,lapply(.SD, function(x){mean(x)}), by = c("condition",  "variable"), .SDcols = c("value")]
dtf.s.pct.m.agg[,lapply(.SD, function(x){sum(x)}), by = c("condition"), .SDcols = c("value")]

dtf.s.pct.m.agg$variable = factor(dtf.s.pct.m.agg$variable , levels = rev(names(col_celltype)))
dtf.s.pct.m.aggmedian$variable = factor(dtf.s.pct.m.aggmedian$variable , levels = rev(names(col_celltype)))
dtf.s.pct.m.aggsd$variable = factor(dtf.s.pct.m.aggsd$variable , levels = rev(names(col_celltype)))

dtf.s.pct.m.agg$facetvar = "neutrophil subset in infectious diseases (whole blood)"
p.infectdistribution = 
  ggplot(dtf.s.pct.m.agg, aes(x = condition, y = value, fill = variable)) +
  geom_bar(stat = "identity", alpha = 0.8) +
  theme_bw() + mytheme +
  scale_y_continuous(expand = c(0, 0), breaks = seq(0, 1, 0.2), labels = seq(0, 1, 0.2), position = "right") + 
  scale_x_discrete(position = "top") + 
  labs(x = NULL, y = NULL , title = "neutrophil subset in infectious diseases (whole blood)") +
  theme(
    panel.grid.major = element_blank(), 
    panel.border = element_blank(),
    legend.title =  element_blank(),
    legend.text = element_text(color = textcolor, face = "italic"),
    legend.position = "none",
    legend.spacing.y = unit(0.01, "line"), 
    legend.spacing.x = unit(0.04, "line"),
    legend.key.spacing.x = unit(.1, "line"),
    legend.key.spacing.y = unit(.1, "line"),
    legend.key.width = unit(0.2, "line"),
    legend.key.height = unit(0.2, "line"),
    plot.title = element_text(size = textsize, face = "plain", hjust = 0, margin = margin(t = 0.1, r = 0.1, b = 0.1, l = 0.1, unit = "line"))) +
  scale_fill_manual(values = col_celltype, name = "") + coord_flip()  +
  guides(fill = guide_legend(ncol = 1, byrow = T, size = 2.3)) 

# box ----
shape_infect = c(c("wilk2021" = 3, "sinha2021" = 8, "schrepping2020" = 17, "zhang2023" = 18, 
                   "kwok2023" = 15, "combes2021" = 16, "kaiser2024" = 10))
dtf.s.pct.m$condition = factor(dtf.s.pct.m$condition, levels = names(col_infect2))

p.infect.box = ggplot(dtf.s.pct.m, aes(x = condition, y = value, color = condition)) +
  ggbeeswarm::geom_quasirandom(size = .2, stroke = 0.2, width = 0.2, show.legend = T, alpha = 0.5, varwidth = F, aes(shape = pid)) +
  geom_boxplot(width = 0.2, size = 0.2, outlier.shape = NA, alpha = 0.8, fill = NA)+
  labs(y = "proportion in total neutrophils", x = NULL, title = "") +
  scale_y_continuous(breaks = pretty_breaks()) +
  theme_bw() + mytheme + 
  theme(
    axis.text.x = element_blank(),
    plot.margin = unit(c(.2,.2,.2,.2), "line"),
    panel.grid.major.x = element_blank(),
    plot.title = element_text(size = textsize, face = "plain",hjust = 0,margin = margin(t = .1, r = 0, b = .1, l = 0, unit = "pt"))
  ) +
  facet_wrap(.~variable, nrow = 3, scales = "free") +
  scale_color_manual(values = col_infect2) +
  scale_fill_manual(values = col_infect2) +
  scale_shape_manual(values = shape_infect) +
  stat_compare_means(comparisons = list(c("healthy/convalescent", "COVID-19/flu,nonsevere"), c("healthy/convalescent", "COVID-19,severe"), c("healthy/convalescent", "bacterial sepsis"), c("COVID-19,severe", "COVID-19/flu,nonsevere")),method = "wilcox.test",vjust = 1.5, size = 1.8, label = "p.format", color="grey17", bracket.size = 0.1,tip.length = 0.01, step.increase = c(0,0.08,0.08,-0.025)) +
  guides(colour = guide_legend(override.aes = list(size = 2)), shape = guide_legend(override.aes = list(size = 2))) 
```

```{r proportion meta analysis}
dtf.s.pct$source = samplemeta[match(dtf.s.pct$sampleid.pid, samplemeta$sampleid.pid)]$source

dtf.s.pct.infection = dtf.s.pct[pid %in% unique(dtf[grepl("COVID|sepsis|flu|conv", group)]$pid)][grepl("COVID|sepsis|flu|conv|healthy", group)][tissue %in% c("peripheral blood")][grepl("WB", source)]

table(dtf.s.pct.infection$group)
table(dtf.s.pct.infection$pid)
colnames(dtf.s.pct.infection)

dtf.s.pct.infection[,condition:=ifelse(group %in% c("COVID-19,nonsevere", "flu", "flu,pregnancy"), "COVID-19/flu,nonsevere",
                                       ifelse(group %in% c("healthy", "convalescent"), "healthy/convalescent",
                                              group
                                       )
)]
dtf.s.pct.infection = dtf.s.pct.infection[neu.pct >= 0.05][number.neutrophil >= 100]
table(dtf.s.pct.infection$condition)
table(dtf.s.pct.infection$condition, dtf.s.pct.infection$pid)

dtf.s.pct.infection$sex = samplemeta[match(dtf.s.pct.infection$sampleid.pid, samplemeta$sampleid.pid)]$sex

table(dtf.s.pct.infection[sex == "F"]$condition, dtf.s.pct.infection[sex == "F"]$pid)
table(dtf.s.pct.infection[sex == "M"]$condition, dtf.s.pct.infection[sex == "M"]$pid)

dtf.s.pct.m = melt(dtf.s.pct.infection,
                   id.vars = c("condition", "pid", "sampleid.pid"),
                   measure.vars = names(col_celltype))
dtf.s.pct.m = as.data.table(dtf.s.pct.m)

dtf.s.pct.m.agg = dtf.s.pct.m[,lapply(.SD, function(x){mean(x)}), by = c("condition",  "variable"), .SDcols = c("value")]
dtf.s.pct.m.agg[,lapply(.SD, function(x){sum(x)}), by = c("condition"), .SDcols = c("value")]

dtf.s.pct.m.agg$variable = factor(dtf.s.pct.m.agg$variable , levels = rev(names(col_celltype)))

ESlist = list()
for(id in unique(dtf.s.pct.m$variable)){
  ESlist[[paste0(id, ".COVID-19,severe.vs.healthy/convalescent")]] = metaStats(dtf.s.pct.m[order(pid)], geneid = id, pids = c("combes2021", "schrepping2020", "sinha2021", "wilk2021"), case = "COVID-19,severe", control = "healthy/convalescent")
  
  ESlist[[paste0(id, ".COVID-19/flu,nonsevere.vs.healthy/convalescent")]] = metaStats(dtf.s.pct.m[order(pid)], geneid = id, pids = c("combes2021", "schrepping2020", "zhang2023"), case = "COVID-19/flu,nonsevere", control = "healthy/convalescent")
  
  ESlist[[paste0(id, ".bacterial sepsis.vs.healthy/convalescent")]] = metaStats(dtf.s.pct.m[order(pid)], geneid = id, pids = c("combes2021", "kaiser2024", "kwok2023", "sinha2021"), case = "bacterial sepsis", control = "healthy/convalescent")
}

datadf = rbindlist(ESlist, idcol = 'group')
datadf = as.data.table(datadf)
datadf = datadf[rowname == "summary"]
datadf$group = sapply(strsplit(datadf$group, "\\."), "[[", 2)
datadf$FDR = p.adjust(datadf$pval, method = "fdr")
datadf$celltype = factor(datadf$gene, levels = names(col_celltype))
datadf[,sig:=ifelse(FDR <= 0.2, "FDR ≤ 0.2", "FDR > 0.2")]

nth <- function(x, i) {
  x[(i - 1)%%length(x) + 1]
}
ci.value <- -qnorm((1 - 0.95) / 2)

datadf$lower <- datadf$ES.g - ci.value * datadf$ES.se
datadf$upper <- datadf$ES.g + ci.value * datadf$ES.se

datadf$group = factor(datadf$group, levels = c("bacterial sepsis", "COVID-19,severe", "COVID-19/flu,nonsevere"))

p.es.summary.infect = 
  ggplot(data = datadf, aes(x = celltype, y = ES.g, color = group, fill = group, alpha = sig)) +
  geom_hline(yintercept = 0, size = 0.3, alpha = 1, color = "grey67") +
  geom_tile(width = 0.7, height = 0.1, color = NA, position = position_dodge(0.7)) +
  geom_errorbar(data = datadf, aes(x = celltype, ymin =lower, ymax = upper,  group = group), 
                position = position_dodge(0.7), width = .7, size = 0.2) +
  labs(x = NULL, y = expression("effect size" ~ italic("vs.") ~ "healthy/convalescent"), title = "meta-analysis of proportion") +
  scale_color_manual(values = col_infect, breaks = names(col_infect), guide = "none") +
  scale_fill_manual(values = col_infect, breaks = names(col_infect), guide = "none") +
  scale_y_continuous(breaks = seq(-10, 2, 1)) +  
  scale_alpha_manual(values = c("FDR > 0.2" = 0.2, "FDR ≤ 0.2" = 1)) +
  theme_bw() + mytheme + 
  theme(plot.title = element_text(face = 1, margin = margin(0, 0, 0, 0, "line")), 
        panel.grid.major.y = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "line"),
        axis.text.x = element_text(size = 10),
        axis.text.y = element_text(size = 10, face = "plain"),
        axis.title = element_text(size = 10),
        legend.position = c(1.25, 0.19),
        legend.justification = c(0, 0),
        legend.text = element_text(size = 9),
        legend.title = element_blank(),
        legend.background =  element_blank(),
        legend.box.background = element_blank(),
        legend.spacing.y = unit(0.01, "line"),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.3, "line")) + 
  coord_flip() +
  scale_x_discrete(limits = rev, position = "top")  +
  guides(color = "none", shape = "none", size = "none")
```

```{r gene exprssion meta analysis}
infect.exprmeta = readRDS(paste0(dir, "infect.exprmeta.selectedgenes.neu100.v10f.rds"))

mat.es.granule = dcast(infect.exprmeta[celltype == "granule subsets"], gene ~ group, value.var = "ES.g")
mat.es.IFN = dcast(infect.exprmeta[celltype == "IFN subsets"], gene ~ group, value.var = "ES.g")

rowname = mat.es.granule$gene
mat.es.granule$gene = NULL
mat.es.IFN$gene = NULL

colnames(mat.es.granule) = paste0(colnames(mat.es.granule), ",granule subsets")
colnames(mat.es.IFN) = paste0(colnames(mat.es.IFN), ",IFN subsets")

mat.es = as.matrix(cbind(mat.es.granule, mat.es.IFN))
rownames(mat.es) = rowname

mat.pval.granule = dcast(infect.exprmeta[celltype == "granule subsets"], gene ~ group, value.var = "pval")
mat.pval.IFN = dcast(infect.exprmeta[celltype == "IFN subsets"], gene ~ group, value.var = "pval")

rowname = mat.pval.granule$gene
mat.pval.granule$gene = NULL
mat.pval.IFN$gene = NULL

colnames(mat.pval.granule) = paste0(colnames(mat.pval.granule), ",granule subsets")
colnames(mat.pval.IFN) = paste0(colnames(mat.pval.IFN), ",IFN subsets")

mat.pval = as.matrix(cbind(mat.pval.granule, mat.pval.IFN))
rownames(mat.pval) = rowname

max(mat.es)
min(mat.es)
ESmax = max(abs(mat.es))
mat.pval[mat.pval > 0.05] = 1
mat.size = 1 - mat.pval

col_fun = circlize::colorRamp2(seq(-ESmax, ESmax, length =13), c(dichromat_pal("DarkRedtoBlue.12")(12)[1:6], "white",dichromat_pal("DarkRedtoBlue.12")(12)[7:12]))

pht =
  Heatmap(as.matrix(mat.es),
          rect_gp = gpar(type = "none"), col = col_fun,
          na_col = "grey",
          cluster_columns = F, cluster_rows = T,
          show_row_dend = F, show_column_dend = F,
          column_dend_side  = "bottom",
          clustering_method_rows = "ward.D2", clustering_method_columns  = "ward.D2",
          clustering_distance_rows  = "pearson", clustering_distance_columns  = "pearson",
          column_names_side = "top", column_names_rot = 90, column_names_gp = gpar(fontsize = textsize),
          row_names_gp = gpar(fontsize = textsize, fontface = "plain"),
          row_names_max_width = unit(30, "line"),
          column_title = "", column_title_side = "bottom", column_title_gp = gpar(fontsize = textsize, fontface = "italic"),
          show_heatmap_legend = showheatmaplegend,
          cell_fun =  function(j, i, x, y, width, height, fill){
            grid.rect(x = x, y = y, width = width, height =  height, gp = gpar(lwd = gridlwd, col = "grey93",fill = NA, alpha=0.2))
            grid.circle(x = x, y = y, r = 8 * mat.size[i, j]  * min(unit.c(width, height)), gp = gpar(fill = col_fun(mat.es[i, j]), col = NA, alpha=0.9))},
          heatmap_legend_param = list(title = "NES", title_gp = gpar( fontsize = textsize), labels_gp = gpar(fontsize = textsize), legend_height = unit(3, "line"), legend_width = unit(.2, "line"), grid_height = unit(3, "line"), grid_width = unit(.1, "line"), at = seq(-ESmax/1.5, ESmax/1.5, length =3), labels=c("low", "0", "high"), by_row = F, direction = "vertical", title_position = "topcenter"))

# granule subsets ----
colorder = c("COVID-19/flu,nonsevere", "COVID-19,severe", "bacterial sepsis")
mat.es.granule = dcast(infect.exprmeta[celltype == "granule subsets"], gene ~ group, value.var = "ES.g")
rowname = mat.es.granule$gene
mat.es.granule$gene = NULL
mat.es.granule = as.matrix(mat.es.granule)
rownames(mat.es.granule) = rowname

mat.pval.granule = dcast(infect.exprmeta[celltype == "granule subsets"], gene ~ group, value.var = "pval")
mat.pval.granule$gene = NULL
mat.pval.granule = as.matrix(mat.pval.granule)
rownames(mat.pval.granule) = rowname

mat.pval = as.matrix(cbind(mat.pval.granule, mat.pval.IFN))
rownames(mat.pval) = rowname

genes2include = c("STMN1", "MKI67", "FUT4" , "AZU1", "MPO", "PRTN3", "CTSG", "ELANE", "DEFA4", "CEACAM6", "DEFA3", 
                  "BPI", "LYZ", "RETN", "ANXA1", "CEACAM8", "OLFM4", 
                  "TCN1",  "LTF", "LCN2",  "MMP8", "CAMP", "CRISP3", "HP", "PGLYRP1", 
                  "ANXA3",  "FCN1", "ARG1",  "PADI4", "CD177", "MMP9", 
                  "NQO2",  "CYP4F3", "S100A8", "S100A9", "S100A12", "TXN",
                  "TSPO", "PLAC8",  "GPI", "OLR1",    "ORM1")

mat.es.granule = mat.es.granule[genes2include, colorder]
mat.pval.granule = mat.pval.granule[genes2include, colorder]

max(mat.es.granule)
min(mat.es.granule)
ESmax = 2.5
mat.es.granule[mat.es.granule > ESmax] = ESmax
mat.es.granule[mat.es.granule < -ESmax] = -ESmax

mat.fdr.granule = apply(mat.pval.granule, 2, function(x){p.adjust(x, method = "fdr")})
mat.size.granule = 1 - mat.fdr.granule
mat.fdr.granule[mat.fdr.granule > 0.2] = 1
mat.size2.granule = 1 - mat.fdr.granule

col_fun = circlize::colorRamp2(seq(-ESmax,ESmax, length = 13), c(dichromat_pal("DarkRedtoBlue.12")(12)[1:6], "white", dichromat_pal("DarkRedtoBlue.12")(12)[7:12]))

colanno = HeatmapAnnotation(
  group = c("COVID-19/flu,nonsevere", "COVID-19,severe", "bacterial sepsis"),
  col = list(group = c("COVID-19/flu,nonsevere" = "#7b72c5", "COVID-19,severe" = "#EFC000", "bacterial sepsis" = "#F97D1C")),
  annotation_legend_param = list(
    group = list(title = "group", title_position = "topleft", title_gp = gpar(fontsize = textsize), labels_gp = gpar(fontsize = textsize, col = textcolor), legend_gp = gpar(fontsize = textsize), direction = "horizontal", nrow = 1, legend_width = unit(.3, "line"), grid_width = unit(.3, "line"), legend_height = unit(.3, "line"), grid_height = unit(.3, "line"))),
  show_legend = F,
  show_annotation_name = F, annotation_name_gp = gpar(fontsize = textsize),
  annotation_name_side = "left",
  simple_anno_size = unit(0.3, "line")
)

pht.granulesub =
  Heatmap(as.matrix(mat.es.granule),
          rect_gp = gpar(type = "none"),
          col = col_fun,
          na_col = "grey",
          cluster_columns = F, cluster_rows = F,
          show_row_dend = F, show_column_dend = F,
          bottom_annotation = colanno,
          column_dend_side  = "bottom",
          clustering_method_rows = "ward.D2", clustering_method_columns = "ward.D2",
          clustering_distance_rows = "pearson", clustering_distance_columns = "pearson",
          show_column_names = F,
          column_names_side = "top", column_names_rot = 45, column_names_gp = gpar(fontsize = textsize),
          row_names_gp = gpar(fontsize = textsize, fontface = "italic"),
          row_names_max_width = unit(30, "line"),
          column_title = "            granule subsets", column_title_side = "top",
          column_title_gp = gpar(fontsize = textsize, fontface = "plain"),
          show_heatmap_legend = F,
          cell_fun =  function(j, i, x, y, width, height, fill){
            grid.rect(x = x, y = y, width = width, height =  height, gp = gpar(lwd = gridlwd, col = "grey53",fill = NA, alpha=0.2))
            grid.circle(x = x, y = y, r = 4 * mat.size.granule[i, j]  * min(unit.c(width, height)), gp = gpar(fill = col_fun(mat.es.granule[i, j]), col = NA, alpha=0.9))
            grid.circle(x = x, y = y, r = 4 * mat.size2.granule[i, j]  * min(unit.c(width, height)), gp = gpar(fill = NA, col = "grey17", alpha=1))},
          heatmap_legend_param = list(title = "ES", title_gp = gpar( fontsize = textsize), labels_gp = gpar(fontsize = textsize), legend_height = unit(3, "line"), legend_width = unit(.2, "line"), grid_height = unit(3, "line"), grid_width = unit(.1, "line"), at = seq(-2, 2, length =3), by_row = F, direction = "vertical", title_position = "topleft"))

# IFN subsets ----
colorder = c("COVID-19/flu,nonsevere", "COVID-19,severe", "bacterial sepsis")
mat.es.IFN = dcast(infect.exprmeta[celltype == "IFN subsets"], gene ~ group, value.var = "ES.g")
rowname = mat.es.IFN$gene
mat.es.IFN$gene = NULL
mat.es.IFN = as.matrix(mat.es.IFN)
rownames(mat.es.IFN) = rowname

mat.pval.IFN = dcast(infect.exprmeta[celltype == "IFN subsets"], gene ~ group, value.var = "pval")
mat.pval.IFN$gene = NULL
mat.pval.IFN = as.matrix(mat.pval.IFN)
rownames(mat.pval.IFN) = rowname

mat.pval = as.matrix(cbind(mat.pval.IFN, mat.pval.IFN))
rownames(mat.pval) = rowname

genes2include = c("MT2A", "MT1X", "HES4", "LY6E", "ISG15", "IFI44L", "IFI44",  "OASL",   
                  "IFI6", "OAS3", "OAS2", "OAS1" , "RSAD2", "HERC5", "SAMD9", "SAMD9L", "CMPK2",
                  "MX1", "IFIT1", "IFIT2", "IFIT3", "IFIT5",  "DDX58",  "XAF1",  "IRF7", "EPSTI1",
                  "APOL6",  "GBP5", "GBP4", "GBP2", "GBP1", "PARP14",
                  "IFIH1", "TNFSF10", "TNFSF13B" , "IFITM1", "IFITM3")
#genes2include = genes.ifn

mat.es.IFN = mat.es.IFN[genes2include, colorder]
mat.pval.IFN = mat.pval.IFN[genes2include, colorder]

max(mat.es.IFN)
min(mat.es.IFN)
ESmax = 2.5
mat.es.IFN[mat.es.IFN > ESmax] = ESmax
mat.es.IFN[mat.es.IFN < -ESmax] = -ESmax

mat.fdr.IFN = apply(mat.pval.IFN, 2, function(x){p.adjust(x, method = "fdr")})
mat.size.IFN = 1 - mat.fdr.IFN
mat.fdr.IFN[mat.fdr.IFN > 0.2] = 1
mat.size2.IFN = 1 - mat.fdr.IFN

col_fun = circlize::colorRamp2(seq(-ESmax,ESmax, length = 13), c(dichromat_pal("DarkRedtoBlue.12")(12)[1:6], "white", dichromat_pal("DarkRedtoBlue.12")(12)[7:12]))

colanno = HeatmapAnnotation(
  group = c("COVID-19/flu,nonsevere", "COVID-19,severe", "bacterial sepsis"),
  col = list(group = c("COVID-19/flu,nonsevere" = "#7b72c5", "COVID-19,severe" = "#EFC000", "bacterial sepsis" = "#F97D1C")),
  annotation_legend_param = list(
    group = list(title = "group", title_position = "topleft", title_gp = gpar(fontsize = textsize), labels_gp = gpar(fontsize = textsize, col = textcolor), legend_gp = gpar(fontsize = textsize), direction = "horizontal", nrow =1, legend_width = unit(.3, "line"), grid_width = unit(.3, "line"), legend_height = unit(.3, "line"), grid_height = unit(.3, "line"))),
  show_legend = F,
  show_annotation_name = F, annotation_name_gp = gpar(fontsize = textsize),
  annotation_name_side = "left",
  simple_anno_size = unit(0.3, "line")
)

pht.IFNsub =
  Heatmap(as.matrix(mat.es.IFN),
          rect_gp = gpar(type = "none"),
          col = col_fun,
          na_col = "grey",
          cluster_columns = F, cluster_rows = F,
          show_row_dend = F, show_column_dend = F,
          bottom_annotation = colanno,
          column_dend_side  = "bottom",
          clustering_method_rows = "ward.D2", clustering_method_columns  = "ward.D2",
          clustering_distance_rows  = "pearson", clustering_distance_columns  = "pearson",
          show_column_names = F,
          column_names_side = "top", column_names_rot = 45, column_names_gp = gpar(fontsize = textsize),
          row_names_gp = gpar(fontsize = textsize, fontface = "italic"),
          row_names_max_width = unit(30, "line"),
          column_title = "    IFN subsets", column_title_side = "top",
          column_title_gp = gpar(fontsize = textsize, fontface = "plain"),
          show_heatmap_legend = T,
          cell_fun =  function(j, i, x, y, width, height, fill){
            grid.rect(x = x, y = y, width = width, height =  height, gp = gpar(lwd = gridlwd, col = "grey53",fill = NA, alpha=0.2))
            grid.circle(x = x, y = y, r = 3.6 * mat.size.IFN[i, j]  * min(unit.c(width, height)), gp = gpar(fill = col_fun(mat.es.IFN[i, j]), col = NA, alpha = 0.9))
            grid.circle(x = x, y = y, r = 3.6 * mat.size2.IFN[i, j]  * min(unit.c(width, height)), gp = gpar(fill = NA, col = "grey17", alpha = 1))},
          heatmap_legend_param = list(title = "effect size", title_gp = gpar( fontsize = textsize), labels_gp = gpar(fontsize = textsize), legend_width = unit(3, "line"), legend_height = unit(.2, "line"), grid_width = unit(3, "line"), grid_height = unit(.1, "line"), at = seq(-2, 2, length = 3), by_row = F, direction = "horizontal", title_position = "topcenter"))

```

```{r figure3}
dt.infect = data.frame(
  "combes2021" = c(2,9,10,4), 
  "schrepping2020" = c(6,4,13,0), 
  "zhang2023" = c(2,6,0,0),
  "sinha2021" = c(4,0,18,8),
  "wilk2021" = c(6,0,11,0), 
  "kaiser2024" = c(6,0,0,7),
  "kwok2023" = c(15,0,0,26),
  "total" = c(41,19,52,45)
)

df.pids = data.table(
  pid = colnames(dt.infect),
  y = 1,
  x = 1:ncol(dt.infect)
)
p.pids = ggplot(df.pids, aes(x = x, y = y, label = pid)) +
  geom_text(angle = 30, size = 2.6, hjust = 0, color = textcolor) + 
  theme_void() +
  coord_cartesian(clip = "off") 

dt.infect = tableGrob(dt.infect,  rows = NULL, cols = NULL,
                      theme = ttheme_default(
                        core = list(
                          fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                          bg_params = list(fill = c("grey97")),
                          padding = unit(c(1.3, 0.3), "line")
                        ),
                        colhead = list(
                          fg_params = list(fontsize = 10, fontface = "plain", rot = 45, hjust = 0.5, vjust = 1, lineheight = 0.8),
                          bg_params = list(fill = c("white")),
                          padding = unit(c(1.3, 0.3), "line")
                        ),
                        rowhead = list(
                          fg_params = list(fontsize = 10, fontface = "plain"),
                          padding = unit(c(1.3, 0.3), "line")
                        )
                      ))

df.subsetlist = data.table(
  subset = names(col_celltype),
  color = col_celltype,
  y = 1,
  x = 1:length(col_celltype)
)
p.subsetlist = ggplot(df.subsetlist, aes(x = x, y = y, label = subset, color = color)) +
  geom_text(angle = -40, size = 2.6, hjust = 0) + 
  scale_color_identity() +           
  theme_void() +
  coord_cartesian(clip = "off") 

df.twocircles = data.table(
  x = rep(1,2),
  y = 1:2,
  color = c("grey17", "grey88"),
  labeltext = c("FDR ≤ 0.2", "FDR > 0.2")
)

p.twocircles = ggplot(df.twocircles, aes(x = x, y = y, label = labeltext, color = color)) +
  geom_point(shape = 21, fill = "grey88", size = 2) +
  geom_text(angle = 0, size = 2.6, hjust = -0.2, color = textcolor) + 
  theme(text = element_text(size = textsize, color = textcolor)) +
  scale_color_identity() +           
  theme_void() +
  coord_cartesian(clip = "off") 

infect.exprmeta.allneu = readRDS(paste0(dir, "infect.exprmeta.selectedgenes.allneu.neu100.v10f.rds"))
infect.exprmeta.allneu$rowname = gsub("\\(.*", "", infect.exprmeta.allneu$rowname)
infect.exprmeta.allneu.norowname = infect.exprmeta.allneu
infect.exprmeta.allneu.norowname$rowname = ""

cairo_pdf(file = paste0(dir.fig, "figure3.pdf"), width = 8, height = 10.2)
pushViewport(viewport(layout = grid.layout(nrow = 462, ncol = 63)))
print(p.infectdistribution + theme(plot.margin = margin(0.1, 0.1, 0.1, 0.1, "line")), vp = viewport(layout.pos.row = 5:53, layout.pos.col = 3:34))

print(p.subsetlist + theme(plot.margin = margin(0, 0, 0, 0, "line")), vp = viewport(layout.pos.row = 39:70, layout.pos.col = 2:22))

print(p.pids + theme(plot.margin = margin(0, 0, 0, 0, "line")), vp = viewport(layout.pos.row = 12:28, layout.pos.col = 35:58))

pushViewport(viewport(layout.pos.row = 24:52, layout.pos.col = 36:58))
grid.draw(dt.infect)
popViewport()

pushViewport(viewport(layout.pos.row = 64:300, layout.pos.col = 39:49))
draw(pht.granulesub, merge_legend = F, heatmap_legend_side = "left",newpage = FALSE)
upViewport(1)

pushViewport(viewport(layout.pos.row = 64:294, layout.pos.col = 50:60))
draw(pht.IFNsub, merge_legend = F, heatmap_legend_side = "bottom",newpage = FALSE)
upViewport(1)

print(p.es.summary.infect +  theme(plot.margin = margin(0.1, 0.2, 0.1, 0.6, "line")), vp = viewport(layout.pos.row = 74:300, layout.pos.col = 2:28))

pushViewport(viewport(layout.pos.row = 301:310, layout.pos.col = 10:50))
draw(Legend(title = NULL, title_gp = gpar(fontsize = 8, col = textcolor), title_position = "topcenter",
            labels = names(col_infect), nrow = 1, by_row = T,
            labels_gp = gpar(fontsize = 9, col = textcolor, fontface = "plain"),
            grid_height = unit(0.5, "line"), grid_width = unit(0.5, "line"),
            gap = unit(.2, "line"), legend_gp = gpar(alpha = 0.8, col = col_infect, fill = col_infect)))
upViewport(1)

print(p.twocircles +  theme(plot.margin = margin(0, 0, 0, 0, "line")), vp = viewport(layout.pos.row = 295:299, layout.pos.col = 50:51))

print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "S100A8" & group %in% "COVID-19,severe"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = " ", stripxface = "italic", labex = 2.1, summaryy = c(0,-0.4,0,0.4)) + scale_x_continuous(breaks = seq(-5,5,1)) +  theme(plot.margin = unit(c(0.4,0.1,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 320:390, layout.pos.col = 13:22))


print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "S100A9" & group %in% "COVID-19,severe"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = " ", stripxface = "italic", labex = 1.6, summaryy = c(0,-0.4,0,0.4)) + scale_x_continuous(breaks = seq(-5,5,1)) +  theme(plot.margin = unit(c(0.4,0.1,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 320:390, layout.pos.col = 23:32))


print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "OLR1" & group %in% "COVID-19,severe"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = " ", stripxface = "italic", labex = 1.05, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.1,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 320:390, layout.pos.col = 33:42))

print(
  metaPlots(infect.exprmeta.allneu[gene %in% "CD274" & group %in% "COVID-19,severe"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = " ", stripxface = "italic", labex = 1.05, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.3,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 320:390, layout.pos.col = 43:61))

print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "ARG1" & group %in% "COVID-19,severe"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = expression("COVID-19,severe" ~ italic("vs.") ~ "healthy/convalescent"), stripxface = "italic", labex = -1, summaryy = c(0,-0.4,0,0.4)) + scale_x_continuous(breaks = seq(-5,5,1)) +  theme(plot.margin = unit(c(0.24,0.1,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 320:390, layout.pos.col = 3:12))


print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "S100A8" & group %in% "bacterial sepsis"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = " ", stripxface = "italic", labex = 0.05, summaryy = c(0,-0.4,0,0.4)) + scale_x_continuous(breaks = seq(-5,5,1)) +  theme(plot.margin = unit(c(0.4,0.1,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 392:462, layout.pos.col = 13:22))

print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "S100A9" & group %in% "bacterial sepsis"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = " ", stripxface = "italic", labex = 2.6, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.1,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 392:462, layout.pos.col = 23:32))

print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "OLR1" & group %in% "bacterial sepsis"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = " ", stripxface = "italic", labex = -1.9, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.1,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 392:462, layout.pos.col = 33:42))

print(
  metaPlots(infect.exprmeta.allneu[gene %in% "CD274" & group %in% "bacterial sepsis"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = " ", stripxface = "italic", labex = 1.5, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.6,0.4,0.1), "line")),
  vp = viewport(layout.pos.row = 392:462, layout.pos.col = 43:60))

print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "ARG1" & group %in% "bacterial sepsis"], boxsize = 5, textsize = 10, mycol = "#FF6718",
            plottitle = expression("bacterial sepsis" ~ italic("vs.") ~ "healthy/convalescent"), stripxface = "italic", labex = -0.6, summaryy = c(0,-0.4,0,0.4)) + scale_x_continuous(breaks = seq(-5,5,1)) +  theme(plot.margin = unit(c(0.24,0.2,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 392:462, layout.pos.col = 3:12))

grid.text(x = unit(0.02, "npc"), y = unit(0.985, "npc"), label = "A", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.02, "npc"), y = unit(0.842, "npc"), label = "B", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.598, "npc"), y = unit(0.854, "npc"), label = "C", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.77, "npc"), y = unit(0.854, "npc"), label = "D", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.608, "npc"), y = unit(0.87, "npc"), label = "meta-analysis of gene expression", gp = gpar(fontsize = textsize, fontface = "plain"), just = "left")
grid.text(x = unit(0.02, "npc"), y = unit(0.282, "npc"), label = "E", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.02, "npc"), y = unit(0.126, "npc"), label = "F", gp = gpar(fontsize = textsize, fontface = "bold"))

grid.brackets(0.42, 0.817, 0.42, 0.771, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.text(x = unit(0.435, "npc"), y = unit(0.794, "npc"), just = "left", label = "granule\nsubsets", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

grid.brackets(0.42, 0.633 , 0.42, 0.587, lwd = 0.7, ticks = 0.5, h = 0.01, type = 4, curvature = 0.3, col = "grey42")
grid.text(x = unit(0.435, "npc"), y = unit(0.61, "npc"), just = "left", label = "interferon-\nrelated\nsubsets", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

dev.off()
```

```{r figureS3}
cairo_pdf(file = paste0(dir.fig, "figureS3.pdf"), width = 12.6, height = 16)
pushViewport(viewport(layout = grid.layout(nrow = 192, ncol = 93)))

print(p.infect.box, vp = viewport(layout.pos.row = 2:58, layout.pos.col = 2:88))

print(
  metaPlots(ESlist$`AZU1.COVID-19/flu,nonsevere.vs.healthy/convalescent`, boxsize = 5, mycol = "grey67",
            plottitle = "COVID-19/flu,nonsevere vs. healthy/convalescent", labex = 0.4, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 61:78, layout.pos.col = 2:30))
print(
  metaPlots(ESlist$`AZU1.COVID-19,severe.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19,severe vs. healthy/convalescent", labex = 1.1, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.01,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 79:97, layout.pos.col = 2:30))
print(
  metaPlots(ESlist$`AZU1.bacterial sepsis.vs.healthy/convalescent`, boxsize = 5, mycol = "grey67",
            plottitle = "bacterial sepsis vs. healthy/convalescent", labex = 1, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,1.1,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 98:116, layout.pos.col = 2:30))

print(
  metaPlots(ESlist$`LTF.COVID-19/flu,nonsevere.vs.healthy/convalescent`, boxsize = 5, mycol = "grey67",
            plottitle = "COVID-19/flu,nonsevere vs. healthy/convalescent", labex = 1, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 61:78, layout.pos.col = 31:59))
print(
  metaPlots(ESlist$`LTF.COVID-19,severe.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19,severe vs. healthy/convalescent", labex = 1.2, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.01,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 79:97, layout.pos.col = 31:59))
print(
  metaPlots(ESlist$`LTF.bacterial sepsis.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "bacterial sepsis vs. healthy/convalescent", labex = 1.4, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,1.1,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 98:116, layout.pos.col = 31:59))

print(
  metaPlots(ESlist$`MMP9.COVID-19/flu,nonsevere.vs.healthy/convalescent`, boxsize = 5, mycol = "grey67",
            plottitle = "COVID-19/flu,nonsevere vs. healthy/convalescent", labex = 0.6, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 61:78, layout.pos.col = 60:88))
print(
  metaPlots(ESlist$`MMP9.COVID-19,severe.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19,severe vs. healthy/convalescent", labex = 1.4, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.01,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 79:97, layout.pos.col = 60:88))
print(
  metaPlots(ESlist$`MMP9.bacterial sepsis.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "bacterial sepsis vs. healthy/convalescent", labex = 2, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,1.1,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 98:116, layout.pos.col = 60:88))

print(
  metaPlots(ESlist$`IFN1.COVID-19/flu,nonsevere.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19/flu,nonsevere vs. healthy/convalescent", labex = 1.4, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 119:136, layout.pos.col = 2:30))

print(
  metaPlots(ESlist$`IFN1.COVID-19,severe.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19,severe vs. healthy/convalescent", labex = 1.2, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.01,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 137:155, layout.pos.col = 2:30))
print(
  metaPlots(ESlist$`IFN1.bacterial sepsis.vs.healthy/convalescent`, boxsize = 5, mycol = "grey67",
            plottitle = "bacterial sepsis vs. healthy/convalescent", labex = 1.3, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,1.1,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 156:174, layout.pos.col = 2:30))

print(
  metaPlots(ESlist$`IFN2.COVID-19/flu,nonsevere.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19/flu,nonsevere vs. healthy/convalescent", labex = 1.5, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 119:136, layout.pos.col = 31:59))

print(
  metaPlots(ESlist$`IFN2.COVID-19,severe.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19,severe vs. healthy/convalescent", labex = 1.3, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.01,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 137:155, layout.pos.col = 31:59))
print(
  metaPlots(ESlist$`IFN2.bacterial sepsis.vs.healthy/convalescent`, boxsize = 5, mycol = "grey67",
            plottitle = "bacterial sepsis vs. healthy/convalescent", labex = -2, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,1.1,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 156:174, layout.pos.col = 31:59))

print(
  metaPlots(ESlist$`IFN3.COVID-19/flu,nonsevere.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19/flu,nonsevere vs. healthy/convalescent", labex = 2.5, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 119:136, layout.pos.col = 60:88))

print(
  metaPlots(ESlist$`IFN3.COVID-19,severe.vs.healthy/convalescent`, boxsize = 5, mycol = "#FF6718",
            plottitle = "COVID-19,severe vs. healthy/convalescent", labex = 1.15, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.01,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 137:155, layout.pos.col = 60:88))
print(
  metaPlots(ESlist$`IFN3.bacterial sepsis.vs.healthy/convalescent`, boxsize = 5, mycol = "grey67",
            plottitle = "bacterial sepsis vs. healthy/convalescent", labex = 0.6, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,1.1,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 156:174, layout.pos.col = 60:88))

print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "S100A8" & group %in% "COVID-19/flu,nonsevere"], boxsize = 5, textsize = 10, mycol = "grey67",
            plottitle = "", stripxface = "italic", labex = 1, summaryy = c(0,-0.4,0,0.4)) + scale_x_continuous(breaks = seq(-5,5,1)) +  theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 175:192, layout.pos.col = 18:33))

print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "S100A9" & group %in% "COVID-19/flu,nonsevere"], boxsize = 5, textsize = 10, mycol = "grey67",
            plottitle = " ", stripxface = "italic", labex = 2, summaryy = c(0,-0.4,0,0.4)) + scale_x_continuous(breaks = seq(-5,5,1)) +  theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 175:192, layout.pos.col = 34:49))
print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "OLR1" & group %in% "COVID-19/flu,nonsevere"], boxsize = 5, textsize = 10, mycol = "grey67",
            plottitle = " ", stripxface = "italic", labex = 1, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 175:192, layout.pos.col = 50:65))

print(
  metaPlots(infect.exprmeta.allneu[gene %in% "CD274" & group %in% "COVID-19/flu,nonsevere"], boxsize = 5, textsize = 10, mycol = "grey67",
            plottitle = " ", stripxface = "italic", labex = 1.2, summaryy = c(0,-0.4,0,0.4)) + theme(plot.margin = unit(c(0.4,0.6,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 175:192, layout.pos.col = 66:89))

print(
  metaPlots(infect.exprmeta.allneu.norowname[gene %in% "ARG1" & group %in% "COVID-19/flu,nonsevere"], boxsize = 5, textsize = 10, mycol = "grey67",
            plottitle = expression("COVID-19/flu,nonsevere" ~ italic("vs.") ~ "healthy/convalescent"), stripxface = "italic", labex = 0.2, summaryy = c(0,-0.4,0,0.4)) + scale_x_continuous(breaks = seq(-5,5,1)) +  theme(plot.margin = unit(c(0.4,0.3,0.4,0.4), "line")),
  vp = viewport(layout.pos.row = 175:192, layout.pos.col = 2:17))

grid.text(x = unit(0.01, "npc"), y = unit(0.99, "npc"), label = "A", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.01, "npc"), y = unit(0.69, "npc"), label = "B", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.01, "npc"), y = unit(0.385, "npc"), label = "C", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.01, "npc"), y = unit(0.095, "npc"), label = "D", gp = gpar(fontsize = textsize, fontface = "bold"))

dev.off()
```

# Figure 4&S4

```{r distribution&ES}
dtf.s.pct.cancertissue = dtf.s.pct[!grepl("peripheral blood",tissue)][!grepl("han2020",pid)]
dtf.s.pct.cancertissue = dtf.s.pct.cancertissue[grepl("cancer",group)]

samplemeta.cancer = fread(paste0(dir, "samplemta.cancer.v10f.csv"))

dtf.s.pct.cancertissue$stage = samplemeta.cancer[match(dtf.s.pct.cancertissue$sampleid.pid, samplemeta.cancer$sampleid.pid)]$stage
dtf.s.pct.cancertissue$type = samplemeta.cancer[match(dtf.s.pct.cancertissue$sampleid.pid, samplemeta.cancer$sampleid.pid)]$type
dtf.s.pct.cancertissue$treatment = samplemeta.cancer[match(dtf.s.pct.cancertissue$sampleid.pid, samplemeta.cancer$sampleid.pid)]$treatment

dtf.s.pct.cancertissue = dtf.s.pct.cancertissue[number.neutrophil >= 50]

dtf.s.pct.cancertissue[,condition:=ifelse(group.detail == "PDAC", "PDAC", group)]
dtf.s.pct.cancertissue[,condition:=ifelse(condition == "kidney cancer", "other cancer", condition)]
table(dtf.s.pct.cancertissue$condition)

dtf.s.pct.cancertissue$sex = samplemeta[match(dtf.s.pct.cancertissue$sampleid.pid, samplemeta$sampleid.pid)]$sex
dtf.s.pct.cancertissue$age = samplemeta[match(dtf.s.pct.cancertissue$sampleid.pid, samplemeta$sampleid.pid)]$age

dtf.s.pct.cancertissue.m = melt(dtf.s.pct.cancertissue,
                                id.vars = c("pid", "sampleid.pid", "condition", "type", "stage", "treatment"),
                                measure.vars = names(col_celltype))
dtf.s.pct.cancertissue.m = as.data.table(dtf.s.pct.cancertissue.m)
dtf.s.pct.cancertissue.m$condition = factor(dtf.s.pct.cancertissue.m$condition, levels = c("healthy", "lung cancer", "PDAC", "GI cancer", "other cancer"))

# bar ----
dtf.s.pct.m.agg = dtf.s.pct.cancertissue.m[pid %in% c("hu2022", "salcher2022", "wu2024", "hu2023", "chan2021")][,lapply(.SD, function(x){mean(x)}), by = c( "variable", "type", "treatment"), .SDcols = c("value")]

dtf.s.pct.m.agg$variable = factor(dtf.s.pct.m.agg$variable , levels = rev(names(col_celltype)))
dtf.s.pct.m.agg$group = paste0(dtf.s.pct.m.agg$type, ", ",dtf.s.pct.m.agg$treatment)

dtf.s.pct.m.agg = dtf.s.pct.m.agg[!grepl("other",group)]
dtf.s.pct.m.agg = dtf.s.pct.m.agg[!grepl("metastasis", group)]
dtf.s.pct.m.agg$group = factor(dtf.s.pct.m.agg$group, levels = rev(c("adjacent, pre-treatment", "cancer, pre-treatment", "cancer, post-treatment")))

dtf.s.pct.m.agg$facetvar = "neutrophil subset in cancer"
p.cancer.distribution = 
  ggplot(dtf.s.pct.m.agg, aes(x = group, y = value, fill = variable)) +
  geom_bar(stat = "identity", alpha = 0.8) +
  theme_bw() + mytheme +
  scale_y_continuous(expand = c(0, 0), breaks = seq(0, 1, 0.2), labels = seq(0, 1, 0.2), position = "left") + 
  scale_x_discrete(position = "top") + 
  facet_wrap(. ~ facetvar )+
  labs(x = NULL, y = NULL , title = NULL) +
  theme(plot.margin = unit(c(.1, .1, .1, .1), "line"),
        text = element_text(size = 9, color = textcolor),
        axis.text = element_text(size = 9, color = textcolor),
        axis.title = element_text(size = 9, color = textcolor),
        panel.grid.major = element_blank(), 
        panel.border = element_blank(),
        legend.title =  element_blank(),
        legend.text = element_text(color = textcolor, face = "italic"),
        legend.position = "none",
        legend.spacing.y = unit(0.01, "line"), 
        legend.spacing.x = unit(0.04, "line"),
        legend.key.spacing.x = unit(.1, "line"),
        legend.key.spacing.y = unit(.1, "line"),
        legend.key.width = unit(0.2, "line"),
        legend.key.height = unit(0.2, "line"),
        plot.title = element_text(size = 9, face = "plain", hjust = 0.5, margin = margin(t = 1, r = 0, b = 1, l = 0, unit = "pt"))) +
  scale_fill_manual(values = col_celltype, name = "") + coord_flip()  +
  guides(fill = guide_legend(ncol = 1, byrow = T, size = 2.3)) 

# effect size ----
dtf.s.pct.cancertissue.m$condition = dtf.s.pct.cancertissue.m$type
ESlist.cancer = list()

for(id in unique(dtf.s.pct.cancertissue.m$variable)){
  ESlist.cancer[[paste0(id, ".cancer.vs.adjacent")]] = metaStats(dtf.s.pct.cancertissue.m[treatment == "pre-treatment"][order(pid)], geneid = id, pids = c("hu2022", "salcher2022", "wu2024"), case = "cancer", control = "adjacent")
}

ESlist.cancer = lapply(ESlist.cancer, function(x){x$rowname = gsub(" \\(.*", "", x$rowname); return(x)})

datadf.cancer.adjacent = rbindlist(ESlist.cancer, idcol = 'group')
datadf.cancer.adjacent = as.data.table(datadf.cancer.adjacent)
datadf.cancer.adjacent.summary = datadf.cancer.adjacent[rowname == "summary"]
datadf.cancer.adjacent.summary$FDR = p.adjust(datadf.cancer.adjacent.summary$pval, method = "fdr")

datadf.cancer.adjacent.summary$celltype = factor(datadf.cancer.adjacent.summary$gene, levels = names(col_celltype))
datadf.cancer.adjacent.summary[,sig:=ifelse(FDR <= 0.2, "FDR ≤ 0.2", "FDR > 0.2")]

nth <- function(x, i) {
  x[(i - 1)%%length(x) + 1]
}
ci.value <- -qnorm((1 - 0.95)/2)

datadf.cancer.adjacent.summary$lower <- datadf.cancer.adjacent.summary$ES.g - ci.value * datadf.cancer.adjacent.summary$ES.se
datadf.cancer.adjacent.summary$upper <- datadf.cancer.adjacent.summary$ES.g + ci.value * datadf.cancer.adjacent.summary$ES.se

p.es.summary.cancer.vs.adjacent = 
  ggplot(data = datadf.cancer.adjacent.summary, aes(x = gene, y = ES.g, alpha = sig)) +
  geom_hline(yintercept = 0, size = 0.3, alpha = 1, color = "grey67") +
  geom_tile(width = 0.7, height = 0.1, position = position_dodge(0.7), color = NA) +
  geom_errorbar(data = datadf.cancer.adjacent.summary, aes(x = celltype, ymin = lower, ymax = upper,  group = group), 
                position = position_dodge(0.7), color = "grey7", width = .7, size = 0.2) +
  labs(x = NULL, title = "meta-analysis of proportion\neffect size", y = expression("cancer" ~ italic("vs.") ~ "adjacent, pre-treatment")) +
  scale_y_continuous(breaks = seq(-10, 2, 1)) +  
  scale_alpha_manual(values = c("FDR > 0.2" = 0.2, "FDR ≤ 0.2" = 1)) +
  theme_bw() + mytheme + 
  theme(plot.title = element_text(face = 1, margin = margin(0, 0, 0, 0, "line"), size = 9), 
        text = element_text(size = 9, color = textcolor),
        axis.text = element_text(size = 9, color = textcolor),
        axis.title = element_text(size = 9, color = textcolor),
        panel.grid.major.y = element_blank(),
        axis.text.y = element_text(size = 9, face = "plain", color = rev(col_celltype)),
        legend.position = c(0.04, 0.02),
        legend.justification = c(0, 0),
        legend.title = element_blank(),
        legend.background =  element_blank(),
        legend.box.background = element_blank(),
        legend.spacing.y = unit(0.01, "line"),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.3, "line")) + 
  coord_flip() +
  scale_x_discrete(limits = rev, position = "bottom")  +
  guides(color = "none", shape = "none", size = "none")
```

```{r treatment}
table(dtf.s.pct.cancertissue[grepl("cancer|adjacent",type)]$pid, dtf.s.pct.cancertissue[grepl("cancer|adjacent",type)]$treatment)
table(dtf.s.pct.cancertissue$pid, dtf.s.pct.cancertissue$treatment)

dtf.s.pct.cancertissue.m$condition = dtf.s.pct.cancertissue.m$treatment

ESlist.treatment = list()
for(id in unique(dtf.s.pct.cancertissue.m$variable)){
  ESlist.treatment[[paste0(id, ".treatment")]] = metaStats(dtf.s.pct.cancertissue.m[grepl("cancer|adjacent",type)][order(pid)], geneid = id, pids = c("chan2021", "hu2023", "maynard2020"), case = "post-treatment", control = "pre-treatment")
}

datadf.treatment = rbindlist(ESlist.treatment, idcol = 'group')
datadf.treatment = as.data.table(datadf.treatment)
datadf.treatment.summary = datadf.treatment[rowname == "summary"]
datadf.treatment.summary$FDR = p.adjust(datadf.treatment.summary$pval, method = "fdr")
datadf.treatment.summary$celltype = factor(datadf.treatment.summary$gene, levels = names(col_celltype))
datadf.treatment.summary[,sig:=ifelse(FDR <= 0.2, "FDR ≤ 0.2", "FDR > 0.2")]

nth <- function(x, i) {
  x[(i - 1)%%length(x) + 1]
}
ci.value <- -qnorm((1 - 0.95)/2)

datadf.treatment.summary$lower = datadf.treatment.summary$ES.g - ci.value * datadf.treatment.summary$ES.se
datadf.treatment.summary$upper = datadf.treatment.summary$ES.g + ci.value * datadf.treatment.summary$ES.se

p.es.summary.treatment = 
  ggplot(data = datadf.treatment.summary, aes(x = celltype, y = ES.g, alpha = sig)) +
  geom_hline(yintercept = 0, size = 0.3, alpha = 1, color = "grey67") +
  geom_tile(width = 0.7, height = 0.1, position = position_dodge(0.7), color = NA) +
  geom_errorbar(data = datadf.treatment.summary, aes(x = celltype, ymin = lower, ymax = upper,  group = group), 
                position = position_dodge(0.7), color = "grey7", width = .7, size = 0.2) +
  labs(x = NULL, y = "meta-analysis of proportion\neffect size", title = expression("post" ~ italic("vs.") ~ "pre-treatment cancer")) +
  scale_y_continuous(breaks = seq(-10, 2, 1)) +  
  scale_alpha_manual(values = c("FDR > 0.2" = 0.2, "FDR ≤ 0.2" = 1)) +
  theme_bw() + mytheme + 
  theme(plot.title = element_text(face = 1, margin = margin(0, 0, 0, 0, "line")), 
        panel.grid.major.y = element_blank(),
        text = element_text(size = 9, color = textcolor),
        plot.margin = unit(c(0, 0, 0, 0), "line"),
        axis.text.y = element_text(size = 9, face = "plain"),
        legend.position = c(0.04, 0.02),
        legend.justification = c(0, 0),
        legend.title = element_blank(),
        legend.background =  element_blank(),
        legend.box.background = element_blank(),
        legend.spacing.y = unit(0.01, "line"),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.3, "line")) + 
  coord_flip() +
  scale_x_discrete(limits = rev, position = "top")  +
  guides(color = "none", shape = "none", size = "none")

datadf.cancer.es = data.table(
  celltype = datadf.cancer.adjacent.summary$celltype,
  "cancer.adjacent" = datadf.cancer.adjacent.summary$ES.g, "post.vs.pre" = datadf.treatment.summary$ES.g)

p.cancer.es = ggplot(datadf.cancer.es, aes(x = cancer.adjacent, y =  post.vs.pre)) +
  geom_point(shape = 16, color = "grey7", alpha = 0.6) +
  geom_text_repel(aes(label = celltype), size = 1.8, color = textcolor, alpha=1, force = 1.5, max.overlaps = 40, force_pull = 1.5, box.padding = 0.02) +
  labs(y = expression("post-" ~ italic("vs.") ~ "pre-treatment cancer"), x = expression("cancer" ~ italic("vs.") ~ "adjacent, pre-treatment"), title = "meta-analysis of proportion\neffect size") +
  geom_vline(xintercept = 0, size = 0.1, color = "grey67") +
  geom_hline(yintercept = 0, size = 0.1, color = "grey67") +
  theme_bw() + mytheme + 
  theme(text = element_text(size = 9, color = textcolor),
        axis.text = element_text(size = 9, color = textcolor),
        axis.title = element_text(size = 9, color = textcolor),
        plot.title = element_text(face = 1, size = 9, margin = margin(0, 0, 0, 0, "line")), 
        plot.margin = unit(c(0, 0, 0, 0), "line"),
        legend.position = c(0.04,0.02),
        legend.justification = c(0,0),
        legend.title = element_blank(),
        legend.background =  element_blank(),
        legend.box.background = element_blank(),
        legend.spacing.y = unit(0.01, "line"),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.3, "line")) +
  stat_cor(color = textcolor, method = "spearman", size = 2.4, alpha = 1, label.y.npc = 0.1) 
```

```{r T&NK}
dtf.tnk = readRDS(paste0(dir, "TNK.umapdf.harmony.pc20.v1b.anno.rds"))
samplemeta.cancer = fread(paste0(dir, "samplemta.cancer.v10f.csv"))

dtf.neu = dtf
dtf.neu$cluster = as.character(dtf.neu$celltype)
dtf.neu[,cluster:=ifelse(grepl("IFN", cluster), "IFN", cluster)]
dtf.neu.s.N = as.data.frame(dcast(dtf.neu[!celltype %in% c("doublet")][,.N, by = c("sampleid.pid", "cluster")], sampleid.pid ~ cluster, value.var = "N"))
rownames(dtf.neu.s.N) = dtf.neu.s.N$sampleid.pid
dtf.neu.s.N$sampleid.pid = NULL
dtf.neu.s.N[is.na(dtf.neu.s.N)] = 0
dtf.neu.s.number = rowSums(dtf.neu.s.N)
dtf.neu.s.pct = apply(dtf.neu.s.N,2,function(x){x/rowSums(dtf.neu.s.N)})
dtf.neu.s.pct = as.data.frame(dtf.neu.s.pct)
dtf.neu.s.pct$sampleid.pid = rownames(dtf.neu.s.pct)
dtf.neu.s.pct$neu.number = dtf.neu.s.number[dtf.neu.s.pct$sampleid.pid]
dtf.neu.s.pct = as.data.table(dtf.neu.s.pct)
dtf.neu.s.pct[, number.allcell:=ifelse(sampleid.pid %in% names(celltotalnumber), celltotalnumber[sampleid.pid], NA)]
dtf.neu.s.pct$neu.pct = dtf.neu.s.pct$neu.number / dtf.neu.s.pct$number.allcell

dtf.tnk$cluster = dtf.tnk$celltype
dtf.tnk.s.N = as.data.frame(dcast(dtf.tnk[!celltype %in% c("doublet")][,.N, by = c("sampleid.pid", "cluster")], sampleid.pid ~ cluster, value.var = "N"))
rownames(dtf.tnk.s.N) = dtf.tnk.s.N$sampleid.pid
dtf.tnk.s.N$sampleid.pid = NULL
dtf.tnk.s.N[is.na(dtf.tnk.s.N)] = 0
dim(dtf.tnk.s.N) 
dtf.tnk.s.number = rowSums(dtf.tnk.s.N)
dtf.tnk.s.pct = apply(dtf.tnk.s.N,2,function(x){x/rowSums(dtf.tnk.s.N)})
dtf.tnk.s.pct = as.data.frame(dtf.tnk.s.pct)
dtf.tnk.s.pct$sampleid.pid = rownames(dtf.tnk.s.pct)
dtf.tnk.s.pct$tnk.number = dtf.tnk.s.number[dtf.tnk.s.pct$sampleid.pid]
dtf.tnk.s.pct = as.data.table(dtf.tnk.s.pct)
dtf.tnk.s.pct[, number.allcell:=ifelse(sampleid.pid %in% names(celltotalnumber), celltotalnumber[sampleid.pid], NA)]
dtf.tnk.s.pct$tnk.pct = dtf.tnk.s.pct$tnk.number / dtf.tnk.s.pct$number.allcell
dtf.tnk.s.pct$number.allcell = NULL

colnames(dtf.neu.s.N) = paste0(colnames(dtf.neu.s.N), ".N")
colnames(dtf.tnk.s.N) = paste0(colnames(dtf.tnk.s.N), ".N")

dtf.neu.s.N$sampleid.pid = rownames(dtf.neu.s.N)
dtf.tnk.s.N$sampleid.pid = rownames(dtf.tnk.s.N)

dtf.s.pct.neu.tnk = Reduce(function(x, y) merge(x, y, by = c("sampleid.pid"), all = F), 
                           list(dtf.neu.s.N, dtf.tnk.s.N, 
                                dtf.neu.s.pct, dtf.tnk.s.pct))

dtf.s.pct.neu.tnk$pid = samplemeta[match(dtf.s.pct.neu.tnk$sampleid.pid, samplemeta$sampleid.pid)]$pid
dtf.s.pct.neu.tnk$group = samplemeta[match(dtf.s.pct.neu.tnk$sampleid.pid, samplemeta$sampleid.pid)]$group
dtf.s.pct.neu.tnk$tissue = samplemeta[match(dtf.s.pct.neu.tnk$sampleid.pid, samplemeta$sampleid.pid)]$tissue
dtf.s.pct.neu.tnk$tissue.detail = samplemeta[match(dtf.s.pct.neu.tnk$sampleid.pid, samplemeta$sampleid.pid)]$tissue.detail

dtf.s.pct.neu.tnk = as.data.table(dtf.s.pct.neu.tnk)

df.ntm.cancer = dtf.s.pct.neu.tnk[!grepl("peripheral blood", tissue)][!grepl("han2020",sampleid.pid)][grepl("cancer", group)]

df.ntm.cancer = df.ntm.cancer[
  tnk.pct >= 0.05 & tnk.number >= 50 
  & neu.pct >= 0.05  & neu.number >= 50]

df.ntm.cancer$type = samplemeta.cancer[match(df.ntm.cancer$sampleid.pid, samplemeta.cancer$sampleid.pid)]$type
df.ntm.cancer$treatment = samplemeta.cancer[match(df.ntm.cancer$sampleid.pid, df.ntm.cancer$sampleid.pid)]$treatment

NTM.Cancer = function(celltype1, celltype2){
  df.ntm.cancer$celltype1 = df.ntm.cancer[,celltype1, with = F]
  df.ntm.cancer$celltype2 = df.ntm.cancer[,celltype2, with = F]
  p =
    ggplot(df.ntm.cancer[group %in% c("lung cancer") & type %in% c("cancer", "adjacent")],
           aes(x = celltype1, y = celltype2)) +
    labs(y = paste0(celltype2, " subset"), x = paste0(celltype1)) +
    geom_point(aes(shape = type), size = 0.5, color = "grey17", alpha = 0.8) + mytheme +
    scale_shape_manual(values = c("cancer" = 17, "adjacent" = 16)) +
    
    scale_y_continuous(breaks = seq(0, 1, 0.1)) +
    theme(legend.position = "none",
          legend.justification = c(0, 0),
          panel.background = element_rect(color = backgroundcol,fill= backgroundcol),
          text = element_text(size = textsize),
          plot.title = element_text(face = "plain"),
          legend.background = element_blank(),
          legend.key = element_blank(),
          legend.title = element_blank(),
          legend.spacing.y = unit(0.04, "line"), 
          legend.spacing.x = unit(0.04, "line"),
          legend.key.width = unit(0.2, "line"),
          legend.key.height = unit(0.2, "line"),
          legend.text = element_markdown(size = 10),
          plot.margin = margin(0.02,0.05,0.02,0.05, "line"),
          axis.title = element_text(size = 10, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
          axis.text =  element_text(size = 8, color = textcolor, margin = margin(0, 0, 0, 0, "line"))) +
    stat_cor(color = textcolor, method = "spearman", size = 2.6, alpha = 0.8) +
    geom_smooth(method ='lm', alpha =0.3, size = 0, span = 0.5, linetype = 0, color = "grey79", fill = "grey79") +
    stat_smooth(method ='lm', geom = "line", alpha = 0.3, linewidth = 0.3, size =1, span=0.5,  color = "#B71B1BFF") 
  return(p)
}
```

```{r expression}
cancer.exprmeta = readRDS(paste0(dir, "cancer.exprmeta.selectedgenes.allneu.v10f.rds"))
cancer.exprmeta[gene %in% c("CD274", "OLR1")]
cancer.exprmeta = cancer.exprmeta[rowname %in% "summary"]

dtfgenes.mean = readRDS(paste0(dir, "dtfgenes.allneu.mean.v10f.rds"))
dtfgenes.mean.mean.cancer = dtfgenes.mean[sampleid.pid %in% dtf.s.pct.cancertissue$sampleid.pid][, lapply(.SD, geomMean), .SDcols = setdiff(colnames(dtfgenes.mean), c("sampleid.pid", "pid", "group", "tissue", "source", "sex", "age"))]
dtfgenes.mean.mean.cancer.t = data.table(gene = setdiff(colnames(dtfgenes.mean), c("sampleid.pid", "pid", "group", "tissue", "source", "sex", "age")), expr.geom =  unlist(dtfgenes.mean.mean.cancer))
summary(dtfgenes.mean.mean.cancer.t$expr.geom)
medianvar = median(dtfgenes.mean.mean.cancer.t$expr.geom)
dtfgenes.mean.mean.cancer.t[expr.geom >= medianvar][,.N]
cancer.exprmeta = cancer.exprmeta[gene %in% neu.cluster.markers2]

cancer.exprmeta$FDR = p.adjust(cancer.exprmeta$pval, method = "fdr")
cancer.exprmeta[, label:=ifelse((FDR <= 0.005 & abs(ES.g) >= 1) | (FDR <= 0.05 & (ES.g < -1.9 | ES.g > 1))| (FDR <= 0.05 & gene %in% neu.cluster.markers2) , gene, NA)]
cancer.exprmeta[, sig:=ifelse(FDR <= 0.05, "FDR≤0.05", "FDR>0.05")]

p.cancer.expr.meta = ggplot(cancer.exprmeta, aes(x = ES.g, y = -log10(pval))) +
  labs(x = expression("cancer" ~ italic("vs.") ~ "adjacent, effect size"), title = "meta-analysis of expression") +
  geom_point(size = .2, shape = 16, alpha = 0.8, aes(color = sig)) +
  geom_vline(xintercept = 0, size = 0.1, color = "grey47", alpha = 0.6) +
  geom_text_repel(aes_string(label = "label"),  size = 1, alpha = 1, 
                  force = 1.5, max.overlaps = 20, force_pull = 1, box.padding = 0, 
                  fontface = "italic", segment.size = 0.03, min.segment.length = 0.1, show.legend = FALSE) +
  mytheme +
  theme(text = element_text(size = 9, color = textcolor),
        axis.text = element_text(size = 9, color = textcolor),
        axis.title = element_text(size = 9, color = textcolor),
        plot.title = element_text(face = 1, size = 9, margin = margin(0, 0, 0, 0, "line")), 
        legend.position = c(0.9,1.05),
        legend.justification = c(1,1),
        panel.background = element_blank(),
        legend.background = element_blank(),
        legend.key = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 6),
        legend.spacing.y = unit(0.01, "line"), 
        legend.spacing.x = unit(0.01, "line"),
        legend.key.spacing.x = unit(.01, "line"),
        legend.key.spacing.y = unit(.01, "line"),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.1, "line"),
        legend.box.margin = margin(0, 0, 0, 0, "line")) + 
  scale_color_manual(values = c("FDR≤0.05" = "indianred", "FDR>0.05" = "grey77")) +
  guides(color = guide_legend(override.aes = list(size = 0.8, alpha = 0.8)), shape = "none", alpha = "none")
```

```{r figure4}
df.table = data.table(
  pid = c("hu2022", "salcher2022", "wu2024", "hu2023", "chan2021"),
  adjacent = c(3,16,27,0,0),
  cancer.pre = c(7,13,53,3,3),
  cancer.post = c(0,0,0,11,4)
)

t1 = tableGrob(t(df.table), rows = NULL,
               theme = ttheme_default(
                 core = list(
                   fg_params = list(fontsize = 9, fontface = "plain"),
                   bg_params = list(fill = c("grey95")),
                   padding = unit(c(0.3, 0.7), "line")
                 ),
                 colhead = list(
                   fg_params = list(fontsize = 9, fontface = "plain")
                 ),
                 rowhead = list(
                   fg_params = list(fontsize = 9, fontface = "plain")
                 )
               ))

df.shapes = data.table(
  y = rep(1, 2),
  x = 1:2,
  shape = c(16, 17),
  labeltext = c("adjacent", "cancer")
)

p.shapes = ggplot(df.shapes, aes(x = x, y = y, label = labeltext, shape = shape)) +
  geom_point(color = "grey17", fill = "grey17", size = 2) +
  geom_text(angle = 0, size = 2.6, hjust = -0.2,vjust = 0.5, color = textcolor) + 
  theme(text = element_text(size = textsize, color = textcolor)) +
  scale_shape_identity() +           
  theme_void() +
  coord_cartesian(clip = "off") 

cairo_pdf(file = paste0(dir.fig, "figure4.pdf"), width = 7.4, height = 9)

pushViewport(viewport(layout = grid.layout(nrow = 90, ncol = 71)))

print(p.cancer.distribution, vp = viewport(layout.pos.row = 4:14, layout.pos.col = 3:38))

pushViewport(viewport(layout.pos.row = 1:14, layout.pos.col = 39:70))
grid.draw(t1)
popViewport()

print(p.es.summary.cancer.vs.adjacent + theme(plot.margin = unit(c(0, 0, 0, 0), "line")), vp = viewport(layout.pos.row = 16:54, layout.pos.col = 2:28))

print(
  metaPlots(ESlist.cancer$VEGFA.cancer.vs.adjacent, boxsize = 5, mycol = "grey17",
            plottitle = NULL, labex = -0.7, textsize = 9) + scale_x_continuous(breaks = -3:3) + theme(plot.margin = unit(c(0.1, 0.3, 0.1, 0.1), "line")),
  vp = viewport(layout.pos.row = 16:30, layout.pos.col = 29:45))

print(
  metaPlots(ESlist.cancer$`CCL3/4.cancer.vs.adjacent`, boxsize = 5, mycol = "grey17",
            plottitle = NULL, labex = 1.35, textsize = 9) + theme(plot.margin = unit(c(0.1, 0.3, 0.1, 0.1), "line")),
  vp = viewport(layout.pos.row = 35:49, layout.pos.col = 29:45))

print(p.cancer.es, vp = viewport(layout.pos.row = 15:35, layout.pos.col = 46:69))
print(p.cancer.expr.meta + theme(plot.margin = unit(c(0.5, 0, 0, 0), "line")), vp = viewport(layout.pos.row = 36:54, layout.pos.col = 46:69))

print(NTM.Cancer("Treg", "VEGFA") + labs(y = NULL), vp = viewport(layout.pos.row = 58:72, layout.pos.col = 40:54))
print(NTM.Cancer("NK", "VEGFA") + labs(y = NULL), vp = viewport(layout.pos.row = 58:72, layout.pos.col = 56:70))
print(NTM.Cancer("CD4 T exhausted", "VEGFA") + labs(y = NULL), vp = viewport(layout.pos.row = 58:72, layout.pos.col = 24:38))
print(NTM.Cancer("CD8 T exhausted", "VEGFA"), vp = viewport(layout.pos.row = 58:72, layout.pos.col = 6:22))

print(NTM.Cancer("Treg", "CCL3/4") + labs(y = NULL), vp = viewport(layout.pos.row = 73:87, layout.pos.col = 40:54))
print(NTM.Cancer("NK", "CCL3/4") + labs(y = NULL), vp = viewport(layout.pos.row = 73:87, layout.pos.col = 56:70))
print(NTM.Cancer("CD4 T exhausted", "CCL3/4") + labs(y = NULL), vp = viewport(layout.pos.row = 73:87, layout.pos.col = 24:38))
print(NTM.Cancer("CD8 T exhausted", "CCL3/4"), vp = viewport(layout.pos.row = 73:87, layout.pos.col = 6:22))

grid.text(x = unit(0.5, "npc"), y = unit(0.38, "npc"), label = "proportion of T and NK subsets in total T and NK cells", gp = gpar(fontsize = textsize, fontface = "plain"))
grid.text(x = unit(0.04, "npc"), y = unit(0.2, "npc"), rot = 90, label = "proportion of neutrophil subsets\nin total neutrophils", gp = gpar(fontsize = textsize, fontface = "plain", lineheight = 0.8))

grid.text(x = unit(0.01, "npc"), y = unit(0.98, "npc"), label = "A", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.01, "npc"), y = unit(0.83, "npc"), label = "B", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.38, "npc"), y = unit(0.835, "npc"), label = "C", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.38, "npc"), y = unit(0.625, "npc"), label = "D", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.65, "npc"), y = unit(0.845, "npc"), label = "E", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.65, "npc"), y = unit(0.595, "npc"), label = "F", gp = gpar(fontsize = textsize, fontface = "bold"))
grid.text(x = unit(0.01, "npc"), y = unit(0.385, "npc"), label = "G", gp = gpar(fontsize = textsize, fontface = "bold"))
print(p.shapes, vp = viewport(layout.pos.row = 88:89, layout.pos.col = 33:38))
dev.off()
```

```{r figureS4}
dtf.tnk = readRDS(paste0(dir, "TNK.umapdf.harmony.pc20.v1b.anno.rds"))
dtf.tnk$tissue = samplemeta[match(dtf.tnk$sampleid.pid, samplemeta$sampleid.pid)]$tissue
dtf.tnk$group = samplemeta[match(dtf.tnk$sampleid.pid, samplemeta$sampleid.pid)]$group
dtf.tnk$pid = gsub("a$|b$|c$|d$","",dtf.tnk$pid)
dtf.tnk$pid = gsub("myin2024b","yin2024",dtf.tnk$pid)
dtf.tnk$pid = gsub("reyfman2018","reyfman2019",dtf.tnk$pid)
dtf.tnk$pid = gsub("xue2022","xue2023",dtf.tnk$pid)
dtf.tnk = dtf.tnk[!grepl("doublet",celltype)]
dtf.tnk[,tissue2:=ifelse(tissue %in% c("peripheral blood", "lung"), tissue, "other")]

pointsizen = 0.01

### pid 
dtf.tnk$facetvar = "study"
p.umap.pid =
  ScatterPlotRast(df = as.data.frame(dtf.tnk), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = pointsizen, alphalevel = 0.2, colorby = "pid", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed = 13, mycol = col_pid, backgroudcol = "white", plottitle = "", rasterdpi = 300, nacol = "grey42", facetlabelface = "plain") +
  labs(title = NULL)+ 
  theme(text = element_text(size = textsize),
        legend.text = element_text(size = 9, margin = margin(0,0,0,0.1, "line")),
        legend.position = c(1.01,0.5),
        legend.justification = c(0,0.5),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.3,"line"),
        legend.spacing.x = unit(.05, "line"),
        legend.spacing.y = unit(.05, "line"),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) + 
  guides(color = guide_legend(ncol = 2, label.hjust = 0, override.aes = list(size = 1.4, alpha = 1))) 

### group 
dtf.tnk$facetvar = "condition"
dtf.tnk$group = factor(dtf.tnk$group,levels = names(col_group))

p.umap.group = ScatterPlotRast(df = as.data.frame(dtf.tnk), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = pointsizen, alphalevel = 0.2, colorby = "group", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed =13, mycol = col_group, backgroudcol = "white", plottitle = "", rasterdpi = 300, nacol = "grey42", facetlabelface = "plain") +
  labs(title = NULL)+ 
  theme(text = element_text(size = textsize),
        legend.text = element_text(margin = margin(0,0,0,0.1, "line")),
        legend.position = c(1.01,0.5),
        legend.justification = c(0,0.5),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.7,"line"),
        legend.spacing.x = unit(.05, "line"),
        legend.spacing.y = unit(.05, "line"),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) + 
  guides(color = guide_legend(ncol = 2, label.hjust = 0, override.aes = list(size=2, alpha = 1))) 

### tissue 
col_tissue2 =c(
  "peripheral blood" = "#73DAFF",
  "lung"= "#DC3023",  
  "other" = "grey77"
)
length(unique(dtf.tnk$tissue))
dtf.tnk$facetvar = "tissue"
dtf.tnk$tissue2 = factor(dtf.tnk$tissue2,levels = names(col_tissue2))

p.umap.tissue = ScatterPlotRast(df = as.data.frame(dtf.tnk), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = pointsizen, alphalevel = 0.2, colorby = "tissue2", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed =13, mycol = col_tissue2, backgroudcol = "white", plottitle = "", rasterdpi = 300, nacol = "grey42", facetlabelface = "plain") +
  labs(title = NULL)+ 
  theme(text = element_text(size = textsize),
        legend.text = element_text(margin = margin(0,0,0,0.1, "line")),
        legend.position = c(1.01,0.5),
        legend.justification = c(0,0.5),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.7,"line"),
        legend.spacing.x = unit(.05, "line"),
        legend.spacing.y = unit(.05, "line"),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) + 
  guides(color = guide_legend(ncol=1, label.hjust = 0, override.aes = list(size=2, alpha = 1))) 

### celltype 
col_celltype_TNK = c(`CD4 T Eff/Mem` = "#FFA631", `CD4 T exhausted` = "#F9906F", 
                     `CD4 T IFN` = "#214DC8", `CD4 T Naive` = "#55BB8A", `CD8 T cytotoxic` = "#06A5C7", 
                     `CD8 T exhausted` = "#DC3023", `CD8 T Naive` = "#8CC269", NK = "#B0A4E3", 
                     NKT = "#EAFF56", `other T` = "grey87", `Prolif T/NK` = "#F8D626", 
                     `T GZMK` = "#FFF143", `T/NK CCL3/4` = "#acf300", Treg = "#E4C6D0", 
                     `γδ T` = "#F3D3E7")

df.label = as.data.table(dtf.tnk)[,lapply(.SD,median),by = celltype,.SDcols=c("UMAP_1","UMAP_2")]
colnames(df.label)[2:3] = c("x","y")
df.label$labeltext = df.label[,"celltype",with=F]
df.label[, y:=ifelse(celltype=="CD8 T exhausted", y +1, y)]
df.label[, y:=ifelse(celltype=="T/NK CCL3/4", y +0.2, y)]

dtf.tnk$facetvar = "cell type and state"
p.umap.celltype = 
  ScatterPlotRast(df = as.data.frame(dtf.tnk), featureX = "UMAP_1", featureY = "UMAP_2", shuffle = T, pointsize = pointsizen, alphalevel = 0.2, colorby = "celltype", groupby = "facetvar", legendColumn = 4, colcateg="discrete", numberofrows = 5, legendposition = "none", seed =13, mycol = col_celltype_TNK, backgroudcol = "white", plottitle = "", rasterdpi = 300, nacol = "grey42", labeldf = df.label, facetlabelface = "plain", labelsize = 1.3) +
  labs(title = NULL)+ 
  theme(text = element_text(size = textsize),
        legend.text = element_text(margin = margin(0,0,0,0.1, "line")),
        legend.position = c(1.04,0.5),
        legend.justification = c(0,0.5),
        legend.key.width = unit(0.2,"line"),
        legend.key.height = unit(0.7,"line"),
        legend.spacing.x = unit(.05, "line"),
        legend.spacing.y = unit(.05, "line"),
        plot.title = element_text(size = textsize, color = textcolor, face = "plain", hjust = 0.5)) + 
  guides(color = guide_legend(ncol=1, label.hjust = 0, override.aes = list(size=2, alpha = 1))) 

### plots 
cairo_pdf(file = paste0(dir.fig, "figureS4.pdf"),width = 9,height = 5)
pushViewport(viewport(layout=grid.layout(nrow = 32, ncol = 65)))

print(p.umap.pid,vp=viewport(layout.pos.row = 2:16, layout.pos.col = 2:16))
print(p.umap.tissue,vp=viewport(layout.pos.row = 2:16, layout.pos.col = 38:52))

print(p.umap.group,vp=viewport(layout.pos.row = 17:31, layout.pos.col = 2:16))
print(p.umap.celltype,vp=viewport(layout.pos.row = 17:31, layout.pos.col = 38:52))
dev.off()
```

# Figure 5&S5

```{r bar&box}
dtf.s.pct = readRDS(paste0(dir, "dtf.s.pct.v10f.rds"))

dtf.s.pct.sex = dtf.s.pct[group == "healthy"][tissue %in% c("peripheral blood")][grepl("WB",source) | grepl("wigerblad2022|gupta2020|montaldo2022",pid)]
dtf.s.pct.sex[,.N]
dtf.s.pct.sex$sex = samplemeta[match(dtf.s.pct.sex$sampleid.pid, samplemeta$sampleid.pid)]$sex
dtf.s.pct.sex$age = samplemeta[match(dtf.s.pct.sex$sampleid.pid, samplemeta$sampleid.pid)]$age
dtf.s.pct.sex = dtf.s.pct.sex[!sex %in% c("unknown")]
dtf.s.pct.sex = dtf.s.pct.sex[neu.pct >= 0.05 | is.na(neu.pct)][number.neutrophil >= 100]

table(dtf.s.pct.sex$sex, dtf.s.pct.sex$pid)

dtf.s.pct.sex.m = melt(dtf.s.pct.sex,
                       id.vars = c("pid", "sampleid.pid", "sex"),
                       measure.vars = names(col_celltype))
dtf.s.pct.sex.m = as.data.table(dtf.s.pct.sex.m)

# number of datasets
length(unique(dtf.s.pct.sex.m$pid)) # 9
# number of samples
length(unique(dtf.s.pct.sex.m$sampleid.pid)) # 53
# number of samples from healthy males
length(unique(dtf.s.pct.sex.m[sex %in% "M"]$sampleid.pid)) # 29
# number of samples from healthy females
length(unique(dtf.s.pct.sex.m[sex %in% "F"]$sampleid.pid)) # 24
# number of cells from healthy males
dtf[sampleid.pid %in% unique(dtf.s.pct.sex.m[sex %in% "M"]$sampleid.pid)][,.N] # 92798
# number of cells from healthy females
dtf[sampleid.pid %in% unique(dtf.s.pct.sex.m[sex %in% "F"]$sampleid.pid)][,.N] # 103063

# box ----
p.box.prop.main = ggplot(dtf.s.pct.sex.m[variable %in% c("IFN2", "IFN3")], aes(x = sex, y = value)) +
  ggbeeswarm::geom_quasirandom(size = .1, stroke = 0.2, aes(color = sex), width = 0.2, show.legend = T, alpha = 0.5, varwidth = F, shape = 16) +
  geom_boxplot(width = 0.3, size = 0.2, outlier.shape = NA, alpha = 0.8, fill = NA, aes(color = sex)) +
  labs(y = "proportion", x = NULL, title = "") +
  scale_y_continuous(breaks = pretty_breaks()) +
  theme_bw() + mytheme + 
  theme(
    plot.margin = unit(c(.2,.2,.2,.2), "line"),
    legend.position = "none",
    panel.grid.major.x = element_blank(),
    plot.background = element_rect(color = "grey97",fill = "grey97"),
    plot.title = element_text(size = textsize, face = "plain",hjust = 0,margin = margin(t = .1, r = 0, b = .1, l = 0, unit = "pt"))) +
  facet_wrap(. ~ variable, nrow = 2, scales = "free") +
  scale_y_continuous(breaks = seq(0, 0.1, 0.05)) +
  scale_color_manual(values = c("F" = "#ffab65", "M" = "#65b9ff")) +
  stat_compare_means(comparisons = list(c("F", "M")), method = "wilcox.test", vjust = 1.5, size = 2, label = "p.format", color = "grey17", bracket.size = 0.1, tip.length = 0.01) +
  guides(colour = guide_legend(override.aes = list(size = 2)), shape = guide_legend(override.aes = list(size = 2))) 

p.box.prop = ggplot(dtf.s.pct.sex.m, aes(x = sex, y = value, color = pid)) +
  ggbeeswarm::geom_quasirandom(size =1, stroke = 0.2, width = 0.2, show.legend = T, alpha = 0.5, varwidth = F, shape = 16) +
  geom_boxplot(width = 0.3, size = 0.2, outlier.shape = NA, alpha = 0.8, fill = NA, color = "grey42") +
  labs(y = "proportion in total neutrophils", x = NULL, title = "") +
  scale_y_continuous(breaks = pretty_breaks()) +
  theme_bw() + mytheme + 
  theme(
    plot.margin = unit(c(.2,.2,.2,.2), "line"),
    panel.grid.major.x = element_blank(),
    plot.title = element_text(size = textsize, face = "plain",hjust = 0,margin = margin(t = .1, r = 0, b = .1, l = 0, unit = "pt"))) +
  facet_wrap(.~variable,  nrow = 3, scales = "free") +
  scale_y_continuous(breaks = pretty_breaks()) +
  scale_color_manual(values = col_pid) +
  stat_compare_means(comparisons = list(c("F", "M")), method = "wilcox.test", vjust = 1.5, size = 1.8, label = "p.format", color = "grey17", bracket.size = 0.1,tip.length = 0.01) +
  guides(colour = guide_legend(override.aes = list(size = 2)), shape = guide_legend(override.aes = list(size = 2))) 

# bar ----
dtf.s.pct.m.agg = dtf.s.pct.sex.m[,lapply(.SD, function(x){mean(x)}), by = c("sex",  "variable"), .SDcols = c("value")]
dtf.s.pct.m.agg[,lapply(.SD, function(x){sum(x)}), by = c("sex"), .SDcols = c("value")]

dtf.s.pct.m.agg$variable = factor(dtf.s.pct.m.agg$variable, levels = rev(names(col_celltype)))
table(dtf.s.pct.m.agg$variable)
dtf.s.pct.m.agg$sex = factor(dtf.s.pct.m.agg$sex, levels = c("M", "F"))
dtf.s.pct.m.agg$facetvar = "neutrophil subset proportion\nbetween sex (whole blood)"
p.sex.distribution = 
  ggplot(dtf.s.pct.m.agg, aes(x = sex, y = value, fill = variable)) +
  geom_bar(stat = "identity", alpha = 0.8) +
  theme_bw() + mytheme +
  scale_y_continuous(expand = c(0, 0),breaks = seq(0, 1, 0.2), labels = seq(0 , 1, 0.2), position = "right") + 
  scale_x_discrete(position = "top") + 
  labs(x = NULL, y = NULL , title = "neutrophil subset proportion\nbetween sex (whole blood)") +
  theme(plot.margin = unit(c(.1,.1,.1,.1), "line"),
        panel.grid.major = element_blank(), 
        panel.border = element_blank(),
        legend.title =  element_blank(),
        legend.text = element_text(color = textcolor, face = "italic"),
        legend.position = "none",
        legend.spacing.y = unit(0.01, "line"), 
        legend.spacing.x = unit(0.04, "line"),
        plot.background = element_rect(color = "grey97",fill = "grey97"),
        legend.key.width = unit(0.2, "line"),
        legend.key.height = unit(0.2, "line"),
        plot.title = element_text(size = textsize, face = "plain",hjust = 0.5, margin = margin(t = 1, r = 0, b = 1, l = 0, unit = "pt"))) +
  scale_fill_manual(values = col_celltype, name= "") + coord_flip()  +
  guides(fill = guide_legend(ncol = 1, byrow = T, size = 2.3)) 
```

```{r singlecell expression}
meta.sex.persample = readRDS(paste0(dir, "sex.exprmeta.selectedgenes.allneu.v10f.rds"))
meta.sex.persample = meta.sex.persample[rowname == "summary" & !is.na(pval)]

meta.sex.persample$FDR = p.adjust(meta.sex.persample$pval, method = "fdr")
meta.sex.persample[, sig:=ifelse(FDR <= 0.2, "FDR≤0.2", "FDR>0.2")]

meta.sex.persample[,pval:=ifelse(pval < 1e-08, 1e-08, pval)]

meta.sex.persample[,inset:=ifelse(gene %in% genes.immature.neu, "IMNeuS", 
                                  ifelse(gene %in% genes.total.neu, "TNeuS", 
                                         ifelse(gene %in% genes.ifn, "IFN genes", 
                                                "other")))]
meta.sex.persample[, label:=ifelse(pval < 0.08 & gene %in% signaturegenes | gene %in% c("CEACAM6") , gene, NA)]
meta.sex.persample$inset = factor(meta.sex.persample$inset, levels = names(col_set))

meta.sex.persample = rbind(
  meta.sex.persample[inset == "other"],
  meta.sex.persample[inset != "other"]
)

p.sex.sc.expr.meta = ggplot(meta.sex.persample, aes(x = ES.g, y = -log10(pval))) +
  labs(x = expression("male" ~ italic("vs.") ~ "female, effect size")) +
  geom_point(size = .2, shape = 16, alpha = 0.8, color = "grey42") +
  geom_vline(xintercept = 0, size = 0.1, color = "grey47", alpha = 0.6) +
  geom_text_repel(aes_string(label = "label"),  size = 1.3, alpha = 1, color = textcolor,
                  force = 1.5, max.overlaps = 40, force_pull = 1.5, box.padding = 0.02, 
                  fontface = "italic", segment.size = 0.03, min.segment.length = 0.3, show.legend = FALSE) +
  mytheme +
  theme(legend.position = c(0,1.05),
        legend.justification = c(0,1),
        panel.background = element_rect(color = backgroundcol,fill = backgroundcol),
        text = element_text(size = textsize),
        plot.title = element_text(face = "plain"),
        legend.background = element_blank(),
        legend.text = element_text(size = 6, color = textcolor),
        legend.key = element_blank(),
        legend.title = element_blank(),
        legend.spacing.y = unit(0.01, "line"), 
        legend.spacing.x = unit(0.01, "line"),
        legend.key.spacing.x = unit(.01, "line"),
        legend.key.spacing.y = unit(.01, "line"),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.1, "line"),
        legend.box.margin = margin(0, 0, 0, 0, "line"),
        plot.background = element_rect(color = "grey97",fill = "grey97"),
        axis.title =  element_text(size = 8, color = textcolor),
        axis.text =  element_text(size = 8, color = textcolor)) + 
  scale_color_manual(values = col_set2) +
  guides(color = guide_legend(override.aes = list(size = 0.8, alpha = 0.8)), shape = "none", alpha = "none")
```

```{r purified neutrophils (E-GEAD-397)}
res.ota = fread(paste0(dir, "ota2021.Neu_LDG.sex.DEG.csv"))
res.ota = res.ota[!is.na(padj)]

res.gsea = list()

generanks = res.ota$log2FoldChange
names(generanks) = res.ota$gene
set.seed(1)
res.gsea[["sex.ota"]] = fgseaMultilevel(pathways = geneSets, stats = generanks, nPermSimple = 10000)

# volcano ----
res.ota[,inset:=ifelse(gene %in% genes.immature.neu, "IMNeuS", 
                       ifelse(gene %in% genes.total.neu, "TNeuS", 
                              ifelse(gene %in% genes.ifn, "IFN genes", "other")))]

res.ota = rbind(
  res.ota[inset == "other"],
  res.ota[inset != "other"]
)

res.ota[,pvalue:=ifelse(pvalue < 1e-06, 1e-06, pvalue)]
res.ota[, label:=ifelse(inset!= "other", gene, NA)]
res.ota$inset = factor(res.ota$inset, levels = names(col_set))

res.ota[gene %in% genes.immature.neu & pvalue <= 0.05][,.N]
res.ota[gene %in% genes.total.neu & pvalue <= 0.05][,.N]
res.ota[gene %in% genes.ifn & pvalue <= 0.05][,.N]

p.otz = ggplot(res.ota[!is.na(padj)], aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(shape = 16, size = 0.5, alpha = 0.5, aes(color = inset)) +
  labs(x = "log2FC", title = expression("male" ~ italic("vs.") ~ "female (neutrophils)")) +
  geom_text_repel(aes_string(label = "label", color = "inset"), size = 1.3, alpha = 1, 
                  force = 1.5, max.overlaps = 40, force_pull = 1.5, box.padding = 0.02, 
                  fontface = "italic", segment.size = 0.03, min.segment.length = 0.3, show.legend = FALSE) +
  scale_color_manual(values = col_set) +
  mytheme + 
  theme(
    legend.position = c(0, 0.62),
    legend.justification = c(0, 0),
    panel.background = element_rect(color = backgroundcol,fill= backgroundcol),
    text = element_text(size = textsize),
    plot.title = element_text(face = "plain", margin = margin(0, 0, 0, 0, "line")),
    legend.background = element_blank(),
    legend.key = element_blank(),
    legend.title = element_blank(),
    legend.spacing.y = unit(0.01, "line"), 
    legend.spacing.x = unit(0.04, "line"),
    legend.key.width = unit(0.1, "line"),
    legend.key.height = unit(0.1, "line"),
    plot.background = element_rect(color = "grey97", fill = "grey97"),
    axis.title = element_text(size = 8, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
    axis.text =  element_text(size = 8, color = textcolor, margin = margin(0, 0, 0, 0, "line"))) +
  scale_y_continuous(limits = c(0, 6), breaks = seq(0, 6, 1)) +
  scale_x_continuous(limits = c(-2, 2), breaks = seq(-2, 2, 1)) +
  guides(color = guide_legend(override.aes = list(size = 1.5, alpha = 1)), shape = "none", alpha = "none")

# bar ----
res.ota = res.ota[order(-log2FoldChange)]
res.ota$rank = 1:nrow(res.ota)
res.ota[,inset:=ifelse(gene %in% genes.immature.neu, "IMNeuS", 
                       ifelse(gene %in% genes.total.neu, "TNeuS", 
                              ifelse(gene %in% genes.ifn, "IFN genes", "other")))]

res.ota = rbind(
  res.ota[inset == "other"],
  res.ota[inset != "other"]
)

p.sex.bar.otz = ggplot(res.ota, aes(x = -rank, y = log2FoldChange)) +
  geom_bar(stat = "identity", size = 0.002, width = 0.002, aes(color = inset, fill = inset), position = position_dodge(width = 0.002)) +
  labs(x = NULL, y = NULL, title = NULL) +
  theme_void() +
  theme(legend.position = "none") +
  scale_color_manual(values = c(col_set)) +
  scale_fill_manual(values = c(col_set)) 
```

```{r bulk transcriptome}
dtf.sexMeta = readRDS(paste0(dir,"dtf.sexMeta.rds"))
dtf.sexMeta[numStudies.PBMC > 2 & numStudies.WB > 1][,.N] 
dtf.sexMeta = dtf.sexMeta[numStudies.PBMC > 1 & numStudies.WB > 1][!chr %in% c("X", "Y") & !is.na(chr)]

maxES = 1.5

dtf.sexMeta[,effectSize.WB2:=ifelse(effectSize.WB > maxES, maxES, effectSize.WB)]
dtf.sexMeta[,effectSize.WB2:=ifelse(effectSize.WB2 < -maxES, -maxES, effectSize.WB2)]

dtf.sexMeta[,effectSize.PBMC2:=ifelse(effectSize.PBMC > maxES, maxES, effectSize.PBMC)]
dtf.sexMeta[,effectSize.PBMC2:=ifelse(effectSize.PBMC2 < -maxES, -maxES, effectSize.PBMC2)]

dtf.sexMeta$effectSizeFDR.WB = p.adjust(dtf.sexMeta$effectSizePval.WB, method = "fdr")
dtf.sexMeta$effectSizeFDR.PBMC = p.adjust(dtf.sexMeta$effectSizePval.PBMC, method = "fdr")

dtf.sexMeta[,sig:=ifelse(effectSizeFDR.WB < 0.1 & effectSizeFDR.PBMC < 0.1, "significant.both",
                         ifelse(effectSizeFDR.WB < 0.1 & effectSizeFDR.PBMC > 0.1, "significant.WB",
                                ifelse(effectSizeFDR.WB > 0.1 & effectSizeFDR.PBMC < 0.1, "significant.PBMC", "not significant")))]

dtf.sexMeta$effectSize.diff = dtf.sexMeta$effectSize.WB - dtf.sexMeta$effectSize.PBMC

dtf.sexMeta[,inset:=ifelse(gene %in% genes.immature.neu, "IMNeuS", 
                           ifelse(gene %in% genes.total.neu, "TNeuS",
                                  ifelse(gene %in% genes.ifn, "IFN genes", NA)))]

dtf.sexMeta[,inset2:=ifelse(gene %in% genes.immature.neu, "IMNeuS", 
                            ifelse(gene %in% genes.total.neu, "TNeuS",
                                   ifelse(gene %in% genes.ifn, "IFN genes", "other")))]

dtf.sexMeta[,inset3:=ifelse(!is.na(inset), "yes", "no")]
dtf.sexMeta[,genelabel:=ifelse(!is.na(inset), gene, NA)]

sigcol = c(
  "significant.both" = "#1A5E1F",
  "significant.WB" = "#F8D626",
  "significant.PBMC" = "#F47F17",
  "not significant" = "grey80"
)

sigsize = c(
  "significant.both" = 0.3,
  "significant.WB" = 0.3,
  "significant.PBMC" = 0.3,
  "not significant" = 0.01
)

sigshape = c(
  "significant.both" = 16,
  "significant.WB" = 15,
  "significant.PBMC" = 17,
  "not significant" = 1
)

dtf.sexMeta = rbind(
  dtf.sexMeta[sig == "not significant" & is.na(genelabel)],
  dtf.sexMeta[sig != "not significant" & is.na(genelabel)],
  dtf.sexMeta[sig == "not significant" & !is.na(genelabel)],
  dtf.sexMeta[sig != "not significant" & !is.na(genelabel)]
)

# scatter ----
v = prcomp(cbind(dtf.sexMeta$effectSize.WB2, dtf.sexMeta$effectSize.PBMC2))$rotation
beta = v[2,1]/v[1,1]

p.sex.PBMC.WB = ggplot(dtf.sexMeta,aes(x = effectSize.WB2, y = effectSize.PBMC2)) + mytheme +
  labs(x = "effect size, WB", y = "effect size, PBMC", title = expression("male" ~ italic("vs.") ~ "female (WB & PBMC)")) +
  geom_hline(yintercept = 0, size = 0.2, color = linecolor, alpha = 0.5) +
  geom_vline(xintercept = 0, size = 0.2, color = linecolor, alpha = 0.5) +
  geom_abline(intercept = 0 , slope = beta, color = linecolor, size = 0.2, alpha = 0.5) +
  geom_abline(intercept = 0 , slope = 1, color = linecolor, size = 0.2, alpha = 0.5) +
  geom_point(stroke = 0.2, aes_string(color = "inset2", shape = "sig", size = "sig", alpha = "inset3")) +
  geom_text_repel(aes_string(label = "genelabel", color = "inset2"), size = 1.3, alpha = 1, force = 1.5, max.overlaps = 40, force_pull = 1.5, box.padding = 0.02, fontface = "italic", segment.size = 0.03, min.segment.length = 0.3, show.legend = FALSE) +
  scale_x_continuous(breaks= pretty_breaks()) +
  scale_y_continuous(breaks= pretty_breaks()) +
  scale_color_manual(values = c(sigcol, "IMNeuS" = "#a93434", "TNeuS" = "#1A5E1F", "IFN genes" = "#2E42B8", "other" = "grey67"),
                     labels = c(names(sigcol),
                                "IMNeuS" = "<span style='color:#a93434;'>IMNeuS</span>",
                                "IFN genes" = "<span style='color:#2E42B8;'>IFN genes</span>")) +
  scale_shape_manual(values = sigshape) +
  scale_alpha_manual(values = c("yes" = 0.9, "no" = 0.3)) +
  scale_size_manual(values = sigsize) +
  theme(legend.position = c(0.45, 0),
        legend.justification = c(0, 0),
        panel.background = element_rect(color = backgroundcol, fill= backgroundcol),
        text = element_text(size = textsize),
        plot.title = element_text(face = "plain", margin = margin(0, 0, 0, 0, "line")),
        legend.background = element_blank(),
        legend.key = element_blank(),
        legend.title = element_blank(),
        legend.spacing.y = unit(0.04, "line"),
        legend.spacing.x = unit(0.04, "line"),
        legend.key.spacing.x = unit(.01, "line"),
        legend.key.spacing.y = unit(.01, "line"),
        legend.key.width = unit(0.1, "line"),
        legend.key.height = unit(0.1, "line"),
        legend.box = element_blank(),
        legend.text = element_text(size = 8),
        plot.background = element_rect(color = "grey97", fill = "grey97"),
        axis.title = element_text(size = 8, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
        axis.text =  element_text(size = 8, color = textcolor, margin = margin(0, 0, 0, 0, "line"))) +
  stat_cor(color = textcolor, method = "spearman", size = 2, alpha = 0.8) +
  guides(color = "none", shape = guide_legend(override.aes = list(size = 1.5)), alpha = "none", size = "none")

# bar ----
dtf.sexMeta = dtf.sexMeta[order(-effectSize.WB)]
dtf.sexMeta$rank.WB = 1:nrow(dtf.sexMeta)
dtf.sexMeta = dtf.sexMeta[order(-effectSize.PBMC)]
dtf.sexMeta$rank.PBMC = 1:nrow(dtf.sexMeta)
dtf.sexMeta[,inset:=ifelse(gene %in% genes.immature.neu, "IMNeuS", 
                           ifelse(gene %in% genes.total.neu, "TNeuS",
                                  ifelse(gene %in% genes.ifn, "IFN genes", "other")))]

dtf.sexMeta = rbind(
  dtf.sexMeta[inset == "other"],
  dtf.sexMeta[inset != "other"]
)

generanks = dtf.sexMeta$effectSize.WB
names(generanks) = dtf.sexMeta$gene
set.seed(1)
res.gsea[["sex.WB"]] = fgseaMultilevel(pathways = geneSets, stats = generanks, nPermSimple = 10000)

generanks = dtf.sexMeta$effectSize.PBMC
names(generanks) = dtf.sexMeta$gene
set.seed(1)
res.gsea[["sex.PBMC"]] = fgseaMultilevel(pathways = geneSets, stats = generanks, nPermSimple = 10000)

p.sex.WB = ggplot(dtf.sexMeta, aes(x = -rank.WB, y = effectSize.WB)) +
  geom_bar(stat = "identity", size = 0.002, width = 0.002, aes(color = inset, fill = inset), position = position_dodge(width = 0.002)) +
  labs(x = NULL, y = "effect size\nWB",  title = NULL) +
  theme_void() +
  theme(legend.position = "none") +
  scale_color_manual(values = c(col_set)) +
  scale_fill_manual(values = c(col_set)) 

p.sex.PBMC = ggplot(dtf.sexMeta, aes(x = -rank.PBMC, y = -effectSize.PBMC)) +
  geom_bar(stat = "identity", size = 0.002, width = 0.002, aes(color = inset, fill = inset), position = position_dodge(width = 0.002)) +
  labs(x = NULL, y = "effect size\nPBMC",  title = NULL) +
  theme_void() +
  theme(legend.position = "none") +
  scale_color_manual(values = c(col_set)) +
  scale_fill_manual(values = c(col_set)) +
  coord_flip()

# stats ----
genes2test = genes.immature.neu
N = length(unique(dtf.sexMeta$gene))  # Total number of unique genes in List B
K = nrow(dtf.sexMeta) * 0.01  # Number of top genes to consider in List B
M = length(genes2test)  # Number of genes in List A

#Calculate the overlap
top_K_genes = dtf.sexMeta[order(-effectSize.WB)]$gene[1:K]
k = length(intersect(genes2test, top_K_genes))
k
phyper(k - 1, K, N - K, M, lower.tail = FALSE)

top_K_genes = dtf.sexMeta[order(-effectSize.PBMC)]$gene[1:K]
k = length(intersect(genes2test, top_K_genes))
k
phyper(k - 1, K, N - K, M, lower.tail = FALSE)
```

```{r framingham}
framingham_score_pheno = fread(paste0(dir, "fh_age51-54conorm.score_pheno.csv"))
df.fh.hc = melt(framingham_score_pheno[healthy == 1], id.vars = c("age", "sex", "bmi", "regular_smokers"), measure.vars = c("IMNeuS", "TNeuS", "IFNS"))
df.fh.hc = as.data.table(df.fh.hc)
df.fh.hc[,bmi_categ:=ifelse(bmi>30, "high", "low")]
df.fh.hc$variable = factor(df.fh.hc$variable, levels = c("IMNeuS", "TNeuS", "IFNS"))

p.fh.hc.list.age = list()
for(score in c("IMNeuS", "TNeuS", "IFNS")){
  tmp = df.fh.hc[variable == score]
  tmp$facetvar = "age"
  p.fh.hc.list.age[[score]] = ggplot(tmp, aes(x = age, y = value, color = sex)) +
    geom_point(size = 0.1, alpha = 0.6, shape = 16) +
    geom_smooth(method= "loess", show.legend = F, size = 0.5) +
    facet_wrap(. ~ facetvar, ncol = 1, scales = "free") +
    theme_bw() + mytheme +
    labs(y = score, x = NULL, title = score) +
    theme(
      axis.text = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      axis.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.02, 0.02, 0.2, 0.02, "line")),
      strip.text.y = element_text(size = textsize, color = textcolor, margin = margin(0.1, 0.1, 0.1, 0.1, "line")),
      legend.position = c(-0.14, -0.22),
      legend.key.spacing.x = unit(.01, "line"),
      legend.key.spacing.y = unit(.01, "line"),
      legend.key.width = unit(0.1, "line"),
      legend.key.height = unit(0.1, "line"),
      legend.box = element_blank(),
      legend.background  = element_blank(),
      plot.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      plot.background = element_rect(color = "grey97", fill = "grey97")
    ) +
    scale_y_continuous(breaks = seq(0,10,1)) +
    scale_color_manual(values = c("F" = "#ffab65", "M" = "#65b9ff")) +
    guides(color = guide_legend(override.aes = list(size = 2), nrow = 1))
}

p.fh.hc.list.bmi = list()
for(score in c("IMNeuS", "TNeuS", "IFNS")){
  tmp = df.fh.hc[variable == score]
  tmp$facetvar = "BMI"
  p.fh.hc.list.bmi[[score]] = ggplot(tmp, aes(x = bmi, y = value, color = sex)) +
    geom_point(size = 0.1, alpha = 0.6, shape = 16) +
    geom_smooth(method= "loess", show.legend = F, size = 0.5) +
    facet_wrap(. ~ facetvar, ncol = 1, scales = "free") +
    theme_bw() + mytheme +
    labs(y = NULL, x = NULL, title = " ") +
    theme(
      axis.text = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      axis.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.02, 0.02, 0.2, 0.02, "line")),
      strip.text.y = element_text(size = textsize, color = textcolor, margin = margin(0.1, 0.1, 0.1, 0.1, "line")),
      legend.position = "none",
      legend.key.spacing.x = unit(.01, "line"),
      legend.key.spacing.y = unit(.01, "line"),
      legend.box = element_blank(),
      legend.background  = element_blank(),
      plot.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      plot.background = element_rect(color = "grey97", fill = "grey97")
    ) +
    scale_y_continuous(breaks = seq(0,10,1)) +
    scale_color_manual(values = c("F" = "#ffab65", "M" = "#65b9ff")) +
    guides(color = guide_legend(override.aes = list(size = 2)))
} 

p.fh.hc.list.smoking = list()
for(score in c("IMNeuS", "TNeuS", "IFNS")){
  tmp = df.fh.hc[variable == score][!is.na(bmi)]
  tmp[,smoking:=ifelse(regular_smokers == 1, "smokers", "non-smokers")]
  tmp$facetvar = "age"
  tmp$sex = gsub("F", "females", tmp$sex)
  tmp$sex = gsub("M", "males", tmp$sex)
  p.fh.hc.list.smoking[[score]] = ggplot(tmp, aes(x = age, y = value, color = smoking)) +
    geom_point(size = 0.1, alpha = 0.6, shape = 16) +
    geom_smooth(method= "loess", show.legend = F, size = 0.5) +
    facet_grid(cols = vars(sex)) +
    theme_bw() + mytheme +
    labs(y = score, x = "age", title = paste0(score, " score")) +
    theme(
      axis.text = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      axis.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
      strip.text.x = element_text(size = textsize, color = textcolor, margin = margin(0.02, 0.02, 0.2, 0.02, "line")),
      strip.text.y = element_text(size = textsize, color = textcolor, margin = margin(0.1, 0.1, 0.1, 0.1, "line")),
      legend.position = c(0.2, -0.27),
      legend.key.spacing.x = unit(.01, "line"),
      legend.key.spacing.y = unit(.01, "line"),
      legend.key.width = unit(0.1, "line"),
      legend.key.height = unit(0.1, "line"),
      legend.box = element_blank(),
      legend.text = element_text(size = textsize),
      legend.background  = element_blank(),
      plot.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line"))
    ) +
    scale_y_continuous(breaks = seq(0,10,1)) +
    scale_color_manual(values = c("smokers" = "#F76E5E", "non-smokers" = "darkgreen")) +
    guides(color = guide_legend(override.aes = list(size = 2), nrow = 2))
}
```

```{r glm}
### IMNeuS ----
glm.fit.IMNeuS = glm(IMNeuS ~ age + sex + bmi + regular_smokers,  data = framingham_score_pheno[healthy == 1 ])
sum.glm.fit.IMNeuS = summary(glm.fit.IMNeuS)
sum.glm.fit.IMNeuS 

tb.glm.fit.IMNeuS = sum.glm.fit.IMNeuS$coefficients
colnames(tb.glm.fit.IMNeuS) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.IMNeuS = as.data.frame(tb.glm.fit.IMNeuS[,c("regression\ncoefficient",  "p value")])
tb.glm.fit.IMNeuS = tb.glm.fit.IMNeuS[-1,]
rownames(tb.glm.fit.IMNeuS) = c("age", "sex", "BMI", "smoking")
tb.glm.fit.IMNeuS = tb.glm.fit.IMNeuS[c("age", "sex", "BMI", "smoking"),]
tb.glm.fit.IMNeuS$`p value` = formatC(tb.glm.fit.IMNeuS$`p value`, digits = 1, format = "e")
tb.glm.fit.IMNeuS$`regression\ncoefficient` = formatC(tb.glm.fit.IMNeuS$`regression\ncoefficient`, digits = 1, format = "e")

tbg.glm.fit.IMNeuS = tableGrob(tb.glm.fit.IMNeuS, 
                               theme = ttheme_default(
                                 core = list(
                                   fg_params = list(fontsize = 8, fontface = "plain", lineheight = 0.8),
                                   bg_params = list(fill = c("grey97")),
                                   padding = unit(c(0.3, 0.3), "line")
                                 ),
                                 colhead = list(
                                   fg_params = list(fontsize = 8, fontface = "plain", lineheight = 0.8),
                                   bg_params = list(fill = c("grey91")),
                                   padding = unit(c(0.3, 0.3), "line")
                                 ),
                                 rowhead = list(
                                   fg_params = list(fontsize = 8, fontface = "plain"),
                                   bg_params = list(fill = c("grey97")),
                                   padding = unit(c(0.3, 0.3), "line")
                                 )
                               ))

### TNeuS ----
glm.fit.TNeuS = glm(TNeuS ~ age + sex + bmi + regular_smokers + age*sex + age*bmi,  data = framingham_score_pheno[healthy == 1 ])
sum.glm.fit.TNeuS = summary(glm.fit.TNeuS)

tb.glm.fit.TNeuS = sum.glm.fit.TNeuS$coefficients
colnames(tb.glm.fit.TNeuS) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.TNeuS = as.data.frame(tb.glm.fit.TNeuS[,c("regression\ncoefficient",  "p value")])
tb.glm.fit.TNeuS = tb.glm.fit.TNeuS[-1,]
rownames(tb.glm.fit.TNeuS) = c("age", "sex", "BMI", "smoking", "age:sex", "age:BMI")
tb.glm.fit.TNeuS = tb.glm.fit.TNeuS[c("age", "sex", "BMI", "smoking", "age:sex", "age:BMI"),]
tb.glm.fit.TNeuS$`p value` = formatC(tb.glm.fit.TNeuS$`p value`, digits = 1, format = "e")
tb.glm.fit.TNeuS$`regression\ncoefficient` = formatC(tb.glm.fit.TNeuS$`regression\ncoefficient`, digits = 1, format = "e")

tbg.glm.fit.TNeuS = tableGrob(tb.glm.fit.TNeuS, 
                              theme = ttheme_default(
                                core = list(
                                  fg_params = list(fontsize = 8, fontface = "plain", lineheight = 0.8),
                                  bg_params = list(fill = c("grey97")),
                                  padding = unit(c(0.3, 0.3), "line")
                                ),
                                colhead = list(
                                  fg_params = list(fontsize = 8, fontface = "plain", lineheight = 0.8),
                                  bg_params = list(fill = c("grey91")),
                                  padding = unit(c(0.3, 0.3), "line")
                                ),
                                rowhead = list(
                                  fg_params = list(fontsize = 8, fontface = "plain"),
                                  bg_params = list(fill = c("grey97")),
                                  padding = unit(c(0.3, 0.3), "line")
                                )
                              ))

### IFNS ----
glm.fit.IFNS = glm(IFNS ~ age + sex + bmi + regular_smokers,  data = framingham_score_pheno[healthy == 1 ])
sum.glm.fit.IFNS = summary(glm.fit.IFNS)

tb.glm.fit.IFNS = sum.glm.fit.IFNS$coefficients
colnames(tb.glm.fit.IFNS) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.IFNS = as.data.frame(tb.glm.fit.IFNS[,c("regression\ncoefficient",  "p value")])
tb.glm.fit.IFNS = tb.glm.fit.IFNS[-1,]
rownames(tb.glm.fit.IFNS) = c("age", "sex", "BMI", "smoking")
tb.glm.fit.IFNS = tb.glm.fit.IFNS[c("age", "sex", "BMI", "smoking"),]
tb.glm.fit.IFNS$`p value` = formatC(tb.glm.fit.IFNS$`p value`, digits = 1, format = "e")
tb.glm.fit.IFNS$`regression\ncoefficient` = formatC(tb.glm.fit.IFNS$`regression\ncoefficient`, digits = 1, format = "e")

tbg.glm.fit.IFNS = tableGrob(tb.glm.fit.IFNS, 
                             theme = ttheme_default(
                               core = list(
                                 fg_params = list(fontsize = 8, fontface = "plain", lineheight = 0.8),
                                 bg_params = list(fill = c("grey97")),
                                 padding = unit(c(0.3, 0.3), "line")
                               ),
                               colhead = list(
                                 fg_params = list(fontsize = 8, fontface = "plain", lineheight = 0.8),
                                 bg_params = list(fill = c("grey91")),
                                 padding = unit(c(0.3, 0.3), "line")
                               ),
                               rowhead = list(
                                 fg_params = list(fontsize = 8, fontface = "plain"),
                                 bg_params = list(fill = c("grey97")),
                                 padding = unit(c(0.3, 0.3), "line")
                               )
                             ))
# glm, age group ----

### IMNeuS ----
glm.fit.IMNeuS.below53 = glm(IMNeuS ~ age + sex + bmi + regular_smokers,  data = framingham_score_pheno[healthy == 1 & age < 53])
sum.glm.fit.IMNeuS.below53 = summary(glm.fit.IMNeuS.below53)

tb.glm.fit.IMNeuS.below53 = sum.glm.fit.IMNeuS.below53$coefficients
colnames(tb.glm.fit.IMNeuS.below53) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.IMNeuS.below53 = as.data.frame(tb.glm.fit.IMNeuS.below53[,c("regression\ncoefficient", "standard\nerror", "p value")])
tb.glm.fit.IMNeuS.below53 = tb.glm.fit.IMNeuS.below53[-1,]
rownames(tb.glm.fit.IMNeuS.below53) = c("age", "sex", "BMI", "smoking")
tb.glm.fit.IMNeuS.below53 = tb.glm.fit.IMNeuS.below53[c("age", "sex", "BMI", "smoking"),]
tb.glm.fit.IMNeuS.below53$`p value` = formatC(tb.glm.fit.IMNeuS.below53$`p value`, digits = 1, format = "e")
tb.glm.fit.IMNeuS.below53$`regression\ncoefficient` = formatC(tb.glm.fit.IMNeuS.below53$`regression\ncoefficient`, digits = 1, format = "e")
tb.glm.fit.IMNeuS.below53$`standard\nerror` = formatC(tb.glm.fit.IMNeuS.below53$`standard\nerror`, digits = 1, format = "e")

tbg.glm.fit.IMNeuS.below53 <- tableGrob(tb.glm.fit.IMNeuS.below53, 
                                        theme = ttheme_default(
                                          core = list(
                                            fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                            bg_params = list(fill = c("grey97")),
                                            padding = unit(c(0.3, 0.3), "line")
                                          ),
                                          colhead = list(
                                            fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                            bg_params = list(fill = c("grey91")),
                                            padding = unit(c(0.3, 0.3), "line")
                                          ),
                                          rowhead = list(
                                            fg_params = list(fontsize = 10, fontface = "plain"),
                                            bg_params = list(fill = c("grey97")),
                                            padding = unit(c(0.3, 0.3), "line")
                                          )
                                        ))


glm.fit.IMNeuS.above53 = glm(IMNeuS ~ age + sex + bmi + regular_smokers,  data = framingham_score_pheno[healthy == 1 & age >= 53])
sum.glm.fit.IMNeuS.above53 = summary(glm.fit.IMNeuS.above53)
tb.glm.fit.IMNeuS.above53 = sum.glm.fit.IMNeuS.above53$coefficients
colnames(tb.glm.fit.IMNeuS.above53) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.IMNeuS.above53 = as.data.frame(tb.glm.fit.IMNeuS.above53[,c("regression\ncoefficient", "standard\nerror", "p value")])
tb.glm.fit.IMNeuS.above53 = tb.glm.fit.IMNeuS.above53[-1,]
rownames(tb.glm.fit.IMNeuS.above53) = c("age", "sex", "BMI", "smoking")
tb.glm.fit.IMNeuS.above53 = tb.glm.fit.IMNeuS.above53[c("age", "sex", "BMI", "smoking"),]
tb.glm.fit.IMNeuS.above53$`p value` = formatC(tb.glm.fit.IMNeuS.above53$`p value`, digits = 1, format = "e")
tb.glm.fit.IMNeuS.above53$`regression\ncoefficient` = formatC(tb.glm.fit.IMNeuS.above53$`regression\ncoefficient`, digits = 1, format = "e")
tb.glm.fit.IMNeuS.above53$`standard\nerror` = formatC(tb.glm.fit.IMNeuS.above53$`standard\nerror`, digits = 1, format = "e")

tbg.glm.fit.IMNeuS.above53 <- tableGrob(tb.glm.fit.IMNeuS.above53, 
                                        theme = ttheme_default(
                                          core = list(
                                            fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                            bg_params = list(fill = c("grey97")),
                                            padding = unit(c(0.3, 0.3), "line")
                                          ),
                                          colhead = list(
                                            fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                            bg_params = list(fill = c("grey91")),
                                            padding = unit(c(0.3, 0.3), "line")
                                          ),
                                          rowhead = list(
                                            fg_params = list(fontsize = 10, fontface = "plain"),
                                            bg_params = list(fill = c("grey97")),
                                            padding = unit(c(0.3, 0.3), "line")
                                          )
                                        ))

### TNeuS ----
glm.fit.TNeuS.below53 = glm(TNeuS ~ age + sex + bmi + regular_smokers,  data = framingham_score_pheno[healthy == 1 & age < 53])
sum.glm.fit.TNeuS.below53 = summary(glm.fit.TNeuS.below53)
sum.glm.fit.TNeuS.below53


tb.glm.fit.TNeuS.below53 = sum.glm.fit.TNeuS.below53$coefficients
colnames(tb.glm.fit.TNeuS.below53) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.TNeuS.below53 = as.data.frame(tb.glm.fit.TNeuS.below53[,c("regression\ncoefficient", "standard\nerror", "p value")])
tb.glm.fit.TNeuS.below53 = tb.glm.fit.TNeuS.below53[-1,]
rownames(tb.glm.fit.TNeuS.below53) = c( "age", "sex", "BMI", "smoking")
tb.glm.fit.TNeuS.below53 = tb.glm.fit.TNeuS.below53[c("age", "sex", "BMI", "smoking"),]
tb.glm.fit.TNeuS.below53$`p value` = formatC(tb.glm.fit.TNeuS.below53$`p value`, digits = 1, format = "e")
tb.glm.fit.TNeuS.below53$`regression\ncoefficient` = formatC(tb.glm.fit.TNeuS.below53$`regression\ncoefficient`, digits = 1, format = "e")
tb.glm.fit.TNeuS.below53$`standard\nerror` = formatC(tb.glm.fit.TNeuS.below53$`standard\nerror`, digits = 1, format = "e")

tbg.glm.fit.TNeuS.below53 <- tableGrob(tb.glm.fit.TNeuS.below53, 
                                       theme = ttheme_default(
                                         core = list(
                                           fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                           bg_params = list(fill = c("grey97")),
                                           padding = unit(c(0.3, 0.3), "line")
                                         ),
                                         colhead = list(
                                           fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                           bg_params = list(fill = c("grey91")),
                                           padding = unit(c(0.3, 0.3), "line")
                                         ),
                                         rowhead = list(
                                           fg_params = list(fontsize = 10, fontface = "plain"),
                                           bg_params = list(fill = c("grey97")),
                                           padding = unit(c(0.3, 0.3), "line")
                                         )
                                       ))


glm.fit.TNeuS.above53 = glm(TNeuS ~ age + sex + bmi + regular_smokers,  data = framingham_score_pheno[healthy == 1 & age >= 53])
sum.glm.fit.TNeuS.above53 = summary(glm.fit.TNeuS.above53)
sum.glm.fit.TNeuS.above53

tb.glm.fit.TNeuS.above53 = sum.glm.fit.TNeuS.above53$coefficients
colnames(tb.glm.fit.TNeuS.above53) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.TNeuS.above53 = as.data.frame(tb.glm.fit.TNeuS.above53[,c("regression\ncoefficient", "standard\nerror", "p value")])
tb.glm.fit.TNeuS.above53 = tb.glm.fit.TNeuS.above53[-1,]
rownames(tb.glm.fit.TNeuS.above53) = c("age", "sex", "BMI", "smoking")
tb.glm.fit.TNeuS.above53 = tb.glm.fit.TNeuS.above53[c("age", "sex", "BMI", "smoking"),]
tb.glm.fit.TNeuS.above53$`p value` = formatC(tb.glm.fit.TNeuS.above53$`p value`, digits = 1, format = "e")
tb.glm.fit.TNeuS.above53$`regression\ncoefficient` = formatC(tb.glm.fit.TNeuS.above53$`regression\ncoefficient`, digits = 1, format = "e")
tb.glm.fit.TNeuS.above53$`standard\nerror` = formatC(tb.glm.fit.TNeuS.above53$`standard\nerror`, digits = 1, format = "e")

tbg.glm.fit.TNeuS.above53 <- tableGrob(tb.glm.fit.TNeuS.above53, 
                                       theme = ttheme_default(
                                         core = list(
                                           fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                           bg_params = list(fill = c("grey97")),
                                           padding = unit(c(0.3, 0.3), "line")
                                         ),
                                         colhead = list(
                                           fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                           bg_params = list(fill = c("grey91")),
                                           padding = unit(c(0.3, 0.3), "line")
                                         ),
                                         rowhead = list(
                                           fg_params = list(fontsize = 10, fontface = "plain"),
                                           bg_params = list(fill = c("grey97")),
                                           padding = unit(c(0.3, 0.3), "line")
                                         )
                                       ))

### IFNS ----
glm.fit.IFNS.below53 = glm(IFNS ~ age + sex + bmi + regular_smokers,  data = framingham_score_pheno[healthy == 1 & age < 53])
sum.glm.fit.IFNS.below53 = summary(glm.fit.IFNS.below53)
tb.glm.fit.IFNS.below53 = sum.glm.fit.IFNS.below53$coefficients
colnames(tb.glm.fit.IFNS.below53) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.IFNS.below53 = as.data.frame(tb.glm.fit.IFNS.below53[,c("regression\ncoefficient", "standard\nerror", "p value")])
tb.glm.fit.IFNS.below53 = tb.glm.fit.IFNS.below53[-1,]
rownames(tb.glm.fit.IFNS.below53) = c("age", "sex", "BMI", "smoking")
tb.glm.fit.IFNS.below53 = tb.glm.fit.IFNS.below53[c("age", "sex", "BMI", "smoking"),]
tb.glm.fit.IFNS.below53$`p value` = formatC(tb.glm.fit.IFNS.below53$`p value`, digits = 1, format = "e")
tb.glm.fit.IFNS.below53$`regression\ncoefficient` = formatC(tb.glm.fit.IFNS.below53$`regression\ncoefficient`, digits = 1, format = "e")
tb.glm.fit.IFNS.below53$`standard\nerror` = formatC(tb.glm.fit.IFNS.below53$`standard\nerror`, digits = 1, format = "e")

tbg.glm.fit.IFNS.below53 <- tableGrob(tb.glm.fit.IFNS.below53, 
                                      theme = ttheme_default(
                                        core = list(
                                          fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                          bg_params = list(fill = c("grey97")),
                                          padding = unit(c(0.3, 0.3), "line")
                                        ),
                                        colhead = list(
                                          fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                          bg_params = list(fill = c("grey91")),
                                          padding = unit(c(0.3, 0.3), "line")
                                        ),
                                        rowhead = list(
                                          fg_params = list(fontsize = 10, fontface = "plain"),
                                          bg_params = list(fill = c("grey97")),
                                          padding = unit(c(0.3, 0.3), "line")
                                        )
                                      ))


glm.fit.IFNS.above53 = glm(IFNS ~ age + sex + bmi + regular_smokers,  data = framingham_score_pheno[healthy == 1 & age >= 53])
sum.glm.fit.IFNS.above53 = summary(glm.fit.IFNS.above53)
tb.glm.fit.IFNS.above53 = sum.glm.fit.IFNS.above53$coefficients
colnames(tb.glm.fit.IFNS.above53) = c("regression\ncoefficient", "standard\nerror", "t", "p value")
tb.glm.fit.IFNS.above53 = as.data.frame(tb.glm.fit.IFNS.above53[,c("regression\ncoefficient", "standard\nerror", "p value")])
tb.glm.fit.IFNS.above53 = tb.glm.fit.IFNS.above53[-1,]
rownames(tb.glm.fit.IFNS.above53) = c("age", "sex", "BMI", "smoking")
tb.glm.fit.IFNS.above53 = tb.glm.fit.IFNS.above53[c("age", "sex", "BMI", "smoking"),]
tb.glm.fit.IFNS.above53$`p value` = formatC(tb.glm.fit.IFNS.above53$`p value`, digits = 1, format = "e")
tb.glm.fit.IFNS.above53$`regression\ncoefficient` = formatC(tb.glm.fit.IFNS.above53$`regression\ncoefficient`, digits = 1, format = "e")
tb.glm.fit.IFNS.above53$`standard\nerror` = formatC(tb.glm.fit.IFNS.above53$`standard\nerror`, digits = 1, format = "e")

tbg.glm.fit.IFNS.above53 <- tableGrob(tb.glm.fit.IFNS.above53, 
                                      theme = ttheme_default(
                                        core = list(
                                          fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                          bg_params = list(fill = c("grey97")),
                                          padding = unit(c(0.3, 0.3), "line")
                                        ),
                                        colhead = list(
                                          fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                                          bg_params = list(fill = c("grey91")),
                                          padding = unit(c(0.3, 0.3), "line")
                                        ),
                                        rowhead = list(
                                          fg_params = list(fontsize = 10, fontface = "plain"),
                                          bg_params = list(fill = c("grey97")),
                                          padding = unit(c(0.3, 0.3), "line")
                                        )
                                      ))
```

```{r figure5}
dt.ota = data.table(
  " " = c("females", "males"),
  "neutrophils" = c(58,20)
)

dt.ota <- tableGrob(t(dt.ota), 
                    theme = ttheme_default(
                      core = list(
                        fg_params = list(fontsize = 10, fontface = "plain", lineheight = 0.8),
                        bg_params = list(fill = c("grey97")),
                        padding = unit(c(0.3, 0.3), "line")
                      ),
                      colhead = list(
                        fg_params = list(fontsize = 10, fontface = "plain"),
                        padding = unit(c(0.3, 0.3), "line")
                      ),
                      rowhead = list(
                        fg_params = list(fontsize = 10, fontface = "plain"),
                        padding = unit(c(0.3, 0.3), "line")
                      )
                    ))

dt.meta = data.table(
  " " = c("datasets", "females", "males"),
  "WB" = c("10", "337", "279"),
  "PBMC" = c("9", "219", "190")
)

dt.meta <- tableGrob(t(dt.meta), 
                     theme = ttheme_default(
                       core = list(
                         fg_params = list(fontsize = 10, fontface = "plain",lineheight = 0.8),
                         bg_params = list(fill = c("grey97")),
                         padding = unit(c(0.3, 0.3), "line")
                       ),
                       colhead = list(
                         fg_params = list(fontsize = 10, fontface = "plain"),
                         padding = unit(c(0.3, 0.3), "line")
                       ),
                       rowhead = list(
                         fg_params = list(fontsize = 10, fontface = "plain"),
                         padding = unit(c(0.3, 0.3), "line")
                       )
                     ))

df.subsetlist = data.table(
  subset = names(col_celltype),
  color = col_celltype,
  y = 1,
  x = 1:length(col_celltype)
)
df.subsetlist[, label:=ifelse(subset %in% c("IFN 2", "IFN 3"), paste0("\u2217", subset), subset)]

p.subsetlist <- ggplot(df.subsetlist, aes(x = x, y = y, label = label, color = color)) +
  geom_text(angle = -45, size = 1.2, hjust = 0, family = "Arial Unicode MS") + 
  scale_color_identity() +           
  theme_void() +
  coord_cartesian(clip = "off") 

df.subsetlist = data.table(
  subset = names(col_celltype),
  color = col_celltype,
  x = rep(1:3, each = 7),
  y = rev(rep(1:7, 3))
)

p.subsetlist <- ggplot(df.subsetlist, aes(x = x, y = y, label = subset, color = color)) +
  geom_text(angle = 0, size = 2) + 
  scale_color_identity() +           
  theme_void() +
  coord_cartesian(clip = "off") 

df.genesetlist = data.table(
  subset = paste0("| ", names(col_set)),
  color = col_set,
  x = 1,
  y = length(col_set):1
)

p.genesetlist <- ggplot(df.genesetlist, aes(x = x, y = y, label = subset, color = color)) +
  geom_text(angle = 0, size = 2.4, hjust = 0) + 
  scale_color_identity() +           
  theme_void() +
  coord_cartesian(clip = "off") 

cairo_pdf(file = paste0(dir.fig, "figure5.pdf"), width = 8.6, height = 5.8)
pushViewport(viewport(layout = grid.layout(nrow = 106, ncol = 352)))

print(ggplot() + theme_void() + theme(plot.background = element_rect(color = "grey97", fill = "grey97")), vp = viewport(layout.pos.row = 2:32, layout.pos.col =  3:208))
print(ggplot() + theme_void() + theme(plot.background = element_rect(color = "grey97", fill = "grey97")), vp = viewport(layout.pos.row = 34:105, layout.pos.col =  3:95))
print(ggplot() + theme_void() + theme(plot.background = element_rect(color = "grey97", fill = "grey97"), plot.margin = margin(0.2, 0.1, 1, 0.4, "line")), vp = viewport(layout.pos.row = 34:105, layout.pos.col =  98:208))

print(ggplot() + theme_void() + theme(plot.background = element_rect(color = "grey97", fill = "grey97")), vp = viewport(layout.pos.row = 2:30, layout.pos.col =  211:351))
print(ggplot() + theme_void() + theme(plot.background = element_rect(color = "grey97", fill = "grey97")), vp = viewport(layout.pos.row = 32:105, layout.pos.col =  211:351))

print(p.sex.distribution + theme(plot.margin = margin(0.2, 0.04, 0.04, 0.04, "line")), vp = viewport(layout.pos.row = 3:19, layout.pos.col = 5:80))
print(p.box.prop.main + theme(plot.margin = margin(0.01, 0.04, 0.01, 0.04, "line")), vp = viewport(layout.pos.row = 2:32, layout.pos.col = 82:127))
print(p.sex.sc.expr.meta + theme(plot.margin = margin(0.1, 0.1, 0.1, 0.1, "line")), vp = viewport(layout.pos.row = 3:31, layout.pos.col = 130:207))
print(p.subsetlist + theme(plot.margin = margin(0.1, 0.1, 0.1, 0.1, "line")), vp = viewport(layout.pos.row = 20:31, layout.pos.col = 15:60))


grid.text(x = unit(0.62, "npc"), y = unit(0.977, "npc"), just = "left", label = "Immune cells in peripheral blood", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

print(p.umap.celltype.neu.main + theme(plot.margin = margin(0.2,0.2,0.2,0.2, "line")), vp = viewport(layout.pos.row = 6:27, layout.pos.col = 216:265))

pushViewport(viewport(layout.pos.row = 11:24, layout.pos.col = 268:348))
grid.arrange(grobs=pscores.blood, nrow = 1, as.table = TRUE, newpage = FALSE)
upViewport(1)

# purified neutrophils ----
grid.text(x = unit(0.05, "npc"), y = unit(0.66, "npc"), just = "left", label = "Bulk RNA-seq of neutrophils \n(E-GEAD-397)", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

pushViewport(viewport(layout.pos.row = 38:52, layout.pos.col = 3:94))
grid.draw(dt.ota)
popViewport()

print(p.otz + theme(plot.margin = margin(0.1, 0.1, 0, 0.1, "line")), vp = viewport(layout.pos.row = 51:97, layout.pos.col = 4:94))
print(p.sex.bar.otz + theme(plot.margin = margin(0.2, 0.1, 0, 0.1, "line")), vp = viewport(layout.pos.row = 93:103, layout.pos.col = 16:92))

grid.text(x = unit(0.09, "npc"), y = unit(0.064, "npc"), just = "left", label = paste0("NES:  ", round(res.gsea$sex.ota[pathway == "genes.immature.neu"]$NES, 2), " (pval ", formatC(res.gsea$sex.ota[pathway == "genes.immature.neu"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#a93434", lineheight = 0.8))
grid.text(x = unit(0.09, "npc"), y = unit(0.047, "npc"), just = "left", label = paste0("NES: ", round(res.gsea$sex.ota[pathway == "genes.total.neu"]$NES, 2), " (pval ", formatC(res.gsea$sex.ota[pathway == "genes.total.neu"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#1A5E1F", lineheight = 0.8))
grid.text(x = unit(0.09, "npc"), y = unit(0.03, "npc"), just = "left", label = paste0("NES: ", round(res.gsea$sex.ota[pathway == "genes.ifn"]$NES, 2), " (pval ", formatC(res.gsea$sex.ota[pathway == "genes.ifn"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#2E42B8", lineheight = 0.8))

# bulk transcriptome ----
grid.text(x = unit(0.31, "npc"), y = unit(0.66, "npc"), just = "left", label = "Meta-analysis of bulk transcriptome\n", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

pushViewport(viewport(layout.pos.row = 35:52, layout.pos.col = 102:205))
grid.draw(dt.meta)
popViewport()

print(p.sex.PBMC.WB + theme(plot.margin = margin(0.1, 0.1, 0, 0.1, "line")), vp = viewport(layout.pos.row = 51:97, layout.pos.col = 114:207))
print(p.sex.WB + theme(plot.margin = margin(0.2, 0.1, 0.1, 0.1, "line")), vp = viewport(layout.pos.row = 93:102, layout.pos.col = 130:203))
print(p.sex.PBMC + theme(plot.margin = margin(0.1, 0.1, 0.1, 0.1, "line")), vp = viewport(layout.pos.row = 57:87, layout.pos.col = 104:122))

grid.text(x = unit(0.42, "npc"), y = unit(0.064, "npc"), just = "left", label = paste0("NES:  ", round(res.gsea$sex.WB[pathway == "genes.immature.neu"]$NES, 2), " (pval ", formatC(res.gsea$sex.WB[pathway == "genes.immature.neu"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#a93434", lineheight = 0.8))
grid.text(x = unit(0.42, "npc"), y = unit(0.047, "npc"), just = "left", label = paste0("NES: ", round(res.gsea$sex.WB[pathway == "genes.total.neu"]$NES, 2), " (pval ", formatC(res.gsea$sex.WB[pathway == "genes.total.neu"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#1A5E1F", lineheight = 0.8))
grid.text(x = unit(0.42, "npc"), y = unit( 0.03, "npc"), just = "left", label = paste0("NES: ", round(res.gsea$sex.WB[pathway == "genes.ifn"]$NES, 2), " (pval ", formatC(res.gsea$sex.WB[pathway == "genes.ifn"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#2E42B8", lineheight = 0.8))

grid.text(x = unit(0.282, "npc"), y = unit(0.215, "npc"), just = "left", rot = 90, label = paste0("NES:  ", round(res.gsea$sex.PBMC[pathway == "genes.immature.neu"]$NES, 2), " (pval ", formatC(res.gsea$sex.PBMC[pathway == "genes.immature.neu"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#a93434", lineheight = 0.8))
grid.text(x = unit(0.292, "npc"), y = unit(0.215, "npc"), just = "left", rot = 90, label = paste0("NES:  ", round(res.gsea$sex.PBMC[pathway == "genes.total.neu"]$NES, 2), " (pval ", formatC(res.gsea$sex.PBMC[pathway == "genes.total.neu"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#1A5E1F", lineheight = 0.8))
grid.text(x = unit(0.302, "npc"), y = unit(0.215, "npc"), just = "left", rot = 90, label = paste0("NES: ", round(res.gsea$sex.PBMC[pathway == "genes.ifn"]$NES, 2), " (pval ", formatC(res.gsea$sex.PBMC[pathway == "genes.ifn"]$pval, digits = 2, format = "e"), ")"), gp = gpar(fontsize = 7, fontface = "plain", col = "#2E42B8", lineheight = 0.8))

print(p.genesetlist + theme(plot.margin = margin(0.2, 0.1, 0.01, 0.1, "line")), vp = viewport(layout.pos.row = 98:103, layout.pos.col = 100:110))

# framingham ----
grid.text(x = unit(0.63, "npc"), y = unit(0.68, "npc"), just = "left", label = "Framingham cohort study\n823 healthy subjects (433 females, 390 males)", gp = gpar(fontsize = textsize, fontface = "plain", col = textcolor, lineheight = 0.8))

print(p.fh.hc.list.age$IMNeuS + theme(plot.margin = margin(0.2,0,0.2,0, "line")), vp = viewport(layout.pos.row = 38:60, layout.pos.col = 212:253))
print(p.fh.hc.list.bmi$IMNeuS + theme(plot.margin = margin(0.2,0,0.2,0, "line")), vp = viewport(layout.pos.row = 38:60, layout.pos.col = 254:286))

pushViewport(viewport(layout.pos.row = 38:60, layout.pos.col = 286:352))
grid.draw(tbg.glm.fit.IMNeuS)
popViewport()

print(p.fh.hc.list.age$TNeuS + theme(plot.margin = margin(0.2, 0, 0.2, 0, "line")), vp = viewport(layout.pos.row = 61:83, layout.pos.col = 212:253))
print(p.fh.hc.list.bmi$TNeuS + theme(plot.margin = margin(0.2,0,0.2,0, "line")), vp = viewport(layout.pos.row = 61:83, layout.pos.col = 254:286))

pushViewport(viewport(layout.pos.row = 61:83, layout.pos.col = 286:352))
grid.draw(tbg.glm.fit.TNeuS)
popViewport()

print(p.fh.hc.list.age$IFNS + theme(plot.margin = margin(0.2, 0, 0.2, 0, "line")), vp = viewport(layout.pos.row = 84:106, layout.pos.col = 212:253))
print(p.fh.hc.list.bmi$IFNS + theme(plot.margin = margin(0.2,0,0.2,0, "line")), vp = viewport(layout.pos.row = 84:106, layout.pos.col = 254:286))

pushViewport(viewport(layout.pos.row = 84:106, layout.pos.col = 286:352))
grid.draw(tbg.glm.fit.IFNS)
popViewport()

# other ----
grid.text(x = unit(0.012, "npc"), y = unit(0.98, "npc"), label = "A", gp = gpar(fontsize =8, fontface = "bold"))
grid.text(x = unit(0.24, "npc"), y = unit(0.98, "npc"), label = "B", gp = gpar(fontsize =8, fontface = "bold"))
grid.text(x = unit(0.372, "npc"), y = unit(0.98, "npc"), label = "C", gp = gpar(fontsize =8, fontface = "bold"))
grid.text(x = unit(0.61, "npc"), y = unit(0.98, "npc"), label = "D", gp = gpar(fontsize =8, fontface = "bold"))

grid.text(x = unit(0.012, "npc"), y = unit(0.675, "npc"), label = "E", gp = gpar(fontsize =8, fontface = "bold"))
grid.text(x = unit(0.29, "npc"), y = unit(0.675, "npc"), label = "F", gp = gpar(fontsize =8, fontface = "bold"))

grid.text(x = unit(0.61, "npc"), y = unit(0.635, "npc"), label = "G", gp = gpar(fontsize =8, fontface = "bold"))
grid.text(x = unit(0.61, "npc"), y = unit(0.418, "npc"), label = "H", gp = gpar(fontsize =8, fontface = "bold"))
grid.text(x = unit(0.61, "npc"), y = unit(0.205, "npc"), label = "I", gp = gpar(fontsize =8, fontface = "bold"))

dev.off()
```

```{r figureS5}
cairo_pdf(file = paste0(dir.fig, "figureS5.pdf"), width = 10, height = 9)
pushViewport(viewport(layout = grid.layout(nrow = 90, ncol = 40)))
print(p.box.prop, vp = viewport(layout.pos.row = 1:40, layout.pos.col = 1:40))

pushViewport(viewport(layout.pos.row = 47:57, layout.pos.col = 3:14))
grid.draw(tbg.glm.fit.IMNeuS.below53)
popViewport()

pushViewport(viewport(layout.pos.row = 58:68, layout.pos.col = 3:14))
grid.draw(tbg.glm.fit.IMNeuS.above53)
popViewport()

pushViewport(viewport(layout.pos.row = 47:57, layout.pos.col = 15:26))
grid.draw(tbg.glm.fit.TNeuS.below53)
popViewport()

pushViewport(viewport(layout.pos.row = 58:68, layout.pos.col = 15:26))
grid.draw(tbg.glm.fit.TNeuS.above53)
popViewport()

pushViewport(viewport(layout.pos.row = 47:57, layout.pos.col = 27:38))
grid.draw(tbg.glm.fit.IFNS.below53)
popViewport()

pushViewport(viewport(layout.pos.row = 58:68, layout.pos.col = 27:38))
grid.draw(tbg.glm.fit.IFNS.above53)
popViewport()

grid.text(x= unit(0.012, "npc"), y = unit(0.98, "npc"), label = "A", gp = gpar(fontsize = 10, fontface = "bold"))
grid.text(x= unit(0.012, "npc"), y = unit(0.54, "npc"), label = "B", gp = gpar(fontsize = 10, fontface = "bold"))
grid.text(x= unit(0.012, "npc"), y = unit(0.22, "npc"), label = "C", gp = gpar(fontsize = 10, fontface = "bold"))
grid.text(x= unit(0.03, "npc"), y = unit(0.53, "npc"), label= "Framingham cohort study, 823 healthy subjects", gp = gpar(fontsize = 10, fontface = "plain"), hjust = 0)

print(p.fh.hc.list.smoking$IMNeuS + theme(plot.margin = margin(0.2, 0, 0.2, 0, "line")), vp = viewport(layout.pos.row = 70:89, layout.pos.col = 3:13))

print(p.fh.hc.list.smoking$TNeuS + theme(plot.margin = margin(0.2, 0, 0.2, 0, "line")), vp = viewport(layout.pos.row = 70:89, layout.pos.col = 15:25))

print(p.fh.hc.list.smoking$IFNS + theme(plot.margin = margin(0.2, 0, 0.2, 0, "line")), vp = viewport(layout.pos.row = 70:89, layout.pos.col = 27:37))

grid.text(x= unit(0.18, "npc"), y = unit(0.5, "npc"), label= "IMNeuS score", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)
grid.text(x= unit(0.47, "npc"), y = unit(0.5, "npc"), label= "TNeuS score", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)
grid.text(x= unit(0.78, "npc"), y = unit(0.5, "npc"), label= "IFNS score", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)

grid.text(x= unit(0.03, "npc"), y = unit(0.46, "npc"), label= "< 53 y/o", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)
grid.text(x= unit(0.03, "npc"), y = unit(0.34, "npc"), label= "≥ 53 y/o", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)

grid.text(x= unit(0.33, "npc"), y = unit(0.46, "npc"), label= "< 53 y/o", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)
grid.text(x= unit(0.33, "npc"), y = unit(0.34, "npc"), label= "≥ 53 y/o", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)

grid.text(x= unit(0.63, "npc"), y = unit(0.46, "npc"), label= "< 53 y/o", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)
grid.text(x= unit(0.63, "npc"), y = unit(0.34, "npc"), label= "≥ 53 y/o", gp = gpar(fontsize = 9, fontface = "plain"), hjust = 0)

dev.off()
```

# Figure 6&S6

```{r framingham survival functions}
print_cox_model_summary <- function(cox_model)
{
  summary_cox_model <- summary(cox_model)
  summary_cox_model_data_frame <- as.data.frame(summary_cox_model$coefficients)[, c(2, 5)]
  colnames(summary_cox_model_data_frame) <- c("hazard ratio", "pval")
  summary_cox_model_data_frame$`hazard ratio` <- paste0(round(summary_cox_model_data_frame$`hazard ratio`, 2), " (", round(summary_cox_model$conf.int[, 3], 2), ", ", round(summary_cox_model$conf.int[, 4], 2), ")")
  summary_cox_model_data_frame$pval <- as.character(ifelse(summary_cox_model_data_frame$pval < 0.01, format(summary_cox_model_data_frame$pval, digits = 2, scientific = TRUE), round(summary_cox_model_data_frame$pval, 2)))
  return(summary_cox_model_data_frame)
}

evaluate_framingham_mortality_models <- function(n_years_cutoff, score_pheno_matrix, score_name, score_name_print)
{
  score_pheno_matrix$status <- 1
  score_pheno_matrix$status[is.na(score_pheno_matrix$death_date) | score_pheno_matrix$death_date > 365 * n_years_cutoff] <- 0
  score_pheno_matrix$death_date[is.na(score_pheno_matrix$death_date) | score_pheno_matrix$death_date > 365 * n_years_cutoff] <- 365 * n_years_cutoff
  score_pheno_matrix$death_date <- score_pheno_matrix$death_date / 365
  score_pheno_matrix$disease_status <- as.factor(1 - score_pheno_matrix$healthy)
  print(sum(score_pheno_matrix$status == 1))
  print(sum(score_pheno_matrix$status == 0))
  #
  cox_model_all_var <- coxph(as.formula(paste0("Surv(death_date, status) ~ ", score_name, " + age + as.factor(sex) + bmi + disease_status + as.factor(regular_smokers)")), data = score_pheno_matrix)
  
  #cox_model_all_var <- coxph(as.formula(paste0("Surv(death_date, status) ~ ", score_name, " + strata(age) + as.factor(sex) + bmi + strata(disease_status) + as.factor(regular_smokers)")), data = score_pheno_matrix)
  
  ph_test <- cox.zph(cox_model_all_var)
  ph_test
  
  cox_model_all_var_summary <- print_cox_model_summary(cox_model_all_var)
  rownames(cox_model_all_var_summary) <- c(score_name_print, "age", "sex", "BMI", "disease", "smoking")
  #
  lm_score_adjust <- lm(as.formula(paste0(score_name, " ~ age + as.factor(sex) + bmi + disease_status + as.factor(regular_smokers)")), data = score_pheno_matrix)
  score_pheno_matrix$residuals <- NA
  score_pheno_matrix$residuals[as.numeric(names(lm_score_adjust$residuals))] <- lm_score_adjust$residuals
  score_pheno_matrix$score_discretized <- "high"
  score_pheno_matrix$score_discretized[score_pheno_matrix$residuals < 0] <- "low"
  score_pheno_matrix$score_discretized[is.na(score_pheno_matrix$residuals)] <- NA
  score_pheno_matrix$score_discretized <- factor(score_pheno_matrix$score_discretized, levels = c("low", "high"))
  #
  residual_cox_model_discretized <- coxph(Surv(death_date, status) ~ score_discretized, data = score_pheno_matrix)
  residual_cox_model_discretized_summary <- print_cox_model_summary(residual_cox_model_discretized)
  rownames(residual_cox_model_discretized_summary) <- paste0("adjusted discretized ", score_name_print)
  #
  residual_cox_model <- coxph(Surv(death_date, status) ~ residuals, data = score_pheno_matrix)
  residual_cox_model_summary <- print_cox_model_summary(residual_cox_model)
  rownames(residual_cox_model_summary) <- paste0("adjusted ", score_name_print)
  #
  km_fit <- survfit(Surv(death_date, status) ~ score_discretized, data = score_pheno_matrix)
  # km_plot <- autoplot(km_fit, conf.int = TRUE) +
  #   labs(color = paste0("Adjusted ", score_name_print, "\n(discretized)"), fill = paste0("Adjusted ", score_name_print, "\n(discretized)")) +
  #   theme(legend.position = c(0.3, 0.25)) +
  #   xlab("Time (years)") + ylab("Survival rate") +
  #   scale_fill_manual(values = c("#1A5E1F", "#EE6C00")) +
  #   scale_colour_manual(values = c("#1A5E1F", "#EE6C00"))
  
  return(list(cox_model_all_var_summary = cox_model_all_var_summary, residual_cox_model_discretized_summary = residual_cox_model_discretized_summary, residual_cox_model_summary = residual_cox_model_summary, km_fit = km_fit))
}

evaluate_framingham_mortality_models2 <- function(n_years_cutoff, score_pheno_matrix, score_name, score_name_print)
{
  score_pheno_matrix$status <- 1
  score_pheno_matrix$status[is.na(score_pheno_matrix$death_date) | score_pheno_matrix$death_date > 365 * n_years_cutoff] <- 0
  score_pheno_matrix$death_date[is.na(score_pheno_matrix$death_date) | score_pheno_matrix$death_date > 365 * n_years_cutoff] <- 365 * n_years_cutoff
  score_pheno_matrix$death_date <- score_pheno_matrix$death_date / 365
  score_pheno_matrix$disease_status <- as.factor(1 - score_pheno_matrix$healthy)
  print(sum(score_pheno_matrix$status == 1))
  print(sum(score_pheno_matrix$status == 0))
  #
  cox_model_all_var <- coxph(as.formula(paste0("Surv(death_date, status) ~ ", score_name, " + age +  bmi + disease_status + as.factor(regular_smokers)")), data = score_pheno_matrix)
  cox_model_all_var_summary <- print_cox_model_summary(cox_model_all_var)
  rownames(cox_model_all_var_summary) <- c(score_name_print, "age", "BMI", "disease", "smoking")
  #
  lm_score_adjust <- lm(as.formula(paste0(score_name, " ~ age + bmi + disease_status + as.factor(regular_smokers)")), data = score_pheno_matrix)
  score_pheno_matrix$residuals <- NA
  score_pheno_matrix$residuals[as.numeric(names(lm_score_adjust$residuals))] <- lm_score_adjust$residuals
  score_pheno_matrix$score_discretized <- "high"
  score_pheno_matrix$score_discretized[score_pheno_matrix$residuals < 0] <- "low"
  score_pheno_matrix$score_discretized[is.na(score_pheno_matrix$residuals)] <- NA
  score_pheno_matrix$score_discretized <- factor(score_pheno_matrix$score_discretized, levels = c("low", "high"))
  #
  residual_cox_model_discretized <- coxph(Surv(death_date, status) ~ score_discretized, data = score_pheno_matrix)
  residual_cox_model_discretized_summary <- print_cox_model_summary(residual_cox_model_discretized)
  rownames(residual_cox_model_discretized_summary) <- paste0("adjusted discretized ", score_name_print)
  #
  residual_cox_model <- coxph(Surv(death_date, status) ~ residuals, data = score_pheno_matrix)
  residual_cox_model_summary <- print_cox_model_summary(residual_cox_model)
  rownames(residual_cox_model_summary) <- paste0("adjusted ", score_name_print)
  km_fit <- survfit(Surv(death_date, status) ~ score_discretized, data = score_pheno_matrix)
  
  return(list(cox_model_all_var_summary = cox_model_all_var_summary, residual_cox_model_discretized_summary = residual_cox_model_discretized_summary, residual_cox_model_summary = residual_cox_model_summary, km_fit = km_fit))
}
```

```{r fh survival}
col_survival = c("#214DC8", "#FF7F0E")

dichromat_pal("DarkRedtoBlue.12")(12)[c(2,11)]

# all 
framingham_score_pheno = fread(paste0(dir, "fh_age51-54conorm.score_pheno.csv"))

scores_to_test <- c("IMNeuS", "TNeuS", "IFNS")

survival_summary = list()
for(score_to_test in scores_to_test){
  cox_model_summaries <- evaluate_framingham_mortality_models(7, framingham_score_pheno, score_to_test, score_to_test)
  survival_summary[[score_to_test]] = cox_model_summaries
}

### IMNeuS ----
dt.surv.IMNeuS = survival_summary$IMNeuS$cox_model_all_var_summary
rownames(dt.surv.IMNeuS)[1] = "score"

dt.surv.IMNeuS.adj = rbind(survival_summary$IMNeuS$residual_cox_model_summary,
                           survival_summary$IMNeuS$residual_cox_model_discretized_summary)
rownames(dt.surv.IMNeuS.adj) = c("score\nadjusted", "score\nadjusted\n discretized")

tb.surv.IMNeuS <- tableGrob(dt.surv.IMNeuS, 
                            theme = ttheme_default(
                              core = list(
                                fg_params = list(fontsize = 10, fontface = "plain", hjust = 1, x = 0.99),
                                bg_params = list(fill = c("grey97")),
                                padding = unit(c(0.35, 0.3), "line")
                              ),
                              colhead = list(
                                fg_params = list(fontsize = 10, fontface = "plain"),
                                bg_params = list(fill = c("grey89")),
                                padding = unit(c(0.35, 0.3), "line")
                              ),
                              rowhead = list(
                                fg_params = list(fontsize = 10, fontface = "plain"),
                                padding = unit(c(0.3, 0.3), "line")
                              )
                            ))


tb.surv.IMNeuS.adj <- tableGrob(dt.surv.IMNeuS.adj, 
                                theme = ttheme_default(
                                  core = list(
                                    fg_params = list(fontsize = 10, fontface = "plain", hjust = 1, x = 0.99),
                                    bg_params = list(fill = c("grey97")),
                                    padding = unit(c(0.35, 0.3), "line")
                                  ),
                                  colhead = list(
                                    fg_params = list(fontsize = 10, fontface = "plain"),
                                    bg_params = list(fill = c("grey89")),
                                    padding = unit(c(0.35, 0.3), "line")
                                  ),
                                  rowhead = list(
                                    fg_params = list(fontsize = 10, fontface = "plain",lineheight = 0.7),
                                    
                                    padding = unit(c(0.3, 0.3), "line")
                                  )
                                ))

p.surv.IMNeuS = autoplot(survival_summary$IMNeuS$km_fit, conf.int = TRUE) +
  labs(color = "IMNeuS adjusted \n(discretized)", fill = "IMNeuS adjusted \n(discretized)") +
  xlab("time (years)") + ylab("survival rate") +
  scale_fill_manual(values = col_survival) +
  scale_colour_manual(values = col_survival) +
  theme_bw() + 
  theme(legend.position = "none",
        panel.grid.minor = element_blank(),
        panel.grid.major = element_line(color = "grey67",linewidth=0.05,linetype = 2),
        panel.border = element_blank(),
        plot.margin = margin(0, 0, 0, 0, "line"),
        axis.text = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
        axis.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
        axis.line = element_line(color = linecolor,linewidth=linesize)) 


### TNeuS ----
dt.surv.TNeuS = survival_summary$TNeuS$cox_model_all_var_summary
rownames(dt.surv.TNeuS)[1] = "TNeuS"
rownames(dt.surv.TNeuS) = NULL

dt.surv.TNeuS.adj = rbind(survival_summary$TNeuS$residual_cox_model_summary,
                          survival_summary$TNeuS$residual_cox_model_discretized_summary)
rownames(dt.surv.TNeuS.adj) = c("TNeuS\nadjusted", "TNeuS\nadjusted\n discretized")
rownames(dt.surv.TNeuS.adj) = c(" \n", " \n\n")
dt.surv.TNeuS.adj$pval = paste0("   ",dt.surv.TNeuS.adj$pval)
dt.surv.TNeuS.adj$`hazard ratio` = paste0(" ",dt.surv.TNeuS.adj$`hazard ratio`)

tb.surv.TNeuS <- tableGrob(dt.surv.TNeuS,  rows = NULL,
                           theme = ttheme_default(
                             core = list(
                               fg_params = list(fontsize = 10, fontface = "plain", hjust = 1, x = 0.99),
                               bg_params = list(fill = c("grey97")),
                               padding = unit(c(0.35, 0.3), "line")
                             ),
                             colhead = list(
                               fg_params = list(fontsize = 10, fontface = "plain"),
                               bg_params = list(fill = c("grey89")),
                               padding = unit(c(0.35, 0.3), "line")
                             ),
                             rowhead = list(
                               fg_params = list(fontsize = 10, fontface = "plain"),
                               padding = unit(c(0.3, 0.3), "line")
                             )
                           ))


tb.surv.TNeuS.adj <- tableGrob(dt.surv.TNeuS.adj, 
                               theme = ttheme_default(
                                 core = list(
                                   fg_params = list(fontsize = 10, fontface = "plain", hjust = 1, x = 0.99),
                                   bg_params = list(fill = c("grey97")),
                                   padding = unit(c(0.35, 0.3), "line")
                                 ),
                                 colhead = list(
                                   fg_params = list(fontsize = 10, fontface = "plain"),
                                   bg_params = list(fill = c("grey89")),
                                   padding = unit(c(0.35, 0.3), "line")
                                 ),
                                 rowhead = list(
                                   fg_params = list(fontsize = 10, fontface = "plain",lineheight = 0.7),
                                   
                                   padding = unit(c(0.3, 0.3), "line")
                                 )
                               ))


p.surv.TNeuS = autoplot(survival_summary$TNeuS$km_fit, conf.int = TRUE) +
  labs(color = "TNeuS adjusted \n(discretized)", fill = "TNeuS adjusted \n(discretized)") +
  labs(x = "time (years)", y = NULL) +
  scale_fill_manual(values = col_survival) +
  scale_colour_manual(values = col_survival) +
  theme_bw() + 
  theme(legend.position = "none",
        panel.grid.minor = element_blank(),
        panel.grid.major = element_line(color = "grey67",linewidth=0.05,linetype = 2),
        panel.border = element_blank(),
        plot.margin = margin(0, 0, 0, 0, "line"),
        axis.text = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
        axis.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
        axis.line = element_line(color = linecolor,linewidth=linesize)) 


### IFNS ----
dt.surv.IFNS = survival_summary$IFNS$cox_model_all_var_summary
rownames(dt.surv.IFNS)[1] = "IFNS"
rownames(dt.surv.IFNS) = NULL

dt.surv.IFNS.adj = rbind(survival_summary$IFNS$residual_cox_model_summary,
                         survival_summary$IFNS$residual_cox_model_discretized_summary)
rownames(dt.surv.IFNS.adj) = c("IFNS\nadjusted", "IFNS\nadjusted\n discretized")
rownames(dt.surv.IFNS.adj) = c(" \n", " \n\n")
dt.surv.IFNS.adj$pval = paste0("   ",dt.surv.IFNS.adj$pval)
dt.surv.IFNS.adj$`hazard ratio` = paste0(" ",dt.surv.IFNS.adj$`hazard ratio`)

tb.surv.IFNS <- tableGrob(dt.surv.IFNS,  rows = NULL,
                          theme = ttheme_default(
                            core = list(
                              fg_params = list(fontsize = 10, fontface = "plain", hjust = 1, x = 0.99),
                              bg_params = list(fill = c("grey97")),
                              padding = unit(c(0.35, 0.3), "line")
                            ),
                            colhead = list(
                              fg_params = list(fontsize = 10, fontface = "plain"),
                              bg_params = list(fill = c("grey89")),
                              padding = unit(c(0.35, 0.3), "line")
                            ),
                            rowhead = list(
                              fg_params = list(fontsize = 10, fontface = "plain"),
                              padding = unit(c(0.3, 0.3), "line")
                            )
                          ))


tb.surv.IFNS.adj <- tableGrob(dt.surv.IFNS.adj, 
                              theme = ttheme_default(
                                core = list(
                                  fg_params = list(fontsize = 10, fontface = "plain", hjust = 1, x = 0.99),
                                  bg_params = list(fill = c("grey97")),
                                  padding = unit(c(0.35, 0.3), "line")
                                ),
                                colhead = list(
                                  fg_params = list(fontsize = 10, fontface = "plain"),
                                  bg_params = list(fill = c("grey89")),
                                  padding = unit(c(0.35, 0.3), "line")
                                ),
                                rowhead = list(
                                  fg_params = list(fontsize = 10, fontface = "plain",lineheight = 0.7),
                                  
                                  padding = unit(c(0.3, 0.3), "line")
                                )
                              ))

p.surv.IFNS = autoplot(survival_summary$IFNS$km_fit, conf.int = TRUE) +
  labs(color = "IFNS adjusted \n(discretized)", fill = "IFNS adjusted \n(discretized)") +
  labs(x = "time (years)", y = NULL) +
  scale_fill_manual(values = col_survival) +
  scale_colour_manual(values = col_survival) +
  theme_bw() + 
  theme(legend.position = "none",
        panel.grid.minor = element_blank(),
        panel.grid.major = element_line(color = "grey67",linewidth=0.05,linetype = 2),
        panel.border = element_blank(),
        plot.margin = margin(0, 0, 0, 0, "line"),
        axis.text = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
        axis.title = element_text(size = textsize, color = textcolor, margin = margin(0, 0, 0, 0, "line")),
        axis.line = element_line(color = linecolor,linewidth=linesize)) 
```

```{r correlations}
framingham_score_pheno = fread(paste0(dir, "fh_age51-54conorm.score_pheno.csv"))
framingham_score_pheno2 = fread(paste0(dir, "framingham_conormalized_by_healthy_matrix.csv"))
framingham_score_pheno2 = framingham_score_pheno2[order(match(shareid, framingham_score_pheno$shareid))]
all.equal(framingham_score_pheno2$shareid, framingham_score_pheno$shareid)
framingham_score_pheno2 = cbind(framingham_score_pheno2, framingham_score_pheno[,c("IMNeuS", "TNeuS", "IFNS" , "IMNeuS_adj", "TNeuS_adj", "IFNS_adj")])

framingham_score_pheno2$disease <- as.factor(1 - framingham_score_pheno2$healthy)
framingham_score_pheno2$LDL_plasma_estimated <- framingham_score_pheno2$total_cholestrol_plasma - framingham_score_pheno2$HDL_cholestrol_plasma - framingham_score_pheno2$triglycerides_plasma / 5
list_of_lab_scores <- c("glucose_plasma", "A1C",  "triglycerides_plasma", "albumin_urine", "creatinine_serum", "HDL_cholestrol_plasma")

framingham_score_pheno2$diabetes_yesno = factor(framingham_score_pheno2$diabetes)
framingham_score_pheno2[,diabetes_yesno:=ifelse(diabetes %in% c(1), "diabetes", "no diabetes")]
framingham_score_pheno2[,BMI_categ:=ifelse(bmi > median(framingham_score_pheno2$bmi, na.rm = T), "higher BMI", "lower BMI")]
framingham_score_pheno2[,sex_more:=ifelse(sex %in% c("F"), "female", "male")]

FraminghamCorrelationHT <- function(dtf, ptitle){
  list_of_gene_scores <- c("IMNeuS", "TNeuS", "IFNS")
  list_of_lab_scores <- c("glucose_plasma", "A1C",  "triglycerides_plasma", "HDL_cholestrol_plasma")
  corr_estimate <- matrix(0, nrow = length(list_of_gene_scores), ncol = length(list_of_lab_scores))
  rownames(corr_estimate) <- list_of_gene_scores
  colnames(corr_estimate) <- list_of_lab_scores
  corr_p_value <- matrix(0, nrow = length(list_of_gene_scores), ncol = length(list_of_lab_scores))
  rownames(corr_p_value) <- list_of_gene_scores
  colnames(corr_p_value) <- list_of_lab_scores
  
  for(gene_score in list_of_gene_scores){
    for(lab_score in list_of_lab_scores){
      cor_test <- cor.test(dtf[[lab_score]], dtf[[gene_score]], method = "spearman")
      corr_estimate[gene_score, lab_score] <- cor_test$estimate
      corr_p_value[gene_score, lab_score] <- cor_test$p.value
    }
  }
  
  colnames(corr_estimate) <- c("\nglucose\n(plasma)", "\nHbA1C", "\ntriglycerides\n(plasma)",  "\nHDL\n(plasma)")
  colnames(corr_p_value) <- colnames(corr_estimate)
  rownames(corr_estimate) <- c("IMNeuS", "TNeuS", "IFNS")
  rownames(corr_p_value) <- rownames(corr_estimate)
  
  corr_estimate
  corr_p_value
  
  corr_p_symbol = corr_p_value
  corr_p_symbol[corr_p_value < 0.001] = "***"
  corr_p_symbol[corr_p_value >= 0.001 & corr_p_value < 0.01] = "**"
  corr_p_symbol[corr_p_value >= 0.01 & corr_p_value <= 0.05] = "*"
  corr_p_symbol[corr_p_value > 0.05] = ""
  
  col_fun = circlize::colorRamp2(seq(-0.3, 0.3, length = 13), c(dichromat_pal("DarkRedtoBlue.12")(12)[1:6],  "white", dichromat_pal("DarkRedtoBlue.12")(12)[7:12]))
  
  pht =
    Heatmap(corr_estimate,
            #rect_gp = gpar(type = "none"), 
            col = col_fun,
            na_col = "grey",
            cluster_columns = F, cluster_rows = F,
            show_row_dend = F ,show_column_dend = F,
            column_dend_reorder = F,
            column_dend_side  = "top",
            row_dend_gp = gpar(lwd = 0.3),
            column_dend_gp = gpar(lwd = 0.3),
            clustering_method_rows = "ward.D", clustering_method_columns  = "ward.D",
            clustering_distance_rows  = "spearman", clustering_distance_columns  = "spearman",
            show_column_names = T, show_row_names = T,
            column_names_side = "bottom", column_names_rot = 0, column_names_gp = gpar(fontsize = 10),
            row_names_gp = gpar(fontsize = 10, fontface = "plain"),
            row_names_side = "left",
            row_names_max_width = unit(30, "line"),
            column_title = ptitle, column_title_side = "top", column_title_gp = gpar(fontsize = 10, fontface = "plain"),
            column_names_centered = T,
            show_heatmap_legend = T,
            layer_fun = function(j, i, x, y, width, height, fill) {
              v = pindex(corr_p_value, i, j)
              v2 = pindex(corr_p_symbol, i, j)
              l = v <= 0.05
              grid.text(v2[l], x[l], y[l], gp = gpar(fontsize = 10))
            },
            heatmap_legend_param = list(title = "correlation", title_gp = gpar(fontsize = 10), labels_gp = gpar(fontsize = 10), legend_width = unit(.2, "line"), legend_height = unit(2, "line"), grid_width = unit(.2, "line"), grid_height = unit(2, "line"), at = seq(-0.2, 0.2, length = 3), by_row = F, direction = "vertical", title_position = "leftcenter-rot"))
  
  return(pht)
}
```

```{r htplots}
pht.all = FraminghamCorrelationHT(framingham_score_pheno2, "Framingham cohort study, 5321 subjects")
pht.females = FraminghamCorrelationHT(framingham_score_pheno2[sex %in% "F"], "Framingham cohort study, 2861 females")
pht.males = FraminghamCorrelationHT(framingham_score_pheno2[sex %in% "M"], "Framingham cohort study, 2460 males")
pht.nodiabetes = FraminghamCorrelationHT(framingham_score_pheno2[diabetes == 0], "Framingham cohort study, 4846 subjects without diabetes")
```

```{r figure6} 
df.lg.highlow = data.table(
  sex = c("", "| high", "| low", ""),
  color = c("white", rev(col_survival), "white"),
  x = 1,
  y = 1:4
)
p.lg.highlow <- ggplot(df.lg.highlow, aes(x = x, y = y, label = sex, color = color)) +
  geom_text(angle = 0, size = 3, hjust = 0) + 
  scale_color_identity() +           
  theme_void()

cairo_pdf(file = paste0(dir.fig, "figure6.pdf"), width = 6, height = 6, family = "Arial Unicode MS")
pushViewport(viewport(layout = grid.layout(nrow = 98, ncol = 68)))

pushViewport(viewport(layout.pos.row = 3:23, layout.pos.col = 6:60))
draw(pht.all, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "top", newpage = FALSE)
upViewport(1)

# IMNeuS survival
pushViewport(viewport(layout.pos.row = 33:58, layout.pos.col = 4:26))
grid.draw(tb.surv.IMNeuS)
popViewport()

pushViewport(viewport(layout.pos.row = 59:66, layout.pos.col = 2:26))
grid.draw(tb.surv.IMNeuS.adj)
popViewport()

print(p.surv.IMNeuS + theme(plot.margin = margin(0.04, 0.04, 0.04, 0.04, "line")), vp = viewport(layout.pos.row = 72:97, layout.pos.col = 5:26))

# TNeuS survival
pushViewport(viewport(layout.pos.row = 33:58, layout.pos.col = 29:46))
grid.draw(tb.surv.TNeuS)
popViewport()

pushViewport(viewport(layout.pos.row = 59:66, layout.pos.col = 28:46))
grid.draw(tb.surv.TNeuS.adj)
popViewport()

print(p.surv.TNeuS + theme(plot.margin = margin(0.04, 0.04, 0.04, 0.04, "line")), vp = viewport(layout.pos.row = 72:97, layout.pos.col = 27:46))

# IFNS survival
pushViewport(viewport(layout.pos.row = 33:58, layout.pos.col = 49:66))
grid.draw(tb.surv.IFNS)
popViewport()

pushViewport(viewport(layout.pos.row = 59:66, layout.pos.col = 48:66))
grid.draw(tb.surv.IFNS.adj)
popViewport()

print(p.surv.IFNS + theme(plot.margin = margin(0.04, 0.04, 0.04, 0.04, "line")), vp = viewport(layout.pos.row = 72:97, layout.pos.col = 47:66))

grid.text(x = unit(0.5, "npc"), y = unit(0.7, "npc"), label = "Cox proportional hazards model, 7-year survival", gp = gpar(fontsize = 10, fontface = "plain", fontfamily = "Arial", lineheight = 0.8), hjust = 0.5 )

print(p.lg.highlow + theme(plot.margin = margin(0.04, 0.04, 0.04, 0.04, "line")), vp = viewport(layout.pos.row = 84:90, layout.pos.col = 6:20))
print(p.lg.highlow + theme(plot.margin = margin(0.04, 0.04, 0.04, 0.04, "line")), vp = viewport(layout.pos.row = 84:90, layout.pos.col = 26:40))
print(p.lg.highlow + theme(plot.margin = margin(0.04, 0.04, 0.04, 0.04, "line")), vp = viewport(layout.pos.row = 84:90, layout.pos.col = 46:60))

grid.text(x = unit(0.26, "npc"), y = unit(0.66, "npc"), label = "IMNeuS", gp = gpar(fontsize = 10, fontface = "plain", fontfamily = "Arial", lineheight = 0.8), hjust = 0.5 )
grid.text(x = unit(0.54, "npc"), y = unit(0.66, "npc"), label = "TNeuS", gp = gpar(fontsize = 10, fontface = "plain", fontfamily = "Arial", lineheight = 0.8), hjust = 0.5 )
grid.text(x = unit(0.84, "npc"), y = unit(0.66, "npc"), label = "IFNS", gp = gpar(fontsize = 10, fontface = "plain", fontfamily = "Arial", lineheight = 0.8), hjust = 0.5 )

grid.text(x = unit(0.1, "npc"), y = unit(0.97, "npc"), label = "A", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
grid.text(x = unit(0.135, "npc"), y = unit(0.66, "npc"), label = "B", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
grid.text(x = unit(0.41, "npc"), y = unit(0.66, "npc"), label = "C", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
grid.text(x = unit(0.705, "npc"), y = unit(0.66, "npc"), label = "D", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))

grid.brackets(0.3, 0.31, 0.145, 0.31, lwd = 0.7, ticks = 0.5, h = 0.04, type = 2, curvature = 1, col = "grey42")
grid.brackets(0.59, 0.31, 0.435, 0.31, lwd = 0.7, ticks = 0.5, h = 0.04, type = 2, curvature = 1, col = "grey42")
grid.brackets(0.885, 0.31, 0.73, 0.31, lwd = 0.7, ticks = 0.5, h = 0.04, type = 2, curvature = 1, col = "grey42")

dev.off()
```

```{r figureS6}
cairo_pdf(file = paste0(dir.fig, "figureS6.pdf"), width = 4.5, height = 4, family = "Arial Unicode MS")
pushViewport(viewport(layout = grid.layout(nrow = 71, ncol = 62)))

pushViewport(viewport(layout.pos.row = 3:23, layout.pos.col = 3:60))
draw(pht.females, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "top", newpage = FALSE)
upViewport(1)

pushViewport(viewport(layout.pos.row = 26:46, layout.pos.col = 3:60))
draw(pht.males, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "top", newpage = FALSE)
upViewport(1)

pushViewport(viewport(layout.pos.row = 49:69, layout.pos.col = 3:60))
draw(pht.nodiabetes, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "top", newpage = FALSE)
upViewport(1)

grid.text(x = unit(0.04, "npc"), y = unit(0.97, "npc"), label = "A", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
grid.text(x = unit(0.04, "npc"), y = unit(0.64, "npc"), label = "B", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
grid.text(x = unit(0.04, "npc"), y = unit(0.31, "npc"), label = "C", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
dev.off()
```

# Figure 7&S7

```{r stanford}
dir_EHR = "/Users/hongzheng/Google Drive/My Drive/projects/neutrophils/analysis/rds/EHR/"
death <- read.csv(paste0(dir_EHR, "stanford_prop_surv_death.csv"))
cancer <- read.csv(paste0(dir_EHR, "stanford_prop_surv_cancer.csv"))
charlson <- read.csv(paste0(dir_EHR, "stanford_prop_surv_charlson.csv"))
chf <- read.csv(paste0(dir_EHR, "stanford_prop_surv_chf.csv"))
ckd <- read.csv(paste0(dir_EHR, "stanford_prop_surv_ckd.csv"))
copd <- read.csv(paste0(dir_EHR, "stanford_prop_surv_copd.csv"))
dementia <- read.csv(paste0(dir_EHR, "stanford_prop_surv_dementia.csv"))
diabetes <- read.csv(paste0(dir_EHR, "stanford_prop_surv_diabetes.csv"))
liver <- read.csv(paste0(dir_EHR, "stanford_prop_surv_liver.csv"))
pvd <- read.csv(paste0(dir_EHR, "stanford_prop_surv_pvd.csv"))

death_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_death.csv"))
cancer_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_cancer.csv"))
charlson_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_charlson.csv"))
chf_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_chf.csv"))
ckd_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_ckd.csv"))
copd_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_copd.csv"))
dementia_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_dementia.csv"))
diabetes_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_diabetes.csv"))
liver_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_liver.csv"))
pvd_iqr <- read.csv(paste0(dir_EHR, "stanford_prop_iqr_surv_pvd.csv"))

update_columns <- function(df) {
  df$surv <- 1 - df$surv 
  df$lower <- 1 - df$lower 
  df$upper <- 1 - df$upper 
  return(df)
}

cancer <- update_columns(cancer)
charlson <- update_columns(charlson)
chf <- update_columns(chf)
ckd <- update_columns(ckd)
copd <- update_columns(copd)
dementia <- update_columns(dementia)
diabetes <- update_columns(diabetes)
liver <- update_columns(liver)
pvd <- update_columns(pvd)

cancer_iqr <- update_columns(cancer_iqr)
charlson_iqr <- update_columns(charlson_iqr)
chf_iqr <- update_columns(chf_iqr)
ckd_iqr <- update_columns(ckd_iqr)
copd_iqr <- update_columns(copd_iqr)
dementia_iqr <- update_columns(dementia_iqr)
diabetes_iqr <- update_columns(diabetes_iqr)
liver_iqr <- update_columns(liver_iqr)
pvd_iqr <- update_columns(pvd_iqr)

#combine data, remove LOWEST.QUARTILE strata from iqr data
death_combined <- rbind(death, filter(death_iqr, strata == "IQR"))
cancer_combined <- rbind(cancer, filter(cancer_iqr, strata == "IQR"))
charlson_combined <- rbind(charlson, filter(charlson_iqr, strata == "IQR"))
chf_combined <- rbind(chf, filter(chf_iqr, strata == "IQR"))
ckd_combined <- rbind(ckd, filter(ckd_iqr, strata == "IQR"))
copd_combined <- rbind(copd, filter(copd_iqr, strata == "IQR"))
dementia_combined <- rbind(dementia, filter(dementia_iqr, strata == "IQR"))
diabetes_combined <- rbind(diabetes, filter(diabetes_iqr, strata == "IQR"))
liver_combined <- rbind(liver, filter(liver_iqr, strata == "IQR"))
pvd_combined <- rbind(pvd, filter(pvd_iqr, strata == "IQR"))

death_hr = "1.98 (1.62 - 2.42)"
death_p = "p < 0.001"

charlson_hr = "1.18 (1.13 - 1.23)"
charlson_p = "p < 0.001"

chf_hr = "1.40 (1.24 - 1.58)"
chf_p = "p < 0.001"

ckd_hr = "1.34 (1.20 - 1.48)"
ckd_p = "p < 0.001"

copd_hr = "1.10 (1.03 - 1.18)"
copd_p = "p = 0.003"

dementia_hr = "1.60 (1.30 - 1.98)"
dementia_p = "p < 0.001"

diabetes_hr = "1.79 (1.61 - 1.98)"
diabetes_p = "p < 0.001"

liver_hr = "1.11 (0.78 - 1.59)"
liver_p = "p = 0.56"

cancer_hr = "1.06 (0.98 - 1.14)"
cancer_p = "p = 0.17"

pvd_hr = "1.27 (1.13 - 1.43)"
pvd_p = "p < 0.001"

death_iqr_hr = "1.43 (1.20 - 1.69)"
death_iqr_p = "p < 0.001"

charlson_iqr_hr = "1.16 (1.12 - 1.20)"
charlson_iqr_p = "p < 0.001"

chf_iqr_hr = "1.37 (1.24 - 1.52)"
chf_iqr_p = "p < 0.001"

ckd_iqr_hr = "1.26 (1.16 - 1.38)"
ckd_iqr_p = "p < 0.001"

copd_iqr_hr = "1.11 (1.05 - 1.17)"
copd_iqr_p = "p < 0.001"

dementia_iqr_hr = "1.51 (1.27 - 1.79)"
dementia_iqr_p = "p < 0.001"

diabetes_iqr_hr = "1.53 (1.41 - 1.67)"
diabetes_iqr_p = "p < 0.001"

liver_iqr_hr = "1.08 (0.78 - 1.50)"
liver_iqr_p = "p = 0.63"

cancer_iqr_hr = "1.08 (1.01 - 1.15)"
cancer_iqr_p = "p = 0.02"

pvd_iqr_hr = "1.13 (1.03 - 1.24)"
pvd_iqr_p = "p = 0.01"
```

```{r national}
cancer_national <- read.csv(paste0(dir_EHR, "national_prop_surv_cancer.csv"))
charlson_national <- read.csv(paste0(dir_EHR, "national_prop_surv_charlson.csv"))
chf_national <- read.csv(paste0(dir_EHR, "national_prop_surv_chf.csv"))
ckd_national <- read.csv(paste0(dir_EHR, "national_prop_surv_ckd.csv"))
copd_national <- read.csv(paste0(dir_EHR, "national_prop_surv_copd.csv"))
dementia_national <- read.csv(paste0(dir_EHR, "national_prop_surv_dementia.csv"))
diabetes_national <- read.csv(paste0(dir_EHR, "national_prop_surv_diabetes.csv"))
liver_national <- read.csv(paste0(dir_EHR, "national_prop_surv_liver.csv"))
pvd_national <- read.csv(paste0(dir_EHR, "national_prop_surv_pvd.csv"))

cancer_national <- update_columns(cancer_national)
charlson_national <- update_columns(charlson_national)
chf_national <- update_columns(chf_national)
ckd_national <- update_columns(ckd_national)
copd_national <- update_columns(copd_national)
dementia_national <- update_columns(dementia_national)
diabetes_national <- update_columns(diabetes_national)
liver_national <- update_columns(liver_national)
pvd_national <- update_columns(pvd_national)

charlson_hr_national = "1.27 (1.21 - 1.33)"
charlson_p_national = "p < 0.001"

chf_hr_national = "1.57 (1.33 - 1.86)"
chf_p_national = "p < 0.001"

ckd_hr_national = "1.39 (1.22 - 1.58)"
ckd_p_national = "p < 0.001"

copd_hr_national = "1.22 (1.14 - 1.30)"
copd_p_national = "p < 0.001"

dementia_hr_national = "1.96 (1.50 - 2.56)"
dementia_p_national = "p < 0.001"

diabetes_hr_national = "1.45 (1.32 - 1.59)"
diabetes_p_national = "p < 0.001"

liver_hr_national = "1.40 (0.79 - 2.49)"
liver_p_national = "p = 0.25"

cancer_hr_national = "1.25 (1.12 - 1.41)"
cancer_p_national = "p < 0.001"

pvd_hr_national = "1.34 (1.21 - 1.48)"
pvd_p_national = "p < 0.001"
```

```{r plots}
col_quartile = c("HIGHEST.QUARTILE" = "#FF7F0E", "IQR" = "goldenrod2", "LOWEST.QUARTILE" = "#214DC8")

EHRplotStanford = function(dt, ptitle, laby, labelx, labely, hjust, vjust, hr1, p1, hr2, p2){
  dt$time = dt$time/ 365
  dt$surv = dt$surv * 100
  dt$lower = dt$lower * 100
  dt$upper = dt$upper * 100
  p = ggplot(dt, aes_string(x = "time", y = "surv", color = "strata", fill = "strata")) +
    geom_line()+
    geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.3, color = NA) + 
    labs(title = ptitle, x = "time (years)", y = laby) +
    scale_x_continuous(breaks = seq(0,10,2), limits = c(0,10)) +
    scale_y_continuous(breaks = pretty_breaks()) +
    scale_color_manual(values = col_quartile) +
    scale_fill_manual(values = col_quartile) +
    theme_bw() + mytheme +
    theme(legend.position = "none") +
    annotate("text", x = labelx, y = labely, hjust = hjust, vjust = vjust, lineheight = 0.9, color = textcolor, size = 3, label = paste0("HR(H): ", hr1, "\n", p1, "\n", "HR(I): ", hr2, "\n", p2))
  return(p)
}

EHRplotStanford2 = function(dt, ptitle, laby, labelx, labely, labely2, hjust, vjust, hr1, p1, hr2, p2, textcolor1 = textcolor, textcolor2 = textcolor){
  dt$time = dt$time/ 365
  dt$surv = dt$surv * 100
  dt$lower = dt$lower * 100
  dt$upper = dt$upper * 100
  p = ggplot(dt, aes_string(x = "time", y = "surv", color = "strata", fill = "strata")) +
    geom_line()+
    coord_cartesian(clip = "off") +
    geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.3, color = NA) + 
    labs(title = ptitle, x = "time (years)", y = laby) +
    scale_x_continuous(breaks = seq(0,10,2), limits = c(0,10)) +
    scale_y_continuous(breaks = pretty_breaks()) +
    scale_color_manual(values = col_quartile) +
    scale_fill_manual(values = col_quartile) +
    theme_bw() + mytheme +
    theme(legend.position = "none") +
    annotation_custom(grob = textGrob(label = paste0("HR(H): ", hr1, "\n", p1), x = unit(labelx, "npc"), y = unit(labely, "npc"), just = c(hjust, vjust), gp = gpar(col = textcolor1, fontsize = 3 * .pt, lineheight = 0.8))) +
    annotation_custom(grob = textGrob(label = paste0("HR(I): ", hr2, "\n", p2), x = unit(labelx, "npc"), y = unit(labely2, "npc"), just = c(hjust, vjust), gp = gpar(col = textcolor2, fontsize = 3 * .pt, lineheight = 0.8))) 
  
  
  return(p)
}

death_plot = EHRplotStanford(death_combined, "survival (Stanford)", "survival rate (%)", labelx = 0.1, labely = 98,  vjust = 1, hjust = 0, death_hr, death_p, death_iqr_hr, death_iqr_p) + scale_y_continuous(breaks = seq(0,100,1)) 
charlson_plot = EHRplotStanford(charlson_combined, "Charlson comorbidities (Stanford)", "comorbidities (%)", labelx = 0, labely = Inf, vjust = 1, hjust = 0, charlson_hr, charlson_p, charlson_iqr_hr, charlson_iqr_p)

chf_plot = EHRplotStanford(chf_combined, "congestive heart failure", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, chf_hr, chf_p, chf_iqr_hr, chf_iqr_p)
ckd_plot = EHRplotStanford(ckd_combined, "chronic kidney disease", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, ckd_hr, ckd_p, ckd_iqr_hr, ckd_iqr_p)
copd_plot = EHRplotStanford(copd_combined, "COPD", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, copd_hr, copd_p, copd_iqr_hr, copd_iqr_p)
cancer_plot = EHRplotStanford(cancer_combined, "cancer", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, cancer_hr, cancer_p, cancer_iqr_hr, cancer_iqr_p)
dementia_plot = EHRplotStanford(dementia_combined, "dementia", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, dementia_hr, dementia_p, dementia_iqr_hr, dementia_iqr_p) 
diabetes_plot = EHRplotStanford(diabetes_combined, "diabetes", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, diabetes_hr, diabetes_p, diabetes_iqr_hr, diabetes_iqr_p)
liver_plot = EHRplotStanford(liver_combined, "liver disease", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, liver_hr, liver_p, liver_iqr_hr, liver_iqr_p)
pvd_plot = EHRplotStanford(pvd_combined, "peripheral vascular diseases", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, pvd_hr, pvd_p, pvd_iqr_hr, pvd_iqr_p)

chf_plot = EHRplotStanford2(chf_combined, "congestive heart failure", "percentage", labelx = 0.01, labely = 1, labely2 = 0.83, vjust = 1, hjust = 0, chf_hr, chf_p, chf_iqr_hr, chf_iqr_p)
ckd_plot = EHRplotStanford2(ckd_combined, "chronic kidney disease", "percentage", labelx = 0.01, labely = 1, labely2 = 0.83, vjust = 1, hjust = 0, ckd_hr, ckd_p, ckd_iqr_hr, ckd_iqr_p)
copd_plot = EHRplotStanford2(copd_combined, "COPD", "percentage", labelx = 0.01, labely = 1, labely2 = 0.83, vjust = 1, hjust = 0, copd_hr, copd_p, copd_iqr_hr, copd_iqr_p)
cancer_plot = EHRplotStanford2(cancer_combined, "cancer", "percentage", labelx = 0.01, labely = 1, labely2 = 0.83, vjust = 1, hjust = 0, cancer_hr, cancer_p, cancer_iqr_hr, cancer_iqr_p, textcolor1 = "grey67")
dementia_plot = EHRplotStanford2(dementia_combined, "dementia", "percentage", labelx = 0.01, labely = 1, labely2 = 0.83, vjust = 1, hjust = 0, dementia_hr, dementia_p, dementia_iqr_hr, dementia_iqr_p) 
diabetes_plot = EHRplotStanford2(diabetes_combined, "diabetes", "percentage", labelx = 0.01, labely = 1, labely2 = 0.83, vjust = 1, hjust = 0, diabetes_hr, diabetes_p, diabetes_iqr_hr, diabetes_iqr_p)
liver_plot = EHRplotStanford2(liver_combined, "liver disease", "percentage", labelx = 0.01, labely = 1, labely2 = 0.83, vjust = 1, hjust = 0, liver_hr, liver_p, liver_iqr_hr, liver_iqr_p, textcolor1 = "grey67", textcolor2 = "grey67")
pvd_plot = EHRplotStanford2(pvd_combined, "peripheral vascular diseases", "percentage", labelx = 0.01, labely = 1, labely2 = 0.83, vjust = 1, hjust = 0, pvd_hr, pvd_p, pvd_iqr_hr, pvd_iqr_p)

EHRplotNational = function(dt, ptitle, laby, labelx, labely, vjust, hjust, hr1, p1, hr2, p2, textcolor = textcolor){
  dt$time = dt$time/ 365
  dt$surv = dt$surv * 100
  dt$lower = dt$lower * 100
  dt$upper = dt$upper * 100
  p = ggplot(dt, aes_string(x = "time", y = "surv", color = "strata", fill = "strata")) +
    geom_line()+
    geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.3, color = NA) + 
    labs(title = ptitle, x = "time (years)", y = laby) +
    scale_x_continuous(breaks = seq(0,5,1), limits = c(0,5))+
    scale_y_continuous(breaks = pretty_breaks()) +
    scale_color_manual(values = col_quartile) +
    scale_fill_manual(values = col_quartile) +
    theme_bw() + mytheme +
    theme(legend.position = "none") +
    annotate("text", x = labelx, y = labely, hjust = hjust, vjust = vjust, lineheight = 0.9, color = textcolor, size = 3,
             label = paste0("HR(H): ", hr1, "\n", p1))
  
  return(p)
}

charlson_national_plot = EHRplotNational(charlson_national, "Charlson comorbidities (National)", "comorbidities (%)", labelx = 0, labely = Inf, vjust = 1, hjust = 0, charlson_hr_national, charlson_p_national)
chf_national_plot = EHRplotNational(chf_national, "congestive heart failure", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, chf_hr_national, chf_p_national)

ckd_national_plot = EHRplotNational(ckd_national, "chronic kidney disease", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, ckd_hr_national, ckd_p_national)
copd_national_plot = EHRplotNational(copd_national, "COPD", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, copd_hr_national, copd_p_national)
cancer_national_plot = EHRplotNational(cancer_national, "cancer", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, cancer_hr_national, cancer_p_national)
dementia_national_plot = EHRplotNational(dementia_national, "dementia", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, dementia_hr_national, dementia_p_national)
diabetes_national_plot = EHRplotNational(diabetes_national, "diabetes", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, diabetes_hr_national, diabetes_p_national)
liver_national_plot = EHRplotNational(liver_national, "liver disease", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, liver_hr_national, liver_p_national, textcolor = "grey67")
pvd_national_plot = EHRplotNational(pvd_national, "peripheral vascular diseases", "percentage", labelx = 0, labely = Inf, vjust = 1, hjust = 0, pvd_hr_national, pvd_p_national)
```

```{r figure7}
df.quartile = data.table(
  subset = paste0("- ", c("highest quartile of ANCs (H)", "IQR of ANCs (I)", "lowest quartile of ANCs")),
  color = col_quartile,
  y = 1,
  x = c(3,2.15,1)
)

p.quartile  <- ggplot(df.quartile , aes(x = x, y = y, label = subset, color = color)) +
  geom_text(angle = 0, size = 3, hjust = 0) + 
  scale_color_identity() +           
  theme_void() +
  coord_cartesian(clip = "off") 

cairo_pdf(file = paste0(dir.fig, "figure7.pdf"), width = 8.2, height = 7)
pushViewport(viewport(layout = grid.layout(nrow = 86, ncol = 86)))

print(charlson_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 3:31, layout.pos.col = 2:29))
print(charlson_national_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 3:31, layout.pos.col = 30:57))
print(death_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 3:31, layout.pos.col = 58:85))

print(p.quartile + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 32:34, layout.pos.col = 20:50))

print(diabetes_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 36:59, layout.pos.col = 2:22))
print(dementia_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 36:59, layout.pos.col = 23:43))
print(chf_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 36:59, layout.pos.col = 44:64))
print(ckd_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 36:59, layout.pos.col = 65:85))

print(copd_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 61:84, layout.pos.col = 23:43))
print(cancer_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 61:84, layout.pos.col = 44:64))
print(liver_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 61:84, layout.pos.col = 65:85))
print(pvd_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 61:84, layout.pos.col = 2:22))

grid.text(x = unit(0.03, "npc"), y = unit(0.98, "npc"), label = "A", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
grid.text(x = unit(0.36, "npc"), y = unit(0.98, "npc"), label = "C", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
grid.text(x = unit(0.69, "npc"), y = unit(0.98, "npc"), label = "D", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))

grid.text(x = unit(0.03, "npc"), y = unit(0.595, "npc"), label = "B", gp = gpar(fontsize = 10, fontface = "bold", fontfamily = "Arial"))
dev.off()
```

```{r figureS7}
df.quartile = data.table(
  subset = paste0("- ", c("highest quartile of ANCs (H)", "lowest quartile of ANCs")),
  color = col_quartile[c(1,3)],
  y = 1,
  x = c(2,1)
)

p.quartile  <- ggplot(df.quartile , aes(x = x, y = y, label = subset, color = color)) +
  geom_text(angle = 0, size = 3, hjust = 0) + 
  scale_color_identity() +           
  theme_void() +
  coord_cartesian(clip = "off") 

cairo_pdf(file = paste0(dir.fig, "figureS7.pdf"), width = 8, height = 4)
pushViewport(viewport(layout = grid.layout(nrow = 54, ncol = 86)))

print(diabetes_national_plot  + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 2:26, layout.pos.col = 2:22))
print(dementia_national_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 2:26, layout.pos.col = 23:43))
print(chf_national_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 2:26, layout.pos.col = 44:64))
print(ckd_national_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 2:26, layout.pos.col = 65:85))

print(copd_national_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 28:51, layout.pos.col = 23:43))
print(cancer_national_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 28:51, layout.pos.col = 44:64))
print(liver_national_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 28:51, layout.pos.col = 65:85))

print(pvd_national_plot + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 28:51, layout.pos.col = 2:22))

print(p.quartile + theme(plot.margin = margin(0.04, 0.1, 0.04, 0.1, "line")), vp = viewport(layout.pos.row = 53:54, layout.pos.col = 27:43))

dev.off()
```

